// @syncfusion/ej2/ej2-resource/scripts/ej2-kanban/blazor/kanban/constant.js
export var ROOT_CLASS = 'e-kanban';
export var DEVICE_CLASS = 'e-device';
export var ICON_CLASS = 'e-icons';
export var CARD_CLASS = 'e-card';
export var HEADER_CLASS = 'e-kanban-header';
export var CONTENT_CLASS = 'e-kanban-content';
export var HEADER_ICON_CLASS = 'e-header-icon';
export var CLONED_CARD_CLASS = 'e-cloned-card';
export var TARGET_MULTI_CLONE_CLASS = 'e-target-multi-clone';
export var DROPPED_CLONE_CLASS = 'e-target-dropped-clone';
export var CARD_SELECTION_CLASS = 'e-selection';
export var CONTENT_ROW_CLASS = 'e-content-row';
export var DRAGGED_CLONE_CLASS = 'e-target-dragged-clone';
export var DRAGGED_CARD_CLASS = 'e-kanban-dragged-card';
export var SWIMLANE_HEADER_CLASS = 'e-swimlane-header';
export var SWIMLANE_ROW_CLASS = 'e-swimlane-row';
export var SWIMLANE_ROW_EXPAND_CLASS = 'e-swimlane-row-expand';
export var SWIMLANE_ROW_COLLAPSE_CLASS = 'e-swimlane-row-collapse';
export var CONTENT_CELLS_CLASS = 'e-content-cells';
export var CARD_CONTAINER_CLASS = 'e-card-container';
export var COLLAPSED_CLASS = 'e-collapsed';
export var DROPPING_CLASS = 'e-dropping';
export var MULTI_COLUMN_KEY_CLASS = 'e-column-key';
export var MULTI_ACTIVE_CLASS = 'e-multi-active';
export var SHOW_ADD_BUTTON = 'e-show-add-button';
export var MULTI_CARD_CONTAINER_CLASS = 'e-multi-card-container';
export var LIMITS_CLASS = 'e-limits';
export var SHOW_ADD_FOCUS = 'e-show-add-focus';
export var POPUP_CONTAINER_CLASS = 'e-mobile-popup-container';
export var POPUP_HEADER_CLASS = 'e-popup-header';
export var POPUP_CONTENT_CLASS = 'e-popup-content';
export var CLOSE_CLASS = 'e-close';
export var CLOSE_ICON_CLASS = 'e-close-icon';
export var cardSpace = 16;
export var bottomSpace = 25;
export var DRAGGABLE_CLASS = 'e-draggable';
export var DROPPABLE_CLASS = 'e-droppable';
export var DRAG_CLASS = 'e-drag';
export var DROP_CLASS = 'e-drop';
export var DISABLED_CLASS = 'e-disabled';
export var CLS_RTL = 'e-rtl';
export var TOOLBAR_SWIMLANE_NAME_CLASS = 'e-toolbar-swimlane-name';
export var SWIMLANE_OVERLAY_CLASS = 'e-swimlane-overlay';
export var SWIMLANE_CONTENT_CLASS = 'e-swimlane-content';
export var SWIMLANE_RESOURCE_CLASS = 'e-swimlane-resource';
export var SWIMLANE_TREE_CLASS = 'e-swimlane-tree';
export var WINDOW_WIDTH = 'WindowWidth';
export var TOOLTIP_TEXT_CLASS = 'e-tooltip-text';
export var POPUP_WRAPPER_CLASS = 'e-mobile-popup-wrapper';
