// @syncfusion/ej2/ej2-resource/scripts/ej2-charts/common/model/constants.js
export var loaded = 'loaded';
export var load = 'load';
export var animationComplete = 'animationComplete';
export var legendRender = 'legendRender';
export var textRender = 'textRender';
export var pointRender = 'pointRender';
export var seriesRender = 'seriesRender';
export var axisLabelRender = 'axisLabelRender';
export var axisRangeCalculated = 'axisRangeCalculated';
export var axisMultiLabelRender = 'axisMultiLabelRender';
export var tooltipRender = 'tooltipRender';
export var chartMouseMove = 'chartMouseMove';
export var chartMouseClick = 'chartMouseClick';
export var pointClick = 'pointClick';
export var pointMove = 'pointMove';
export var chartMouseLeave = 'chartMouseLeave';
export var chartMouseDown = 'chartMouseDown';
export var chartMouseUp = 'chartMouseUp';
export var zoomComplete = 'zoomComplete';
export var dragComplete = 'dragComplete';
export var resized = 'resized';
export var beforePrint = 'beforePrint';
export var annotationRender = 'annotationRender';
export var scrollStart = 'scrollStart';
export var scrollEnd = 'scrollEnd';
export var scrollChanged = 'scrollChanged';
export var stockEventRender = 'stockEventRender';
