// @syncfusion/ej2/ej2-resource/scripts/ej2-data-grid/grid/actions/show-hide.js
import { isNullOrUndefined } from '@syncfusion/ej2-base';
import { iterateArrayOrObject, isActionPrevent } from '../base/util';
import * as events from '../base/constant';
var ShowHide = (function () {
    function ShowHide(parent) {
        this.parent = parent;
    }
    ShowHide.prototype.show = function (columnName, showBy) {
        var keys = this.getToggleFields(columnName);
        var columns = this.getColumns(keys, showBy);
        this.parent.notify(events.tooltipDestroy, { module: 'edit' });
        columns.forEach(function (value) {
            value.visible = true;
        });
        this.setVisible(columns);
    };
    ShowHide.prototype.hide = function (columnName, hideBy) {
        var keys = this.getToggleFields(columnName);
        var columns = this.getColumns(keys, hideBy);
        this.parent.notify(events.tooltipDestroy, { module: 'edit' });
        columns.forEach(function (value) {
            value.visible = false;
        });
        this.setVisible(columns);
    };
    ShowHide.prototype.getToggleFields = function (key) {
        var finalized = [];
        if (typeof key === 'string') {
            finalized = [key];
        }
        else {
            finalized = key;
        }
        return finalized;
    };
    ShowHide.prototype.getColumns = function (keys, getKeyBy) {
        var _this = this;
        var columns = iterateArrayOrObject(keys, function (key, index) {
            return iterateArrayOrObject(_this.parent.getColumns(), function (item, index) {
                if (item[getKeyBy] === key) {
                    return item;
                }
                return undefined;
            })[0];
        });
        return columns;
    };
    ShowHide.prototype.setVisible = function (columns) {
        if (isActionPrevent(this.parent)) {
            this.parent.notify(events.preventBatch, {
                instance: this, handler: this.setVisible,
                arg1: columns
            });
            return;
        }
        columns = isNullOrUndefined(columns) ? this.parent.getColumns() : columns;
        if (this.parent.allowSelection && this.parent.getSelectedRecords().length) {
            this.parent.clearSelection();
        }
        this.parent.notify(events.columnVisibilityChanged, columns);
        if (this.parent.columnQueryMode !== 'All') {
            this.parent.refresh();
        }
    };
    return ShowHide;
}());
export { ShowHide };
