// @syncfusion/ej2/ej2-resource/scripts/ej2-smithchart/smithchart/legend/legend.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Property, Complex, ChildProperty } from '@syncfusion/ej2-base';
import { Font } from '../utils/utils';
import { Theme } from '../model/theme';
var LegendTitle = (function (_super) {
    __extends(LegendTitle, _super);
    function LegendTitle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(true)
    ], LegendTitle.prototype, "visible", void 0);
    __decorate([
        Property('')
    ], LegendTitle.prototype, "text", void 0);
    __decorate([
        Property('')
    ], LegendTitle.prototype, "description", void 0);
    __decorate([
        Property('center')
    ], LegendTitle.prototype, "textAlignment", void 0);
    __decorate([
        Complex(Theme.legendLabelFont, Font)
    ], LegendTitle.prototype, "font", void 0);
    return LegendTitle;
}(ChildProperty));
export { LegendTitle };
var LegendLocation = (function (_super) {
    __extends(LegendLocation, _super);
    function LegendLocation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(0)
    ], LegendLocation.prototype, "x", void 0);
    __decorate([
        Property(0)
    ], LegendLocation.prototype, "y", void 0);
    return LegendLocation;
}(ChildProperty));
export { LegendLocation };
var LegendItemStyleBorder = (function (_super) {
    __extends(LegendItemStyleBorder, _super);
    function LegendItemStyleBorder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(1)
    ], LegendItemStyleBorder.prototype, "width", void 0);
    __decorate([
        Property(null)
    ], LegendItemStyleBorder.prototype, "color", void 0);
    return LegendItemStyleBorder;
}(ChildProperty));
export { LegendItemStyleBorder };
var LegendItemStyle = (function (_super) {
    __extends(LegendItemStyle, _super);
    function LegendItemStyle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(9)
    ], LegendItemStyle.prototype, "width", void 0);
    __decorate([
        Property(9)
    ], LegendItemStyle.prototype, "height", void 0);
    __decorate([
        Complex({}, LegendItemStyleBorder)
    ], LegendItemStyle.prototype, "border", void 0);
    return LegendItemStyle;
}(ChildProperty));
export { LegendItemStyle };
var LegendSize = (function (_super) {
    __extends(LegendSize, _super);
    function LegendSize() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(null)
    ], LegendSize.prototype, "width", void 0);
    __decorate([
        Property(null)
    ], LegendSize.prototype, "height", void 0);
    return LegendSize;
}(ChildProperty));
export { LegendSize };
var LegendBorder = (function (_super) {
    __extends(LegendBorder, _super);
    function LegendBorder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(1)
    ], LegendBorder.prototype, "width", void 0);
    __decorate([
        Property(null)
    ], LegendBorder.prototype, "color", void 0);
    return LegendBorder;
}(ChildProperty));
export { LegendBorder };
var LegendSettings = (function (_super) {
    __extends(LegendSettings, _super);
    function LegendSettings() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(false)
    ], LegendSettings.prototype, "visible", void 0);
    __decorate([
        Property('bottom')
    ], LegendSettings.prototype, "position", void 0);
    __decorate([
        Property('center')
    ], LegendSettings.prototype, "alignment", void 0);
    __decorate([
        Property('circle')
    ], LegendSettings.prototype, "shape", void 0);
    __decorate([
        Property(null)
    ], LegendSettings.prototype, "rowCount", void 0);
    __decorate([
        Property(null)
    ], LegendSettings.prototype, "columnCount", void 0);
    __decorate([
        Property(8)
    ], LegendSettings.prototype, "itemPadding", void 0);
    __decorate([
        Property('')
    ], LegendSettings.prototype, "description", void 0);
    __decorate([
        Property(true)
    ], LegendSettings.prototype, "toggleVisibility", void 0);
    __decorate([
        Complex({}, LegendTitle)
    ], LegendSettings.prototype, "title", void 0);
    __decorate([
        Complex({}, LegendLocation)
    ], LegendSettings.prototype, "location", void 0);
    __decorate([
        Complex({}, LegendSize)
    ], LegendSettings.prototype, "size", void 0);
    __decorate([
        Complex({}, LegendItemStyle)
    ], LegendSettings.prototype, "itemStyle", void 0);
    __decorate([
        Complex({}, LegendBorder)
    ], LegendSettings.prototype, "border", void 0);
    __decorate([
        Complex(Theme.legendLabelFont, Font)
    ], LegendSettings.prototype, "font", void 0);
    return LegendSettings;
}(ChildProperty));
export { LegendSettings };
