// @syncfusion/ej2/ej2-resource/scripts/ej2-inputs/blazor/sf-colorpicker.js
import { Browser, EventHandler, isNullOrUndefined } from '@syncfusion/ej2-base';
import { getScrollableParent } from '@syncfusion/ej2-popups';
var HSV_CONTAINER = '.e-hsv-container';
var COLOR_PICKER = '.e-split-colorpicker';
var COLOR_PALETTE = 'e-color-palette';
var PALETTE = '.e-palette';
var HSV_COLOR = '.e-hsv-color';
var SET_OFFSET = 'SetOffset';
var DROPDOWN_BTN = '.e-dropdown-btn';
var HANDLER = '.e-handler';
var MODEL = '.e-colorpicker.e-modal';
var EMPTY = '';
var CONTAINER = '.e-container';
var SPLIT_BUTTON = '.sf-colorpicker';
var INPUTS = '.e-selected-value';
var CONTROL_BTN = '.e-switch-ctrl-btn';
var PIXEL = 'px';
var SCROLL = 'scroll';
var DEFAULT = '100%';
var SfColorPicker = (function () {
    function SfColorPicker(element, dotnetRef, inline) {
        this.element = element;
        this.dotnetRef = dotnetRef;
        this.element.blazor__instance = this;
        if (inline) {
            this.addScrollEvents(true);
            this.setPaletteWidth(this.element.querySelector(CONTAINER), false);
        }
    }
    SfColorPicker.prototype.getOffset = function (element) {
        var colorPicker = element.querySelector(HSV_CONTAINER);
        if (!colorPicker) {
            return { IsDevice: Browser.isDevice };
        }
        var offset = colorPicker.getBoundingClientRect();
        var color = colorPicker.querySelector(HSV_COLOR);
        var handler = colorPicker.querySelector(HANDLER);
        if (handler) {
            handler.focus();
        }
        return { Left: offset.left, Top: offset.top, Width: offset.width, Height: offset.height, Right: offset.right, ClientLeft: color.offsetLeft, ClientTop: color.offsetTop, ClientWidth: color.offsetWidth, ClientHeight: color.offsetHeight,
            IsDevice: Browser.isDevice };
    };
    SfColorPicker.prototype.setOffset = function (element, zIndex) {
        var offset = this.getOffset(element);
        if (zIndex && !Browser.isDevice) {
            this.setZIndex(this.element, zIndex, COLOR_PICKER);
            this.setZIndex(this.element, zIndex, DROPDOWN_BTN);
        }
        if (offset) {
            this.dotnetRef.invokeMethodAsync(SET_OFFSET, offset);
        }
    };
    SfColorPicker.prototype.setZIndex = function (element, zIndex, cls) {
        var btnEle = element.querySelector(cls);
        if (btnEle) {
            btnEle.style.zIndex = zIndex + 1 + EMPTY;
        }
    };
    SfColorPicker.prototype.setPaletteWidth = function (container, modeSwitch, zIndex) {
        if (Browser.isDevice && !isNullOrUndefined(zIndex)) {
            var model = container.querySelector(MODEL);
            if (model) {
                model.style.zIndex = (zIndex - 1) + EMPTY;
                document.body.insertBefore(model, container.parentElement);
                model.style.display = EMPTY;
            }
        }
        if (container.querySelector(HSV_CONTAINER)) {
            container.style.width = EMPTY;
        }
        else {
            var width = parseInt(getComputedStyle(container).borderBottomWidth, 10);
            container.style.width = container.children[0].offsetWidth + width + width + PIXEL;
            var containers = container.querySelector(INPUTS);
            if (containers) {
                containers.style.width = DEFAULT;
                containers = container.querySelector(CONTROL_BTN);
                if (containers) {
                    containers.style.width = DEFAULT;
                }
            }
        }
        if (modeSwitch) {
            var colorPickerBtn = this.element.querySelector(SPLIT_BUTTON);
            if (colorPickerBtn && !isNullOrUndefined(colorPickerBtn.blazor__instance)) {
                colorPickerBtn.blazor__instance.setPosition(true);
            }
        }
    };
    SfColorPicker.prototype.scrollHandler = function (e) {
        if (!this.element.parentElement) {
            EventHandler.remove(e.target, SCROLL, this.scrollHandler);
            return;
        }
        this.setOffset(this.element);
    };
    SfColorPicker.prototype.addScrollEvents = function (add) {
        var elements = getScrollableParent(this.element);
        for (var _i = 0, elements_1 = elements; _i < elements_1.length; _i++) {
            var element = elements_1[_i];
            add ? EventHandler.add(element, SCROLL, this.scrollHandler, this) :
                EventHandler.remove(element, SCROLL, this.scrollHandler);
        }
    };
    SfColorPicker.prototype.focusIn = function (element, inline) {
        if (inline) {
            var container = element.querySelector(CONTAINER);
            if (container) {
                if (container.classList.contains(COLOR_PALETTE)) {
                    this.setFocus(PALETTE);
                }
                else {
                    this.setFocus(HANDLER);
                }
            }
        }
        else {
            this.setFocus(COLOR_PICKER);
        }
    };
    SfColorPicker.prototype.setFocus = function (cls) {
        var btn = this.element.querySelector(cls);
        if (btn) {
            btn.focus();
        }
    };
    return SfColorPicker;
}());
var ColorPicker = {
    initialize: function (element, dotnetRef, inline) {
        if (!isNullOrUndefined(element)) {
            new SfColorPicker(element, dotnetRef, inline);
            return element.blazor__instance.getOffset(element);
        }
        return null;
    },
    getOffset: function (element, container) {
        if (!isNullOrUndefined(element) && !isNullOrUndefined(element.blazor__instance) && !isNullOrUndefined(container)) {
            element.blazor__instance.setPaletteWidth(container, true);
            return element.blazor__instance.getOffset(container);
        }
        return null;
    },
    focusIn: function (element, inline) {
        if (!isNullOrUndefined(element)) {
            element.blazor__instance.focusIn(element, inline);
        }
    },
    destroy: function (element) {
        if (!isNullOrUndefined(element)) {
            element.blazor__instance.addScrollEvents(false);
        }
    }
};
export default ColorPicker;
