// @syncfusion/ej2/ej2-resource/scripts/ej2-inputs/blazor/sf-numerictextbox.js
import { Browser, EventHandler } from '@syncfusion/ej2-base';
var INTREGEXP = new RegExp('^(-)?(\\d*)$');
var ENTER = 13;
var ARROW_UP = 38;
var ARROW_DOWN = 40;
var BACK_SPACE = 8;
var WHEEL_DELTA = 120;
var DELTA = 3;
var LEFT_BUTTON = 0;
var RIGHT_BUTTON = 3;
var MOBILE_INTERVEL_TIME = 600;
var INTERVEL_TIME = 30;
var TIMEOUT = 150;
var MOUSE_BUTTON = 2;
var IE_VERSION = '11.0';
var INCREMENT = 'increment';
var DECREMENT = 'decrement';
var MOUSE_MOVE = 'mouseleave';
var ROOT = 'e-input-group-icon';
var MOUSE_UP = 'mouseup';
var MOUSE_WHEEL = 'mousewheel DOMMouseScroll';
var SERVER_ACTION = 'ServerAction';
var SERVER_VALUE_UPDATE = 'ServerupdateValue';
var SPIN_UP = 'e-spin-up';
var SPIN_DOWN = 'e-spin-down';
var DECIMAL = 'decimal';
var FOCUS = 'focus';
var BLUR = 'blur';
var KEY_PRESS = 'keypress';
var KEY_DOWN = 'keydown';
var PASTE = 'paste';
var INVOKE_PASTE = 'InvokePasteHandler';
var SfNumericTextBox = (function () {
    function SfNumericTextBox(wrapperElement, element, dotnetRef, options) {
        this.wrapperElement = wrapperElement;
        this.element = element;
        this.options = options;
        this.element.blazor__instance = this;
        this.dotNetRef = dotnetRef;
    }
    SfNumericTextBox.prototype.initialize = function () {
        this.spinButtonEvents();
        EventHandler.add(this.element, FOCUS, this.focusHandler, this);
        EventHandler.add(this.element, BLUR, this.focusOutHandler, this);
        EventHandler.add(this.element, KEY_PRESS, this.keyPressHandler, this);
        EventHandler.add(this.element, KEY_DOWN, this.keyDownHandler, this);
        EventHandler.add(this.element, PASTE, this.pasteHandler, this);
    };
    SfNumericTextBox.prototype.keyPressHandler = function (event) {
        if (!this.options.enabled || this.options.readonly) {
            return true;
        }
        var action = event.keyCode;
        if (event.which === LEFT_BUTTON || event.metaKey || event.ctrlKey || action === BACK_SPACE || action === ENTER) {
            return true;
        }
        var currentChar = String.fromCharCode(event.which);
        var text = this.element.value;
        var inputElement = this.element;
        text = text.substring(0, inputElement.selectionStart) + currentChar + text.substring(inputElement.selectionEnd);
        if (!this.numericRegex().test(text)) {
            event.preventDefault();
            event.stopPropagation();
            return false;
        }
        else {
            return true;
        }
    };
    ;
    SfNumericTextBox.prototype.pasteHandler = function (event) {
        if (!(!this.options.enabled || this.options.readonly)) {
            var pasteValue = event.clipboardData.getData('text/plain');
            if (!this.numericRegex().test(pasteValue)) {
                event.preventDefault();
            }
            else {
                this.dotNetRef.invokeMethodAsync('InvokePasteHandler', pasteValue);
            }
        }
    };
    SfNumericTextBox.prototype.keyDownHandler = function (event) {
        if (!this.options.readonly) {
            if (event.keyCode === ARROW_UP) {
                event.preventDefault();
                this.dotNetRef.invokeMethodAsync(SERVER_ACTION, INCREMENT, event);
            }
            else if (event.keyCode === ARROW_DOWN) {
                event.preventDefault();
                this.dotNetRef.invokeMethodAsync(SERVER_ACTION, DECREMENT, event);
            }
        }
    };
    ;
    SfNumericTextBox.prototype.numericRegex = function () {
        var decimalSeparator = this.options.decimalSeparator;
        var fractionRule = '*';
        if (decimalSeparator === '.') {
            decimalSeparator = '\\' + decimalSeparator;
        }
        if (this.options.decimals === 0 && this.options.validateDecimalOnType) {
            return INTREGEXP;
        }
        if (this.options.decimals && this.options.validateDecimalOnType) {
            fractionRule = '{0,' + this.options.decimals + '}';
        }
        return new RegExp('^(-)?(((\\d+(' + decimalSeparator + '\\d' + fractionRule +
            ')?)|(' + decimalSeparator + '\\d' + fractionRule + ')))?$');
    };
    ;
    SfNumericTextBox.prototype.mouseWheel = function (event) {
        event.preventDefault();
        var delta;
        var rawEvent = event;
        if (rawEvent.wheelDelta) {
            delta = rawEvent.wheelDelta / WHEEL_DELTA;
        }
        else if (rawEvent.detail) {
            delta = -rawEvent.detail / DELTA;
        }
        if (delta > 0) {
            this.dotNetRef.invokeMethodAsync(SERVER_ACTION, INCREMENT, event);
        }
        else if (delta < 0) {
            this.dotNetRef.invokeMethodAsync(SERVER_ACTION, DECREMENT, event);
        }
    };
    SfNumericTextBox.prototype.focusHandler = function (event) {
        this.isFocused = true;
        if (!(!this.options.enabled || this.options.readonly)) {
            if (!Browser.isDevice) {
                EventHandler.add(this.element, MOUSE_WHEEL, this.mouseWheel, this);
            }
        }
    };
    SfNumericTextBox.prototype.focusOutHandler = function (event) {
        this.isFocused = false;
        event.preventDefault();
        if (!Browser.isDevice) {
            EventHandler.remove(this.element, MOUSE_WHEEL, this.mouseWheel);
        }
    };
    SfNumericTextBox.prototype.mouseDownOnSpinner = function (event) {
        var _this = this;
        if (this.isFocused) {
            this.isPrevFocused = true;
            event.preventDefault();
        }
        var target = event.currentTarget;
        var action = (target.classList.contains(SPIN_UP)) ? INCREMENT : DECREMENT;
        EventHandler.add(target, MOUSE_MOVE, this.mouseUpClick, this);
        this.timeOut = setInterval(function () {
            _this.isCalled = true;
            _this.dotNetRef.invokeMethodAsync(SERVER_ACTION, action, event);
        }, TIMEOUT);
        EventHandler.add(document, MOUSE_UP, this.mouseUpClick, this);
    };
    SfNumericTextBox.prototype.mouseUpOnSpinner = function (event) {
        if (this.isPrevFocused) {
            this.element.focus();
            if (!Browser.isDevice) {
                this.isPrevFocused = false;
            }
        }
        if (!Browser.isDevice) {
            event.preventDefault();
        }
        if (!this.getElementData(event)) {
            return;
        }
        var target = event.currentTarget;
        var action = (target.classList.contains(SPIN_UP)) ? INCREMENT : DECREMENT;
        EventHandler.remove(target, MOUSE_MOVE, this.mouseUpClick);
        if (!this.isCalled) {
            this.dotNetRef.invokeMethodAsync(SERVER_ACTION, action, event);
        }
        this.isCalled = false;
        EventHandler.remove(document, MOUSE_UP, this.mouseUpClick);
    };
    SfNumericTextBox.prototype.touchMoveOnSpinner = function (event) {
        var target = document.elementFromPoint(event.clientX, event.clientY);
        if (!(target.classList.contains(ROOT))) {
            clearInterval(this.timeOut);
        }
    };
    SfNumericTextBox.prototype.getElementData = function (event) {
        if ((event.which && event.which === RIGHT_BUTTON) || (event.button && event.button === MOUSE_BUTTON)
            || !this.options.enabled || this.options.readonly) {
            return false;
        }
        clearInterval(this.timeOut);
        return true;
    };
    SfNumericTextBox.prototype.mouseUpClick = function (event) {
        event.stopPropagation();
        clearInterval(this.timeOut);
        this.isCalled = false;
        EventHandler.remove(this.spinUp, MOUSE_MOVE, this.mouseUpClick);
        EventHandler.remove(this.spinDown, MOUSE_MOVE, this.mouseUpClick);
    };
    SfNumericTextBox.prototype.selectRange = function (formatValue) {
        var _this = this;
        if (!Browser.isDevice && Browser.info.version === IE_VERSION) {
            this.element.setSelectionRange(0, formatValue.length);
        }
        else {
            var delay = (Browser.isDevice && Browser.isIos) ? MOBILE_INTERVEL_TIME : INTERVEL_TIME;
            setTimeout(function () {
                _this.element.setSelectionRange(0, formatValue.length);
            }, delay);
        }
    };
    SfNumericTextBox.prototype.isDevice = function () {
        return Browser.isDevice;
    };
    SfNumericTextBox.prototype.spinButtonEvents = function () {
        this.spinDown = this.wrapperElement ? this.wrapperElement.querySelector('.' + SPIN_DOWN) : null;
        this.spinUp = this.wrapperElement ? this.wrapperElement.querySelector('.' + SPIN_UP) : null;
        if (this.spinDown && this.spinUp) {
            this.unBindSpinButton();
            this.bindSpinButton();
        }
    };
    SfNumericTextBox.prototype.bindSpinButton = function () {
        EventHandler.add(this.spinUp, Browser.touchStartEvent, this.mouseDownOnSpinner, this);
        EventHandler.add(this.spinDown, Browser.touchStartEvent, this.mouseDownOnSpinner, this);
        EventHandler.add(this.spinUp, Browser.touchEndEvent, this.mouseUpOnSpinner, this);
        EventHandler.add(this.spinDown, Browser.touchEndEvent, this.mouseUpOnSpinner, this);
        EventHandler.add(this.spinUp, Browser.touchMoveEvent, this.touchMoveOnSpinner, this);
        EventHandler.add(this.spinDown, Browser.touchMoveEvent, this.touchMoveOnSpinner, this);
    };
    SfNumericTextBox.prototype.unBindSpinButton = function () {
        EventHandler.remove(this.spinUp, Browser.touchStartEvent, this.mouseDownOnSpinner);
        EventHandler.remove(this.spinDown, Browser.touchStartEvent, this.mouseDownOnSpinner);
        EventHandler.remove(this.spinUp, Browser.touchEndEvent, this.mouseUpOnSpinner);
        EventHandler.remove(this.spinDown, Browser.touchEndEvent, this.mouseUpOnSpinner);
        EventHandler.remove(this.spinUp, Browser.touchMoveEvent, this.touchMoveOnSpinner);
        EventHandler.remove(this.spinDown, Browser.touchMoveEvent, this.touchMoveOnSpinner);
    };
    return SfNumericTextBox;
}());
var NumericTextBox = {
    initialize: function (wrapperElement, element, dotnetRef, options) {
        if (element) {
            new SfNumericTextBox(wrapperElement, element, dotnetRef, options);
        }
        if (element && element.blazor__instance) {
            element.blazor__instance.initialize();
        }
    },
    selectRange: function (inputEle, formatValue) {
        if (inputEle && inputEle.blazor__instance) {
            inputEle.blazor__instance.selectRange(formatValue);
        }
    },
    propertyChanges: function (element, options) {
        if (element && element.blazor__instance) {
            element.blazor__instance.options = options;
        }
    },
    focusIn: function (inputEle) {
        if (inputEle) {
            inputEle.focus();
        }
    },
    focusOut: function (inputEle) {
        if (inputEle) {
            inputEle.blur();
        }
    },
    spinButtonEvents: function (inputEle) {
        if (inputEle && inputEle.blazor__instance) {
            inputEle.blazor__instance.spinButtonEvents();
        }
    }
};
export default NumericTextBox;
