// @syncfusion/ej2/ej2-resource/scripts/ej2-inputs/blazor/sf-textbox.js
import { EventHandler } from '@syncfusion/ej2-base';
var BLUR = 'blur';
var SfTextBox = (function () {
    function SfTextBox(element, dotnetRef) {
        this.element = element;
        this.element.blazor__instance = this;
        this.dotNetRef = dotnetRef;
        this.isDestroyed = false;
    }
    SfTextBox.prototype.initialize = function () {
        EventHandler.add(this.element, BLUR, this.blurHandler, this);
    };
    SfTextBox.prototype.blurHandler = function () {
        if (!this.isDestroyed) {
            this.dotNetRef.invokeMethodAsync('BlurHandler');
        }
        else {
            EventHandler.remove(this.element, BLUR, this.blurHandler);
        }
    };
    return SfTextBox;
}());
var TextBox = {
    initialize: function (element, dotnetRef) {
        if (element) {
            new SfTextBox(element, dotnetRef);
        }
        if (element && element.blazor__instance) {
            element.blazor__instance.initialize();
        }
    },
    focusIn: function (element) {
        element.focus();
    },
    focusOut: function (element) {
        element.blur();
    },
    destroyInput: function (element) {
        if (element && element.blazor__instance) {
            element.blazor__instance.isDestroyed = true;
        }
    }
};
export default TextBox;
