// @syncfusion/ej2/ej2-resource/scripts/ej2-circulargauge/circular-gauge/annotations/annotations.js
import { stringToNumber, getLocationFromAngle, getFontStyle } from '../utils/helper';
import { getElement, getTemplateFunction, measureElementRect } from '../utils/helper';
import { annotationRender } from '../model/constants';
import { createElement } from '@syncfusion/ej2-base';
var Annotations = (function () {
    function Annotations(gauge) {
        this.gauge = gauge;
        this.elementId = gauge.element.id;
    }
    Annotations.prototype.renderAnnotation = function (axis, index) {
        var _this = this;
        var width = parseInt(this.gauge.width, 10);
        var element = createElement('div', {
            id: this.elementId + '_Annotations_' + index
        });
        var parentElement = getElement(this.elementId + '_Secondary_Element');
        document.getElementById(this.elementId + '_Secondary_Element').style.width = width + 'px';
        axis.annotations.map(function (annotation, annotationIndex) {
            if (annotation.content !== null) {
                _this.createTemplate(element, annotationIndex, index);
            }
        });
        if (parentElement && element.childElementCount) {
            parentElement.appendChild(element);
        }
    };
    Annotations.prototype.createTemplate = function (element, annotationIndex, axisIndex) {
        var axis = this.gauge.axes[axisIndex];
        var annotation = axis.annotations[annotationIndex];
        var childElement = createElement('div', {
            id: this.elementId + '_Axis_' + axisIndex + '_Annotation_' + annotationIndex,
            styles: 'position: absolute; z-index:' + annotation.zIndex + ';transform:' +
                (annotation.autoAngle ? 'rotate(' + (annotation.angle - 90) + 'deg)' : 'rotate(0deg)') + ';'
        });
        var argsData = {
            cancel: false, name: annotationRender, content: annotation.content,
            axis: axis, annotation: annotation, textStyle: annotation.textStyle
        };
        this.gauge.trigger(annotationRender, argsData);
        var templateFn;
        var templateElement;
        if (!argsData.cancel) {
            templateFn = getTemplateFunction(argsData.content);
            if (templateFn && templateFn(axis).length) {
                templateElement = Array.prototype.slice.call(templateFn(axis));
                var length_1 = templateElement.length;
                for (var i = 0; i < length_1; i++) {
                    childElement.appendChild(templateElement[i]);
                }
            }
            else {
                childElement.appendChild(createElement('div', {
                    innerHTML: argsData.content,
                    styles: getFontStyle(argsData.textStyle)
                }));
            }
            this.updateLocation(childElement, axis, annotation);
            element.appendChild(childElement);
        }
    };
    Annotations.prototype.updateLocation = function (element, axis, annotation) {
        var location = getLocationFromAngle(annotation.angle - 90, stringToNumber(annotation.radius, axis.currentRadius), this.gauge.midPoint);
        var elementRect = measureElementRect(element);
        element.style.left = (location.x - (elementRect.width / 2)) + 'px';
        element.style.top = (location.y - (elementRect.height / 2)) + 'px';
        element.setAttribute('aria-label', annotation.description || 'Annotation');
    };
    Annotations.prototype.getModuleName = function () {
        return 'Annotations';
    };
    Annotations.prototype.destroy = function (gauge) {
    };
    return Annotations;
}());
export { Annotations };
