// @syncfusion/ej2/ej2-resource/scripts/ej2-kanban/kanban/base/css-constant.js
export var ROOT_CLASS = 'e-kanban';
export var RTL_CLASS = 'e-rtl';
export var DEVICE_CLASS = 'e-device';
export var ICON_CLASS = 'e-icons';
export var TEMPLATE_CLASS = 'e-template';
export var SWIMLANE_CLASS = 'e-swimlane';
export var TABLE_CLASS = 'e-kanban-table';
export var HEADER_CLASS = 'e-kanban-header';
export var HEADER_TABLE_CLASS = 'e-header-table';
export var HEADER_CELLS_CLASS = 'e-header-cells';
export var HEADER_WRAP_CLASS = 'e-header-wrap';
export var HEADER_TITLE_CLASS = 'e-header-title';
export var HEADER_TEXT_CLASS = 'e-header-text';
export var HEADER_ICON_CLASS = 'e-header-icon';
export var STACKED_HEADER_ROW_CLASS = 'e-stacked-header-row';
export var STACKED_HEADER_CELL_CLASS = 'e-stacked-header-cell';
export var CONTENT_CELLS_CLASS = 'e-content-cells';
export var CONTENT_CLASS = 'e-kanban-content';
export var CONTENT_TABLE_CLASS = 'e-content-table';
export var HEADER_ROW_TOGGLE_CLASS = 'e-toggle-header';
export var HEADER_ROW_CLASS = 'e-header-row';
export var CONTENT_ROW_CLASS = 'e-content-row';
export var SWIMLANE_ROW_CLASS = 'e-swimlane-row';
export var SWIMLANE_ROW_EXPAND_CLASS = 'e-swimlane-row-expand';
export var SWIMLANE_ROW_COLLAPSE_CLASS = 'e-swimlane-row-collapse';
export var SWIMLANE_ROW_TEXT_CLASS = 'e-swimlane-text';
export var CARD_ITEM_COUNT_CLASS = 'e-item-count';
export var CARD_WRAPPER_CLASS = 'e-card-wrapper';
export var CARD_CLASS = 'e-card';
export var DRAGGABLE_CLASS = 'e-draggable';
export var DROPPABLE_CLASS = 'e-droppable';
export var DRAG_CLASS = 'e-drag';
export var DROP_CLASS = 'e-drop';
export var DISABLED_CLASS = 'e-disabled';
export var CARD_HEADER_CLASS = 'e-card-header';
export var CARD_CONTENT_CLASS = 'e-card-content';
export var CARD_HEADER_TEXT_CLASS = 'e-card-header-caption';
export var CARD_HEADER_TITLE_CLASS = 'e-card-header-title';
export var CARD_TAGS_CLASS = 'e-card-tags';
export var CARD_TAG_CLASS = 'e-card-tag';
export var CARD_COLOR_CLASS = 'e-card-color';
export var CARD_LABEL_CLASS = 'e-card-label';
export var CARD_FOOTER_CLASS = 'e-card-footer';
export var EMPTY_CARD_CLASS = 'e-empty-card';
export var CARD_FOOTER_CSS_CLASS = 'e-card-footer-css';
export var COLUMN_EXPAND_CLASS = 'e-column-expand';
export var COLUMN_COLLAPSE_CLASS = 'e-column-collapse';
export var COLLAPSE_HEADER_TEXT_CLASS = 'e-collapse-header-text';
export var COLLAPSED_CLASS = 'e-collapsed';
export var DIALOG_CLASS = 'e-kanban-dialog';
export var DIALOG_CONTENT_CONTAINER_CLASS = 'e-kanban-dialog-content';
export var FORM_CLASS = 'e-kanban-form';
export var FORM_WRAPPER_CLASS = 'e-kanban-form-wrapper';
export var ERROR_VALIDATION_CLASS = 'e-kanban-error';
export var FIELD_CLASS = 'e-field';
export var DRAGGED_CLONE_CLASS = 'e-target-dragged-clone';
export var CLONED_CARD_CLASS = 'e-cloned-card';
export var DRAGGED_CARD_CLASS = 'e-kanban-dragged-card';
export var DROPPED_CLONE_CLASS = 'e-target-dropped-clone';
export var DROPPING_CLASS = 'e-dropping';
export var TOGGLE_VISIBLE_CLASS = 'e-toggle-visible';
export var MULTI_CLONE_CONTENT_CELL_CLASS = 'e-multi-content-cell';
export var MULTI_CARD_WRAPPER_CLASS = 'e-multi-card-wrapper';
export var MULTI_ACTIVE_CLASS = 'e-multi-active';
export var TARGET_MULTI_CLONE_CLASS = 'e-target-multi-clone';
export var MULTI_COLUMN_KEY_CLASS = 'e-column-key';
export var CARD_SELECTION_CLASS = 'e-selection';
export var TOOLTIP_CLASS = 'e-kanban-tooltip';
export var TOOLTIP_TEXT_CLASS = 'e-tooltip-text';
export var SCROLL_CONTAINER_CLASS = 'e-cell-scroll-container';
export var SWIMLANE_HEADER_CLASS = 'e-swimlane-header';
export var SWIMLANE_HEADER_TOOLBAR_CLASS = 'e-swimlane-header-toolbar';
export var TOOLBAR_MENU_CLASS = 'e-toolbar-menu';
export var TOOLBAR_MENU_ICON_CLASS = 'e-icon-menu';
export var TOOLBAR_LEVEL_TITLE_CLASS = 'e-toolbar-level-title';
export var TOOLBAR_SWIMLANE_NAME_CLASS = 'e-toolbar-swimlane-name';
export var SWIMLANE_OVERLAY_CLASS = 'e-swimlane-overlay';
export var SWIMLANE_CONTENT_CLASS = 'e-swimlane-content';
export var SWIMLANE_RESOURCE_CLASS = 'e-swimlane-resource';
export var SWIMLANE_TREE_CLASS = 'e-swimlane-tree';
export var LIMITS_CLASS = 'e-limits';
export var MAX_COUNT_CLASS = 'e-max-count';
export var MIN_COUNT_CLASS = 'e-min-count';
export var MAX_COLOR_CLASS = 'e-max-color';
export var MIN_COLOR_CLASS = 'e-min-color';
export var POPUP_EVENT_CLASS = 'e-event-popup';
export var POPUP_HEADER_CLASS = 'e-popup-header';
export var CLOSE_CLASS = 'e-close';
export var POPUP_CONTENT_CLASS = 'e-popup-content';
export var POPUP_WRAPPER_CLASS = 'e-mobile-popup-wrapper';
export var CLOSE_ICON_CLASS = 'e-close-icon';
export var POPUP_OPEN_CLASS = 'e-popup-open';
export var DIALOG_CONTENT_CONTAINER = 'e-kanban-dialog-content';
export var SHOW_ADD_BUTTON = 'e-show-add-button';
export var SHOW_ADD_ICON = 'e-show-add-icon';
export var SHOW_ADD_FOCUS = 'e-show-add-focus';
