// @syncfusion/ej2/ej2-resource/scripts/ej2-pdf-export/implementation/graphics/images/pdf-image.js
import { SizeF } from './../../drawing/pdf-drawing';
import { PdfGraphicsUnit } from './../enum';
import { PdfUnitConverter } from './../unit-convertor';
var PdfImage = (function () {
    function PdfImage() {
    }
    Object.defineProperty(PdfImage.prototype, "width", {
        get: function () {
            return this.imageWidth;
        },
        set: function (value) {
            this.imageWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfImage.prototype, "height", {
        get: function () {
            return this.imageHeight;
        },
        set: function (value) {
            this.imageHeight = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfImage.prototype, "size", {
        set: function (value) {
            this.width = value.width;
            this.height = value.height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfImage.prototype, "physicalDimension", {
        get: function () {
            this.imagePhysicalDimension = this.getPointSize(this.width, this.height, this.horizontalResolution, this.verticalResolution);
            return new SizeF(this.width, this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfImage.prototype, "element", {
        get: function () {
            return this.imageStream;
        },
        enumerable: true,
        configurable: true
    });
    PdfImage.prototype.getPointSize = function (width, height, horizontalResolution, verticalResolution) {
        if (typeof horizontalResolution === 'undefined') {
            var dpiX = PdfUnitConverter.horizontalResolution;
            var dpiY = PdfUnitConverter.verticalResolution;
            var size = this.getPointSize(width, height, dpiX, dpiY);
            return size;
        }
        else {
            var ucX = new PdfUnitConverter(horizontalResolution);
            var ucY = new PdfUnitConverter(verticalResolution);
            var ptWidth = ucX.convertUnits(width, PdfGraphicsUnit.Pixel, PdfGraphicsUnit.Point);
            var ptHeight = ucY.convertUnits(height, PdfGraphicsUnit.Pixel, PdfGraphicsUnit.Point);
            var size = new SizeF(ptWidth, ptHeight);
            return size;
        }
    };
    return PdfImage;
}());
export { PdfImage };
