// @syncfusion/ej2/ej2-resource/scripts/ej2-spreadsheet/workbook/common/event.js
export var workbookDestroyed = 'workbookDestroyed';
export var updateSheetFromDataSource = 'updateSheetFromDataSource';
export var dataSourceChanged = 'dataSourceChanged';
export var dataChanged = 'dataChanged';
export var workbookOpen = 'workbookOpen';
export var beginSave = 'beginSave';
export var sortImport = 'sortImport';
export var ribbonFind = 'ribbonFind';
export var getFilteredCollection = 'getFilteredCollection';
export var saveCompleted = 'saveCompleted';
export var applyNumberFormatting = 'applyNumber';
export var getFormattedCellObject = 'getFormattedCell';
export var refreshCellElement = 'refreshCellElem';
export var setCellFormat = 'setCellFormat';
export var findAllValues = 'findAllValues';
export var textDecorationUpdate = 'textDecorationUpdate';
export var applyCellFormat = 'applyCellFormat';
export var updateUsedRange = 'updateUsedRange';
export var workbookFormulaOperation = 'workbookFormulaOperation';
export var workbookEditOperation = 'workbookEditOperation';
export var checkDateFormat = 'checkDateFormat';
export var getFormattedBarText = 'getFormattedBarText';
export var activeCellChanged = 'activeCellChanged';
export var openSuccess = 'openSuccess';
export var openFailure = 'openFailure';
export var sheetCreated = 'sheetCreated';
export var sheetsDestroyed = 'sheetsDestroyed';
export var aggregateComputation = 'aggregateComputation';
export var beforeSort = 'beforeSort';
export var initiateSort = 'initiateSort';
export var sortComplete = 'sortComplete';
export var sortRangeAlert = 'sortRangeAlert';
export var initiatelink = 'initiatelink';
export var beforeHyperlinkCreate = 'beforeHyperlinkCreate';
export var afterHyperlinkCreate = 'afterHyperlinkCreate';
export var beforeHyperlinkClick = 'beforeHyperlinkClick';
export var afterHyperlinkClick = 'afterHyperlinkClick';
export var addHyperlink = 'addHyperlink';
export var setLinkModel = 'setLinkModel';
export var beforeFilter = 'beforeFilter';
export var initiateFilter = 'initiateFilter';
export var filterComplete = 'filterComplete';
export var filterRangeAlert = 'filterRangeAlert';
export var clearAllFilter = 'clearAllFilter';
export var wrapEvent = 'wrapText';
export var onSave = 'onSave';
export var insert = 'insert';
export var deleteAction = 'delete';
export var insertModel = 'insertModel';
export var deleteModel = 'deleteModel';
export var isValidation = 'isValidation';
export var setValidation = 'setValidation';
export var addHighlight = 'addHighlight';
export var dataValidate = 'dataValidate';
export var findNext = 'findNext';
export var findPrevious = 'findPrevious';
export var goto = 'gotoHandler';
export var findWorkbookHandler = 'findHandler';
export var replaceHandler = 'replace';
export var replaceAllHandler = 'replaceAll';
export var showDialog = 'ShowDialog';
export var findUndoRedo = 'findUndoRedo';
export var findKeyUp = 'findKeyUp';
export var removeValidation = 'removeValidation';
export var removeHighlight = 'removeHighlight';
export var queryCellInfo = 'queryCellInfo';
export var count = 'count';
export var findCount = 'findCount';
export var protectSheetWorkBook = 'protectSheet';
export var updateToggle = 'updateToggleItem';
export var protectsheetHandler = 'protectsheetHandler';
export var replaceAllDialog = 'replaceAllDialog';
export var unprotectsheetHandler = 'unprotectsheetHandler';
export var workBookeditAlert = 'editAlert';
export var setLockCells = 'setLockCells';
export var applyLockCells = 'applyLockCells';
export var setMerge = 'setMerge';
export var applyMerge = 'applyMerge';
export var mergedRange = 'mergedRange';
export var activeCellMergedRange = 'activeCellMergedRange';
export var insertMerge = 'insertMerge';
export var pasteMerge = 'pasteMerge';
export var setCFRule = 'setCFRule';
export var cFInitialCheck = 'cFInitialCheck';
export var clearCFRule = 'clearCFRule';
export var initiateClearCFRule = 'initiateClearCFRule';
export var cFRender = 'cFRender';
export var cFDelete = 'cFDelete';
export var clear = 'clear';
export var clearCF = 'clearCF';
export var clearCells = 'clearCells';
export var setImage = 'setImage';
export var setChart = 'setChart';
export var initiateChart = 'initiateChart';
export var refreshRibbonIcons = 'refreshRibbonIcons';
export var refreshChart = 'refreshChart';
export var refreshChartSize = 'refreshChartSize';
export var updateChart = 'updateChart';
export var deleteChartColl = 'deleteChartColl';
export var initiateChartModel = 'initiateChartModel';
export var focusChartBorder = 'focusChartBorder';
export var saveError = 'saveError';
export var validationHighlight = 'validationHighlight';
export var dataRefresh = 'dataRefresh';
