// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/blazor/constant.js
export var created = 'create';
export var destroyed = 'destroy';
export var load = 'load';
export var initialLoad = 'initial-load';
export var contentChanged = 'content-changed';
export var initialEnd = 'initial-end';
export var iframeMouseDown = 'iframe-click';
export var destroy = 'destroy';
export var toolbarClick = 'toolbarClick';
export var toolbarRefresh = 'toolbar-refresh';
export var refreshBegin = 'refresh-begin';
export var toolbarUpdated = 'toolbar-updated';
export var bindOnEnd = 'bind-on-end';
export var renderColorPicker = 'render-color-picker';
export var htmlToolbarClick = 'html-toolbar-click';
export var markdownToolbarClick = 'markdown-toolbar-click';
export var destroyColorPicker = 'destroy-color-picker';
export var modelChanged = 'model-changed';
export var keyUp = 'keyUp';
export var keyDown = 'keyDown';
export var mouseUp = 'mouseUp';
export var toolbarCreated = 'toolbarCreated';
export var toolbarRenderComplete = 'toolbarRenderComplete';
export var enableFullScreen = 'enableFullScreen';
export var disableFullScreen = 'disableFullScreen';
export var dropDownSelect = 'dropDownSelect';
export var beforeDropDownItemRender = 'beforeDropDownItemRender';
export var execCommandCallBack = 'execCommandCallBack';
export var imageToolbarAction = 'image-toolbar-action';
export var linkToolbarAction = 'link-toolbar-action';
export var windowResize = 'resize';
export var resizeStart = 'resizeStart';
export var onResize = 'resizing';
export var resizeStop = 'resizeStop';
export var undo = 'undo';
export var redo = 'redo';
export var insertLink = 'insertLink';
export var unLink = 'unLink';
export var editLink = 'editLink';
export var openLink = 'openLink';
export var actionBegin = 'actionBegin';
export var actionComplete = 'actionComplete';
export var toolbarStatusUpdate = 'toolbarStatusUpdate';
export var actionSuccess = 'actionSuccess';
export var updateToolbarItem = 'updateToolbarItem';
export var insertImage = 'insertImage';
export var insertCompleted = 'insertCompleted';
export var imageLeft = 'justifyLeft';
export var imageRight = 'justifyRight';
export var imageCenter = 'justifyCenter';
export var imageBreak = 'break';
export var imageInline = 'inline';
export var imageLink = 'insertImgLink';
export var imageAlt = 'imgAltText';
export var imageDelete = 'delete';
export var imageCaption = 'caption';
export var imageSize = 'imageSize';
export var sourceCode = 'sourceCode';
export var updateSource = 'updateSource';
export var toolbarOpen = 'toolbarOpen';
export var beforeDropDownOpen = 'beforeDropDownOpen';
export var selectionSave = 'selection-save';
export var selectionRestore = 'selection-restore';
export var expandPopupClick = 'expandPopupClick';
export var count = 'count';
export var contentFocus = 'contentFocus';
export var contentBlur = 'contentBlur';
export var mouseDown = 'mouseDown';
export var sourceCodeMouseDown = 'sourceCodeMouseDown';
export var editAreaClick = 'editAreaClick';
export var scroll = 'scroll';
export var contentscroll = 'contentscroll';
export var colorPickerChanged = 'colorPickerChanged';
export var tableColorPickerChanged = 'tableColorPickerChanged';
export var focusChange = 'focusChange';
export var selectAll = 'selectAll';
export var selectRange = 'selectRange';
export var getSelectedHtml = 'getSelectedHtml';
export var renderInlineToolbar = 'renderInlineToolbar';
export var paste = 'paste-content';
export var imgModule = 'imageModule';
export var rtlMode = 'rtl-mode';
export var createTable = 'createTable';
export var docClick = 'docClick';
export var tableToolbarAction = 'table-toolbar-action';
export var checkUndo = 'checkUndoStack';
export var readOnlyMode = 'readOnlyMode';
export var pasteClean = 'pasteClean';
export var beforeDialogOpen = 'beforeDialogOpen';
export var dialogOpen = 'dialogOpen';
export var beforeDialogClose = 'beforeDialogClose';
export var dialogClose = 'dialogClose';
export var beforeQuickToolbarOpen = 'beforeQuickToolbarOpen';
export var quickToolbarOpen = 'quickToolbarOpen';
export var quickToolbarClose = 'quickToolbarClose';
export var popupHide = 'popupHide';
export var imageSelected = 'imageSelected';
export var imageUploading = 'imageUploading';
export var imageUploadSuccess = 'imageUploadSuccess';
export var imageUploadFailed = 'imageUploadFailed';
export var imageRemoving = 'imageRemoving';
export var afterImageDelete = 'afterImageDelete';
export var drop = 'drop';
export var xhtmlValidation = 'xhtmlValidation';
export var beforeImageUpload = 'beforeImageUpload';
export var resizeInitialized = 'resizeInitialized';
export var MS_WORD_CLEANUP_PLUGIN = 'ms_word_cleanup_plugin';
export var MS_WORD_CLEANUP = 'ms_word_cleanup';
export var beforePasteUpload = 'beforePasteUpload';
export var beforePasteUploadCallBack = 'beforePasteUploadCallBack';
export var uploading = 'uploading';
export var success = 'success';
export var failure = 'failure';
export var canceling = 'canceling';
export var removing = 'removing';
export var selected = 'selected';
export var beforeQuickToolbarOpenEvent = 'BeforeQuickToolbarOpenEvent';
export var quickToolbarCloseEvent = 'QuickToolbarCloseEvent';
export var actionBeginEvent = 'ActionBeginEvent';
export var actionCompleteEvent = 'ActionCompleteEvent';
export var beforeUpload = 'BeforeUpload';
export var pasteImageUploadFailed = 'PasteImageUploadFailed';
export var pasteImageUploadSuccess = 'PasteImageUploadSuccess';
export var resizeStartEvent = 'ResizeStartEvent';
export var resizeStopEvent = 'ResizeStopEvent';
export var updateClass = 'UpdateClass';
export var showFullScreenClient = 'ShowFullScreenClient';
export var hideFullScreenClient = 'HideFullScreenClient';
export var showImagePopup = 'ShowImagePopup';
export var hideImagePopup = 'HideImagePopup';
export var showLinkPopup = 'ShowLinkPopup';
export var hideLinkPopup = 'HideLinkPopup';
export var showTablePopup = 'ShowTablePopup';
export var hideTablePopup = 'HideTablePopup';
export var showInlinePopup = 'ShowInlinePopup';
export var hideInlinePopup = 'HideInlinePopup';
export var refreshToolbarOverflow = 'RefreshToolbarOverflow';
export var updateUndoRedoStatus = 'UpdateUndoRedoStatus';
export var showImageDialog = 'ShowImageDialog';
export var closeImageDialog = 'CloseImageDialog';
export var imageQuickPopup = '_Image_Quick_Popup';
export var linkQuickPopup = '_Link_Quick_Popup';
export var tableQuickPopup = '_Table_Quick_Popup';
export var inlineQuickPopup = '_Inline_Quick_Popup';
export var resizeID = '-resizable';
export var toolbarCreateTable = '_toolbar_CreateTable';
export var imgResizeId = '_imgResize';
export var IFRAME_HEADER = "\n<!DOCTYPE html> \n    <html>\n         <head>\n            <meta charset='utf-8' /> \n            <style>\n                @charset \"UTF-8\";\n                body {\n                    font-family: \"Roboto\", sans-serif;\n                    font-size: 14px;\n                }\n                html, body{height: 100%;margin: 0;}\n                body.e-cursor{cursor:default}\n                span.e-selected-node\t{background-color: #939393;color: white;}\n                span.e-selected-node.e-highlight {background-color: #1d9dd8;}\n                body{color:#333;word-wrap:break-word;padding: 8px;box-sizing: border-box;}\n                .e-rte-image {border: 0;cursor: pointer;display: block;float: none;height: auto;margin: 5px auto;max-width: 100%;position: relative;}\n                .e-img-caption { display: inline-block; float: none; margin: 5px auto; max-width: 100%;position: relative;}\n                .e-img-caption.e-caption-inline {display: inline-block;float: none;margin: 5px auto;margin-left: 5px;margin-right: 5px;max-width: calc(100% - (2 * 5px));position: relativetext-align: center;vertical-align: bottom;}\n                .e-img-inner {box-sizing: border-box;display: block;font-size: 16px;font-weight: initial;margin: auto;opacity: .9;text-align: center;width: 100%;}\n                .e-img-wrap {display: inline-block;margin: auto;padding: 0;text-align: center;width: 100%;}\n                .e-imgleft {float: left;margin: 0 5px 0 0;text-align: left;}\n                .e-imgright {float: right;margin: 0 0 0 5px;text-align: right;}\n                .e-imgcenter {cursor: pointer;display: block;float: none;height: auto;margin: 5px auto;max-width: 100%;position: relative;}\n                .e-control img:not(.e-resize) {border: 2px solid transparent; z-index: 1000}\n                .e-imginline {display: inline-block;float: none;margin-left: 5px;margin-right: 5px;max-width: calc(100% - (2 * 5px));vertical-align: bottom;}\n                .e-imgbreak {border: 0;cursor: pointer;display: block;float: none;height: auto;margin: 5px auto;max-width: 100%;position: relative;}\n                .e-rte-image.e-img-focus:not(.e-resize) {border: solid 2px #4a90e2;}\n                img::selection { background: transparent;color: transparent;}\n                span.e-rte-imageboxmark {  width: 10px; height: 10px; position: absolute; display: block; background: #4a90e2; border: 1px solid #fff; z-index: 1000;}\n                .e-mob-rte.e-mob-span span.e-rte-imageboxmark { background: #4a90e2; border: 1px solid #fff; }\n                .e-mob-rte span.e-rte-imageboxmark { background: #fff; border: 1px solid #4a90e2; border-radius: 15px; height: 20px; width: 20px; }\n                .e-mob-rte.e-mob-span span.e-rte-imageboxmark { background: #4a90e2; border: 1px solid #fff; }\n                .e-rte-content .e-content img.e-resize { z-index: 1000; }\n                .e-img-caption .e-img-inner { outline: 0; }\n                .e-img-caption .e-rte-image.e-imgright, .e-img-caption .e-rte-image.e-imgleft { float: none; margin: 0;}\n                body{box-sizing: border-box;min-height: 100px;outline: 0 solid transparent;overflow-x: auto;padding: 16px;position: relative;text-align: inherit;z-index: 2;}\n                p{margin: 0 0 10px;margin-bottom: 10px;}\n                li{margin-bottom: 10px;}\n                h1{font-size: 2.17em;font-weight: 400;line-height: 1;margin: 10px 0;}\n                h2{font-size: 1.74em;font-weight: 400;margin: 10px 0;}\n                h3{font-size: 1.31em;font-weight: 400;margin: 10px 0;}\n                h4{font-size: 16px;font-weight: 400;line-height: 1.5;margin: 0;}\n                h5{font-size: 00.8em;font-weight: 400;margin: 0;}\n                h6{font-size: 00.65em;font-weight: 400;margin: 0;}\n                blockquote{margin: 10px 0;margin-left: 0;padding-left: 5px;border-left: solid 2px #5c5c5c;}\n                pre{background-color: inherit;border: 0;border-radius: 0;color: #333;font-size: inherit;line-height: inherit;margin: 0 0 10px;overflow: visible;padding: 0;white-space: pre-wrap;word-break: inherit;word-wrap: break-word;}\n                strong, b{font-weight: 700;}\n                a{text-decoration: none;user-select: auto;}\n                a:hover{text-decoration: underline;};\n                p:last-child, pre:last-child, blockquote:last-child{margin-bottom: 0;}\n                h3+h4, h4+h5, h5+h6{margin-top: 00.6em;}\n                ul:last-child{margin-bottom: 0;}\n                table { border-collapse: collapse; empty-cells: show;}\n                table td,table th {border: 1px solid #BDBDBD; height: 20px; vertical-align: middle;}\n                table.e-alternate-border tbody tr:nth-child(2n) {background-color: #F5F5F5;}\n                table th {background-color: #E0E0E0;}\n                table.e-dashed-border td,table.e-dashed-border th { border: 1px dashed #BDBDBD} \n                table .e-cell-select {border: 1px double #4a90e2;}\n                span.e-table-box { cursor: nwse-resize; display: block; height: 10px; position: absolute; width: 10px; }\n                span.e-table-box.e-rmob {height: 14px;width: 14px;}\n                .e-row-resize, .e-column-resize { background-color: transparent; background-repeat: repeat; bottom: 0;cursor: col-resize;height: 1px;overflow: visible;position: absolute;width: 1px; }\n                .e-row-resize { cursor: row-resize; height: 1px;}\n                .e-table-rhelper { cursor: col-resize; opacity: .87;position: absolute;}\n                .e-table-rhelper.e-column-helper { width: 1px; }\n                .e-table-rhelper.e-row-helper {height: 1px;}\n                .e-reicon::before { border-bottom: 6px solid transparent; border-right: 6px solid; border-top: 6px solid transparent; content: ''; display: block; height: 0; position: absolute; right: 4px; top: 4px; width: 20px; }\n                .e-reicon::after { border-bottom: 6px solid transparent; border-left: 6px solid; border-top: 6px solid transparent; content: ''; display: block; height: 0; left: 4px; position: absolute; top: 4px; width: 20px; z-index: 3; }\n                .e-row-helper.e-reicon::after { top: 10px; transform: rotate(90deg); }\n                .e-row-helper.e-reicon::before { left: 4px; top: -20px; transform: rotate(90deg); }\n                span.e-table-box { background-color: #ffffff; border: 1px solid #BDBDBD; }\n                span.e-table-box.e-rbox-select { background-color: #BDBDBD; border: 1px solid #BDBDBD; }\n                .e-table-rhelper { background-color: #4a90e2;}\n                .e-rtl { direction: rtl; }\n            </style>\n        </head>";
