// @syncfusion/ej2/ej2-resource/scripts/ej2-pivotview/blazor/common/calculated-field.js
import * as cls from '../common/constants';
import { getZindexPartial } from '@syncfusion/ej2-popups';
import { closest, removeClass, Draggable, addClass, remove, createElement, prepend, append, Droppable } from '@syncfusion/ej2-base';
var CalculatedField = (function () {
    function CalculatedField(parent) {
        this.parent = parent;
        this.parent.calculatedFieldModule = this;
    }
    CalculatedField.prototype.getNodeLocation = function (treeElement, id) {
        var position = treeElement.querySelector('li[data-uid=' + '\"' + JSON.parse(id) + '\"' + ']').getBoundingClientRect();
        return JSON.stringify([position.top + (window.scrollY || document.documentElement.scrollTop), position.left]);
    };
    CalculatedField.prototype.setSelectionRange = function (length) {
        if (length) {
            document.querySelector('#' + this.parent.element.id + 'droppable').setSelectionRange(length, length);
        }
        document.querySelector('#' + this.parent.element.id + 'droppable').focus();
    };
    CalculatedField.prototype.getIconInfo = function (top, left) {
        var element = document.elementFromPoint(left, top);
        if (element) {
            if (element.classList.contains(cls.FORMAT)) {
                return JSON.stringify(cls.FORMAT);
            }
            else if (element.classList.contains(cls.CALC_EDIT)) {
                return JSON.stringify(cls.CALC_EDIT);
            }
            else if (element.classList.contains(cls.GRID_REMOVE)) {
                return JSON.stringify(cls.GRID_REMOVE);
            }
            else if (element.classList.contains(cls.CALC_EDITED)) {
                return JSON.stringify(cls.CALC_EDITED);
            }
        }
        return undefined;
    };
    CalculatedField.prototype.emptyFieldName = function (parentID) {
        addClass([document.getElementById(parentID + 'ddlelement')], [cls.EMPTY_FIELD, cls.CALCINPUT]);
        document.getElementById(parentID + 'ddlelement').focus();
    };
    CalculatedField.prototype.editCalculatedFieldInfo = function (isEdit, top, left, title) {
        var treeNode = closest(document.elementFromPoint(left, top), 'li');
        if (isEdit) {
            if (this.parent.options.dataType === 'pivot') {
                addClass(document.querySelector('#' + this.parent.element.id + 'calculateddialog').querySelectorAll('.' + cls.CALC_EDITED), cls.CALC_EDIT);
                removeClass(document.querySelector('#' + this.parent.element.id + 'calculateddialog').querySelectorAll('.' + cls.CALC_EDITED), cls.CALC_EDITED);
                addClass([treeNode.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDITED);
                removeClass([treeNode.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDIT);
                treeNode.querySelector('.' + cls.CALC_EDITED).setAttribute('title', title);
            }
            document.querySelector('#' + this.parent.element.id + 'droppable').value =
                treeNode.getAttribute('data-uid');
        }
        else {
            this.updatedCalculatedFieldInfo(treeNode, title);
        }
    };
    CalculatedField.prototype.updatedCalculatedFieldInfo = function (treeNode, title) {
        if (this.parent.options.dataType === 'pivot') {
            addClass(document.querySelector('#' + this.parent.element.id + 'calculateddialog').querySelectorAll('.' + cls.CALC_EDITED), cls.CALC_EDIT);
            removeClass(document.querySelector('#' + this.parent.element.id + 'calculateddialog').querySelectorAll('.' + cls.CALC_EDITED), cls.CALC_EDITED);
            treeNode.querySelector('.' + cls.CALC_EDIT).setAttribute('title', title);
        }
        document.querySelector('#' + this.parent.element.id + 'droppable').value = '';
    };
    CalculatedField.prototype.updateNodeExpandIcons = function (treeElement, id) {
        var node = treeElement.querySelector('li[data-uid=' + '\"' + id + '\"' + ']');
        var li = [].slice.call(node.querySelectorAll('li'));
        for (var i = 0; i < li.length; i++) {
            var liTextElement = li[i].querySelector('.' + cls.TEXT_CONTENT_CLASS);
            if (li[i].getAttribute('data-type') === cls.CALCULATED_FIELD &&
                liTextElement && li[i].querySelector('.' + cls.LIST_ICON + '.' + cls.CALC_MEMBER) &&
                !li[i].querySelector('.' + cls.GRID_REMOVE)) {
                var removeElement = createElement('span', {
                    className: cls.GRID_REMOVE + ' ' + cls.ICON + ' ' + cls.LIST_ICON
                });
                liTextElement.classList.add(cls.CALC_FIELD_MEMBER);
                liTextElement.appendChild(removeElement);
            }
            if ((li[i].querySelector('.' + cls.CALC_DIMENSION_ICON + ',.' + cls.CALC_MEASURE_ICON + ',.' + cls.MEASURE_ICON) ||
                li[i].querySelector('.' + cls.DIMENSION_TYPE_ICON + ',.' + cls.ATTRIBUTE_TYPE_ICON + ',.' + cls.HIERARCHY_TYPE_ICON) ||
                li[i].querySelector('.' + cls.LEVEL_MEMBERS + ',.' + cls.NAMEDSET_TYPE_ICON))) {
                this.createDraggable(li[i]);
            }
        }
        return this.olapExpand(node);
    };
    CalculatedField.prototype.olapExpand = function (node) {
        if (node && node.querySelector('.' + cls.LIST_ICON) &&
            node.querySelector('.' + cls.ICON_EXPAND) &&
            (node.querySelector('.' + cls.LIST_ICON).className.indexOf(cls.FOLDER_TYPE_ICON) > -1)) {
            var listnode = node.querySelector('.' + cls.LIST_ICON);
            removeClass([listnode], cls.FOLDER_TYPE_ICON);
            addClass([listnode], cls.FOLDER_TYPE_OPEN_ICON);
            return JSON.stringify(false);
        }
        else if (node && node.querySelector('.' + cls.LIST_ICON) &&
            node.querySelector('.' + cls.ICON_COLLAPSE) &&
            (node.querySelector('.' + cls.LIST_ICON).className.indexOf(cls.FOLDER_TYPE_OPEN_ICON) > -1)) {
            node = node.querySelector('.' + cls.LIST_ICON);
            removeClass([node], cls.FOLDER_TYPE_OPEN_ICON);
            addClass([node], cls.FOLDER_TYPE_ICON);
            return JSON.stringify(false);
        }
        return JSON.stringify(true);
    };
    CalculatedField.prototype.updateEditOptions = function (accordId) {
        var acoordItem = [].slice.call(document.querySelectorAll('#' + accordId + ' .' + cls.ACCORD_ITEM));
        addClass(document.querySelectorAll('#' + accordId), cls.CALCACCORD);
        for (var i = 0; i < acoordItem.length; i++) {
            if (acoordItem[i].querySelector('[data-type="CalculatedField"]')) {
                var iconElement = acoordItem[i].querySelector('.' + cls.ACCORD_HEADER + ' .' + cls.TOGGLE_ICON);
                removeClass([iconElement], cls.TOGGLE_ICON);
                addClass([iconElement], cls.ACCORD_HEADER_ICON);
                var editElement = iconElement.querySelector('.' + cls.TOGGLE_COLLAPSE_ICON);
                removeClass([editElement], cls.TOGGLE_COLLAPSE_ICON);
                addClass([editElement], [cls.LIST_ICON, cls.CALC_EDIT]);
                iconElement.appendChild(createElement('span', {
                    className: cls.GRID_REMOVE + ' ' + cls.ICON + ' ' + cls.LIST_ICON
                }));
            }
        }
    };
    CalculatedField.prototype.updateAccordionLabel = function (target) {
        var eventTarget = document.elementFromPoint(JSON.parse(target).ClientX, JSON.parse(target).ClientY);
        var type = eventTarget.parentElement.querySelector('.' + cls.LABEl)
            .innerText;
        var field = closest(eventTarget, '.' + cls.ACCORD_ITEM).
            querySelector('[data-field').getAttribute('data-caption');
        closest(eventTarget, '.' + cls.ACCORD_ITEM).querySelector('.' + cls.LABEl).
            innerText = field + ' (' + type + ')';
        closest(eventTarget, '.' + cls.ACCORD_ITEM).
            querySelector('[data-type').setAttribute('data-type', (eventTarget).getAttribute('data-value'));
    };
    CalculatedField.prototype.accordionClick = function (clientX, clientY, id) {
        var target = document.elementFromPoint(JSON.parse(clientX), JSON.parse(clientY));
        if (closest(target, '.' + cls.ACCORD_HEADER_ICON)) {
            var node = closest(target, '.' + cls.ACCORD_HEADER).querySelector('.' + cls.CALCCHECK + ' input');
            if (node) {
                var actionClass = void 0;
                var optionElement = closest(target, '.' + cls.ACCORD_HEADER_ICON);
                if (optionElement.querySelector('.' + cls.CALC_EDIT) && (target).classList.contains(cls.CALC_EDIT)) {
                    addClass([optionElement.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDITED);
                    removeClass([optionElement.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDIT);
                    actionClass = cls.CALC_EDIT;
                }
                else if (optionElement.querySelector('.' + cls.CALC_EDITED) &&
                    target.classList.contains(cls.CALC_EDITED)) {
                    addClass([optionElement.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDIT);
                    removeClass([optionElement.querySelector('.' + cls.LIST_ICON)], cls.CALC_EDITED);
                    actionClass = cls.CALC_EDITED;
                }
                else if (optionElement.querySelector('.' + cls.GRID_REMOVE) &&
                    target.classList.contains(cls.GRID_REMOVE)) {
                    actionClass = cls.GRID_REMOVE;
                }
                return JSON.stringify([node.getAttribute('data-field'), actionClass, node.getAttribute('id').split(id + '_')[1]]);
            }
        }
        return JSON.stringify([]);
    };
    CalculatedField.prototype.getAccordionValue = function () {
        var fieldText = '';
        var node = [].slice.call(document.querySelectorAll('.' + cls.ACCORDION + ' .' + cls.NODE_CHECK_CLASS));
        for (var i = 0; i < node.length; i++) {
            var field = node[i].parentElement.querySelector('[data-field]').getAttribute('data-field');
            var type = node[i].parentElement.querySelector('[data-type]').getAttribute('data-type');
            if (type.indexOf(cls.CALCULATED_FIELD) === -1) {
                fieldText = fieldText + ('"' + type + '(' + field + ')' + '"');
            }
            else {
                fieldText = fieldText + node[i].parentElement.querySelector('[data-formula]').getAttribute('data-formula');
            }
        }
        return fieldText;
    };
    CalculatedField.prototype.createFormulaDroppable = function (edit, drag, remove, edited, isEdit, fieldName, id) {
        var element = document.getElementById(this.parent.element.id + id);
        if (element.querySelector('.' + cls.FORMULA)) {
            new Droppable(element.querySelector('.' + cls.FORMULA), {});
            var li = [].slice.call(element.querySelectorAll('.' + cls.TREEVIEW + ' ul li'));
            for (var i = 0; i < li.length; i++) {
                var currentFieldName = li[i].getAttribute('data-field');
                if (this.parent.options.dataType === 'olap') {
                    if (li[i].querySelector('.' + cls.MEASURE_ICON)) {
                        li[i].querySelector('.' + cls.LIST_ICON).style.display = 'none';
                    }
                    var liTextElement = li[i].querySelector('.' + cls.TEXT_CONTENT_CLASS);
                    if (li[i].getAttribute('data-type') === 'CalculatedField' &&
                        liTextElement && li[i].querySelector('.' + cls.LIST_ICON + '.' + cls.CALC_MEMBER)) {
                        liTextElement.classList.add(cls.CALC_FIELD_MEMBER);
                        liTextElement.appendChild(createElement('span', {
                            className: cls.GRID_REMOVE + ' ' + cls.ICON + ' ' + cls.LIST_ICON
                        }));
                    }
                    if ((li[i].querySelector('.' + cls.CALC_DIMENSION_ICON + ',.' + cls.CALC_MEASURE_ICON + ',.' + cls.MEASURE_ICON) ||
                        li[i].querySelector('.' + cls.DIMENSION_TYPE_ICON + ',.' + cls.ATTRIBUTE_TYPE_ICON + ',.' + cls.HIERARCHY_TYPE_ICON) ||
                        li[i].querySelector('.' + cls.LEVEL_MEMBERS + ',.' + cls.NAMEDSET_TYPE_ICON))) {
                        this.createDraggable(li[i]);
                    }
                }
                else if (this.parent.options.dataType === 'pivot') {
                    if (li[i].querySelector('.' + cls.TEXT_CONTENT_CLASS + ' span')) {
                        var type = li[i].getAttribute('data-type');
                        var dragElement = createElement('span', {
                            attrs: { 'tabindex': '-1', 'aria-disabled': 'false', 'title': drag },
                            className: cls.ICON + ' ' + cls.DRAG_CLASS
                        });
                        var spaceElement = createElement('div', {
                            className: ' ' + cls.ICON_SPACE
                        });
                        prepend([dragElement], li[i].querySelector('.' + cls.TEXT_CONTENT_CLASS));
                        append([spaceElement, li[i].querySelector('.' + cls.FORMAT)], li[i].querySelector('.' + cls.TEXT_CONTENT_CLASS));
                        if (type === 'CalculatedField') {
                            li[i].querySelector('.' + cls.FORMAT).setAttribute('title', remove);
                            addClass([li[i].querySelector('.' + cls.FORMAT)], cls.GRID_REMOVE);
                            if (isEdit && fieldName === currentFieldName) {
                                element.querySelector('#' + this.parent.element.id + 'droppable').value = li[i].getAttribute('data-uid');
                            }
                            addClass([li[i].querySelector('.' + cls.ICON_SPACE)], [((isEdit && fieldName === currentFieldName) ? cls.CALC_EDITED : cls.CALC_EDIT), cls.ICON, cls.LIST_ICON]);
                            li[i].querySelector('.' + cls.CALC_EDIT + ',.' + cls.CALC_EDITED).setAttribute('title', ((isEdit && fieldName === currentFieldName) ? edited : edit));
                            li[i].querySelector('.' + cls.CALC_EDIT + ',.' + cls.CALC_EDITED).setAttribute('aria-disabled', 'false');
                            li[i].querySelector('.' + cls.CALC_EDIT + ',.' + cls.CALC_EDITED).setAttribute('tabindex', '-1');
                            removeClass([li[i].querySelector('.' + cls.FORMAT)], cls.FORMAT);
                            removeClass([li[i].querySelector('.' + cls.ICON_SPACE)], cls.ICON_SPACE);
                        }
                        this.createDraggable(dragElement);
                    }
                }
            }
        }
    };
    CalculatedField.prototype.createDraggable = function (element) {
        new Draggable(element, {
            dragTarget: '.' + cls.LIST_ITEM,
            clone: true,
            enableTailMode: true,
            enableAutoScroll: true,
            helper: this.calculatedDragClone.bind(this),
            dragStart: this.onCalcDragStart.bind(this),
            drag: this.onDragging.bind(this),
            dragStop: this.onCalcDragStop.bind(this)
        });
    };
    CalculatedField.prototype.onDragging = function (e) {
        var clonedNode = document.querySelector('#' + this.parent.element.id + '_DragClone ' + '.' + cls.ICON);
        if (e.target && (e.target).classList.contains(cls.FORMULA)) {
            removeClass([clonedNode], cls.NO_DRAG_CLASS);
            addClass([(e.target)], cls.COPY_DROP);
        }
        else {
            addClass([clonedNode], cls.NO_DRAG_CLASS);
            removeClass([(e.target)], cls.COPY_DROP);
            addClass([clonedNode], cls.ICON_EXPAND);
            removeClass([clonedNode], cls.LIST_ICON);
        }
    };
    CalculatedField.prototype.onCalcDragStop = function (e) {
        var dropField = document.getElementById(this.parent.element.id + 'droppable');
        var draggedNode = closest(e.element, '.' + cls.LIST_ITEM);
        removeClass([dropField], cls.COPY_DROP);
        removeClass([draggedNode.querySelector('.' + cls.LIST_TEXT_CLASS)], cls.SELECTED_NODE_CLASS);
        var cursorPos = dropField.selectionStart;
        if (e.target.id === (this.parent.element.id + 'droppable')) {
            this.parent.dotNetRef.invokeMethodAsync('UpdateCalcDroppable', draggedNode.getAttribute('data-uid'), cursorPos, e.target.id);
        }
        if (document.getElementById(this.parent.element.id + '_DragClone')) {
            remove(e.helper);
        }
        document.body.style.cursor = 'auto';
        this.parent.isDragging = false;
    };
    CalculatedField.prototype.onCalcDragStart = function (e) {
        var draggedNode = closest(e.element, '.' + cls.LIST_ITEM);
        var clonedNode = document.getElementById(this.parent.element.id + '_DragClone');
        if (clonedNode && draggedNode && ((this.parent.options.dataType === 'pivot' && e.target &&
            e.target.classList.contains(cls.DRAG_CLASS)) || (this.parent.options.dataType === 'olap' &&
            (draggedNode.querySelector('.' + cls.CALC_DIMENSION_ICON + ',.' + cls.CALC_MEASURE_ICON + ',.' + cls.MEASURE_ICON) ||
                draggedNode.querySelector('.' + cls.DIMENSION_TYPE_ICON + ',.' + cls.ATTRIBUTE_TYPE_ICON + ',.' + cls.HIERARCHY_TYPE_ICON) ||
                draggedNode.querySelector('.' + cls.LEVEL_MEMBERS + ',.' + cls.NAMEDSET_TYPE_ICON))))) {
            addClass([draggedNode.querySelector('.' + cls.LIST_TEXT_CLASS)], cls.SELECTED_NODE_CLASS);
            var zIndex = getZindexPartial(draggedNode);
            clonedNode.style.zIndex = zIndex ? zIndex.toString() :
                document.querySelector('.' + cls.CALCDIALOG + '.' + cls.OLAP_CALCDIALOG).style.zIndex + 1;
            clonedNode.style.display = 'inline';
            this.parent.isDragging = true;
        }
        else {
            this.parent.isDragging = false;
        }
        e.bindEvents(e.dragElement);
    };
    CalculatedField.prototype.calculatedDragClone = function (args) {
        var cloneElement = createElement('div', {
            id: this.parent.element.id + '_DragClone',
            className: cls.DRAG_ITEM + ' ' + cls.TREEVIEW_CLASS + ' ' + cls.PIVOTCALC
        });
        var contentElement = createElement('div', {
            className: cls.TEXT_CONTENT_CLASS + ' ' + cls.CALC_FIELD_MEMBER
        });
        contentElement.appendChild(createElement('div', {
            className: cls.ICON
        }));
        contentElement.appendChild(createElement('span', {
            className: cls.LIST_TEXT_CLASS,
            innerHTML: closest(args.element, 'li').getAttribute('data-caption')
        }));
        cloneElement.appendChild(contentElement);
        document.body.appendChild(cloneElement);
        return cloneElement;
    };
    return CalculatedField;
}());
export { CalculatedField };
