// @syncfusion/ej2/ej2-resource/scripts/ej2-pivotview/blazor/common/constants.js
export var ICON = 'e-icons';
export var ICON_HIDDEN = 'e-hide';
export var AXISFIELD_ICON_CLASS = 'e-dropdown-icon';
export var TOGGLE_FIELD_LIST_CLASS = 'e-toggle-field-list';
export var FIELD_LIST_FOOTER_CLASS = 'e-field-list-footer';
export var LEFT_AXIS_PANEL_CLASS = 'e-left-axis-fields';
export var RIGHT_AXIS_PANEL_CLASS = 'e-right-axis-fields';
export var AXIS_CONTENT_CLASS = 'e-axis-content';
export var PIVOT_BUTTON_DIV_CLASS = 'e-pvt-btn-div';
export var PIVOT_BUTTON_CLASS = 'e-pivot-button';
export var PIVOT_BUTTON_CONTENT_CLASS = 'e-content';
export var DRAG_CLONE_CLASS = 'e-button-drag-clone';
export var DRAG_CLASS = 'e-drag';
export var DRAG_DISABLE_CLASS = 'e-disable-drag';
export var DROP_INDICATOR_CLASS = 'e-drop-indicator';
export var INDICATOR_HOVER_CLASS = 'e-drop-hover';
export var MEMBER_EDITOR_DIALOG_CLASS = 'e-member-editor-dialog';
export var EDITOR_TREE_CONTAINER_CLASS = 'e-member-editor-container';
export var CHECK_BOX_FRAME_CLASS = 'e-frame';
export var NODE_CHECK_CLASS = 'e-check';
export var CALCULATED_FIELD = 'CalculatedField';
export var NODE_STOP_CLASS = 'e-stop';
export var DROPPABLE_CLASS = 'e-droppable';
export var ROW_AXIS_CLASS = 'e-rows';
export var COLUMN_AXIS_CLASS = 'e-columns';
export var FILTER_AXIS_CLASS = 'e-filters';
export var GROUP_ROW = 'e-group-row';
export var GROUP_ROW_CLASS = 'e-group-rows';
export var GROUP_COLUMN_CLASS = 'e-group-columns';
export var GROUP_VALUE_CLASS = 'e-group-values';
export var GROUP_FILTER_CLASS = 'e-group-filters';
export var NO_DRAG_CLASS = 'e-drag-restrict';
export var SELECTED_NODE_CLASS = 'e-list-selected';
export var LIST_ITEM = 'e-list-item';
export var TEXT_CONTENT_CLASS = 'e-text-content';
export var FOOTER_CONTENT_CLASS = 'e-footer-content';
export var ADAPTIVE_CONTAINER_CLASS = 'e-adaptive-container';
export var LIST_TEXT_CLASS = 'e-list-text';
export var LIST_SELECT_CLASS = 'e-selected-node';
export var VALUESCONTENT = 'e-valuescontent';
export var HEADERCELL = 'e-headercell';
export var CELLVALUE = 'e-cellvalue';
export var ROW = 'e-row';
export var SELECT_ALL_CLASS = 'e-select-all';
export var PIVOTCALC = 'e-pivot-calc';
export var CALCDIALOG = 'e-pivot-calc-dialog-div';
export var OLAP_CALCDIALOG = 'e-olap-calc-dialog-div';
export var CALCINPUT = 'e-pivot-calc-input';
export var FORMAT = 'e-format';
export var FORMULA = 'e-pivot-formula';
export var TREEVIEW = 'e-pivot-treeview';
export var CALCACCORD = 'e-pivot-accord';
export var AXIS_ROW_CLASS = 'e-axis-row';
export var CALC_EDIT = 'e-edit';
export var CALC_EDITED = 'e-edited';
export var EMPTY_FIELD = 'e-empty-field';
export var CALCCHECK = 'e-pivot-calc-check';
export var MOVABLECONTENT_DIV = 'e-movablecontent';
export var FROZENCONTENT_DIV = 'e-frozencontent';
export var MOVABLEHEADER_DIV = 'e-movableheader';
export var FROZENHEADER_DIV = 'e-frozenheader';
export var HEADERCONTENT = 'e-headercontent';
export var ROW_CELL_CLASS = 'e-rowcell';
export var GRID_REMOVE = 'e-remove-report';
export var GRID_TOOLBAR = 'e-pivot-toolbar';
export var GRID_REPORT_INPUT = 'e-pivotview-report-input';
export var GROUP_CHART_ROW = 'e-group-chart-rows';
export var GROUP_CHART_COLUMN = 'e-group-chart-columns';
export var GROUP_CHART_VALUE = 'e-group-chart-values';
export var GROUP_CHART_VALUE_DROPDOWN = 'e-group-chart-values-drodown';
export var GROUP_CHART_COLUMN_DROPDOWN = 'e-group-chart-columns-dropdown';
export var CHART_GROUPING_BAR_CLASS = 'e-chart-grouping-bar';
export var GRID_GROUPING_BAR_CLASS = 'e-pivot-grouping-bar';
export var MDX_QUERY_CONTENT = 'e-mdx-query-content';
export var HEADER_TEXT = 'e-headertext';
export var STACKED_HEADER_TEXT = 'e-stackedheadertext';
export var HEADER_CELL_DIV = 'e-headercelldiv';
export var STACKED_HEADER_CELL_DIV = 'e-stackedheadercelldiv';
export var STACKED_HEADER_CELL = 'e-stackedheadercell';
export var AXIS_CONTAINER = 'e-axis-container';
export var COLUMN_HEADER = 'e-columnheader';
export var SORT_FILTER_DIV = 'e-sortfilterdiv';
export var RESIZE_HANDLER = 'e-rhandler';
export var COPY_DROP = 'e-copy-drop';
export var ICON_EXPAND = 'e-icon-expandable';
export var LIST_ICON = 'e-list-icon';
export var CALC_DIMENSION_ICON = 'e-calc-dimension-icon';
export var CALC_MEASURE_ICON = 'e-calc-measure-icon';
export var MEASURE_ICON = 'e-measure-icon';
export var DIMENSION_TYPE_ICON = 'e-dimensionCDB-icon';
export var ATTRIBUTE_TYPE_ICON = 'e-attributeCDB-icon';
export var HIERARCHY_TYPE_ICON = 'e-hierarchyCDB-icon';
export var LEVEL_MEMBERS = 'e-level-members';
export var NAMEDSET_TYPE_ICON = 'e-namedSetCDB-icon';
export var NODE_FOCUS = 'e-node-focus';
export var HOVER = 'e-hover';
export var CALC_MEMBER = 'e-calc-member';
export var ACCORD_HEADER = 'e-acrdn-header';
export var ACCORDION = 'e-accordion';
export var LABEl = 'e-label';
export var TOGGLE_ICON = 'e-toggle-icon';
export var ACCORD_HEADER_ICON = 'e-acrdn-header-icon';
export var TOGGLE_COLLAPSE_ICON = 'e-tgl-collapse-icon';
export var ACCORD_ITEM = 'e-acrdn-item';
export var PROCESS = 'e-process';
export var FOLDER_TYPE_ICON = 'e-folderCDB-icon';
export var FOLDER_TYPE_OPEN_ICON = 'e-folderCDB-open-icon';
export var ICON_COLLAPSE = 'e-icon-collapsible';
export var CALC_FIELD_MEMBER = 'e-calcfieldmember';
export var ICON_SPACE = 'e-iconspace';
export var DRAG_ITEM = 'e-drag-item';
export var TREEVIEW_CLASS = 'e-treeview';
export var CHECKBOX_CONTAINER = 'e-checkbox-wrapper';
export var CALC_MEMBER_GROUP_TYPE_ICON = 'e-calcMemberGroupCDB';
export var MEASURE_GROUP_TYPE_ICON = 'e-measureGroupCDB-icon';
export var KPI_ICON = 'e-kpiCDB-icon';
export var KPI_GOAL_ICON = 'e-kpiGoal-icon';
export var KPI_STATUS_ICON = 'e-kpiStatus-icon';
export var KPI_TREND_ICON = 'e-kpiTrend-icon';
export var KPI_VALUE_ICON = 'e-kpiValue-icon';
export var VALUE_AXIS_CLASS = 'e-values';
export var AXIS_FILTER_CLASS = 'e-axis-filter';
export var SORT_CLASS = 'e-sort';
export var FILTER_COMMON_CLASS = 'e-btn-filter';
export var REMOVE_CLASS = 'e-remove';
export var POPUP = 'e-popup';
export var POPUP_OPEN = 'e-popup-open';
export var PREV_ACTIVE_NODE = 'e-prev-active-node';
export var FULL_ROW = 'e-fullrow';
export var PIVOT_VIEW_CLASS = 'e-pivotview';
export var GROUP_PIVOT_ROW = 'e-group-pivot-rows';
export var ROWSHEADER = 'e-rowsheader';
export var FOCUSED_CLASS = 'e-focused';
export var MENU_ITEM_CLASS = 'e-menu-item';
export var GRID_CLASS = 'e-grid';
export var EDITOR_TREE_OUTER_DIV_CLASS = 'e-member-editor-outer-container';
export var TOOLBAR_ITEM = 'e-toolbar-item';
export var TOOLBAR_ITEMS = 'e-toolbar-items';
export var MENU = 'e-menu';
export var MENU_PARENT = 'e-menu-parent';
export var MENU_CARET_ICON = 'e-menu-caret-icon';
export var SELECTED_CLASS = 'e-selected';
export var OVERLAY = 'e-overlay';
export var BUTTON_FOCUSED = 'e-btn-focused';
export var MOVABLE_FIRST = 'e-movablefirst';
export var ROW_SELECT = 'e-rowselect';
export var SPAN_CLICKED = 'e-spanclicked';
export var COLUMNSHEADER = 'e-columnsheader';
export var CELL_SELECTED_BGCOLOR = 'e-cellselectionbackground';
export var CELL_ACTIVE_BGCOLOR = 'e-active';
export var SELECTED_BGCOLOR = 'e-selectionbackground';
export var TABLE = 'e-table';
export var VIRTUALTRACK_DIV = 'e-virtualtrack';
export var ASCENDING_CLASS = 'e-ascending';
export var DESCENDING_CLASS = 'e-descending';
export var VALUESHEADER = 'e-valuesheader';
export var GRID_HEADER = 'e-gridheader';
export var GROUPING_BAR_CLASS = 'e-grouping-bar';
export var SUBTOTAL = 'e-stot';
export var PIVOT_FIELD_LIST_CLASS = 'e-pivotfieldlist';
export var EXPAND_ICON = 'e-expand';
export var COLLAPSE_ICON = 'e-collapse';
export var PIVOT_CHART = 'e-pivotchart';
export var COLOR_PICKER_CONTAINER = 'e-colorpicker-container';
export var FORMAT_FONT_COLOR_PICKER_CONTAINER = 'e-format-font-color-picker-container';
export var FORMAT_BACK_COLOR_PICKER_CONTAINER = 'e-format-back-color-picker-container';
export var FORMAT_COLOR_PICKER = 'e-format-color-picker';
export var FORMAT_FONT_COLOR_PICKER = 'e-format-font-color-picker';
export var COLOR_PICKER = 'e-colorpicker';
export var FORMAT_FONT_COLOR = 'e-format-font-color';
export var SELECTED_COLOR = 'e-selected-color';
export var FORMAT_BACK_COLOR = 'e-format-back-color';
