// @syncfusion/ej2/ej2-resource/scripts/ej2-smithchart/smithchart/utils/utils.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Property, ChildProperty } from '@syncfusion/ej2-base';
var Font = (function (_super) {
    __extends(Font, _super);
    function Font() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property('Segoe UI')
    ], Font.prototype, "fontFamily", void 0);
    __decorate([
        Property('Normal')
    ], Font.prototype, "fontStyle", void 0);
    __decorate([
        Property('Regular')
    ], Font.prototype, "fontWeight", void 0);
    __decorate([
        Property('12px')
    ], Font.prototype, "fontSize", void 0);
    __decorate([
        Property(1)
    ], Font.prototype, "opacity", void 0);
    return Font;
}(ChildProperty));
export { Font };
var Margin = (function (_super) {
    __extends(Margin, _super);
    function Margin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(10)
    ], Margin.prototype, "top", void 0);
    __decorate([
        Property(10)
    ], Margin.prototype, "bottom", void 0);
    __decorate([
        Property(10)
    ], Margin.prototype, "right", void 0);
    __decorate([
        Property(10)
    ], Margin.prototype, "left", void 0);
    return Margin;
}(ChildProperty));
export { Margin };
var Border = (function (_super) {
    __extends(Border, _super);
    function Border() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property(0)
    ], Border.prototype, "width", void 0);
    __decorate([
        Property(1)
    ], Border.prototype, "opacity", void 0);
    __decorate([
        Property('transparent')
    ], Border.prototype, "color", void 0);
    return Border;
}(ChildProperty));
export { Border };
var Size = (function (_super) {
    __extends(Size, _super);
    function Size() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        Property('')
    ], Size.prototype, "width", void 0);
    __decorate([
        Property('')
    ], Size.prototype, "height", void 0);
    return Size;
}(ChildProperty));
export { Size };
var Rect = (function () {
    function Rect(x, y, width, height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
    return Rect;
}());
export { Rect };
var LabelCollection = (function () {
    function LabelCollection() {
    }
    return LabelCollection;
}());
export { LabelCollection };
var LegendSeries = (function () {
    function LegendSeries() {
    }
    return LegendSeries;
}());
export { LegendSeries };
var LabelRegion = (function () {
    function LabelRegion() {
    }
    return LabelRegion;
}());
export { LabelRegion };
var HorizontalLabelCollection = (function (_super) {
    __extends(HorizontalLabelCollection, _super);
    function HorizontalLabelCollection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return HorizontalLabelCollection;
}(LabelCollection));
export { HorizontalLabelCollection };
var RadialLabelCollections = (function (_super) {
    __extends(RadialLabelCollections, _super);
    function RadialLabelCollections() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RadialLabelCollections;
}(HorizontalLabelCollection));
export { RadialLabelCollections };
var LineSegment = (function () {
    function LineSegment() {
    }
    return LineSegment;
}());
export { LineSegment };
var PointRegion = (function () {
    function PointRegion() {
    }
    return PointRegion;
}());
export { PointRegion };
var Point = (function () {
    function Point() {
    }
    return Point;
}());
export { Point };
var ClosestPoint = (function () {
    function ClosestPoint() {
    }
    return ClosestPoint;
}());
export { ClosestPoint };
var MarkerOptions = (function () {
    function MarkerOptions(id, fill, borderColor, borderWidth, opacity) {
        this.id = id;
        this.fill = fill;
        this.borderColor = borderColor;
        this.borderWidth = borderWidth;
        this.opacity = opacity;
    }
    return MarkerOptions;
}());
export { MarkerOptions };
var LabelPosition = (function () {
    function LabelPosition() {
    }
    return LabelPosition;
}());
export { LabelPosition };
var Direction = (function () {
    function Direction() {
        this.counterclockwise = 0;
        this.clockwise = 1;
    }
    return Direction;
}());
export { Direction };
var DataLabelTextOptions = (function () {
    function DataLabelTextOptions() {
    }
    return DataLabelTextOptions;
}());
export { DataLabelTextOptions };
var LabelOption = (function () {
    function LabelOption() {
    }
    return LabelOption;
}());
export { LabelOption };
var SmithchartSize = (function () {
    function SmithchartSize(width, height) {
        this.width = width;
        this.height = height;
    }
    return SmithchartSize;
}());
export { SmithchartSize };
var GridArcPoints = (function () {
    function GridArcPoints() {
    }
    return GridArcPoints;
}());
export { GridArcPoints };
