// @syncfusion/ej2/ej2-resource/scripts/ej2-drawings/drawing/enum/enum.js
export var RotateTransform;
(function (RotateTransform) {
    RotateTransform[RotateTransform["Self"] = 1] = "Self";
    RotateTransform[RotateTransform["Parent"] = 2] = "Parent";
})(RotateTransform || (RotateTransform = {}));
export var ElementAction;
(function (ElementAction) {
    ElementAction[ElementAction["None"] = 0] = "None";
    ElementAction[ElementAction["ElementIsPort"] = 2] = "ElementIsPort";
    ElementAction[ElementAction["ElementIsGroup"] = 4] = "ElementIsGroup";
})(ElementAction || (ElementAction = {}));
export var ConnectorConstraints;
(function (ConnectorConstraints) {
    ConnectorConstraints[ConnectorConstraints["None"] = 1] = "None";
    ConnectorConstraints[ConnectorConstraints["Select"] = 2] = "Select";
    ConnectorConstraints[ConnectorConstraints["Delete"] = 4] = "Delete";
    ConnectorConstraints[ConnectorConstraints["Drag"] = 8] = "Drag";
    ConnectorConstraints[ConnectorConstraints["DragSourceEnd"] = 16] = "DragSourceEnd";
    ConnectorConstraints[ConnectorConstraints["DragTargetEnd"] = 32] = "DragTargetEnd";
    ConnectorConstraints[ConnectorConstraints["DragSegmentThumb"] = 64] = "DragSegmentThumb";
    ConnectorConstraints[ConnectorConstraints["AllowDrop"] = 128] = "AllowDrop";
    ConnectorConstraints[ConnectorConstraints["Bridging"] = 256] = "Bridging";
    ConnectorConstraints[ConnectorConstraints["BridgeObstacle"] = 512] = "BridgeObstacle";
    ConnectorConstraints[ConnectorConstraints["InheritBridging"] = 1024] = "InheritBridging";
    ConnectorConstraints[ConnectorConstraints["PointerEvents"] = 2048] = "PointerEvents";
    ConnectorConstraints[ConnectorConstraints["Tooltip"] = 4096] = "Tooltip";
    ConnectorConstraints[ConnectorConstraints["InheritTooltip"] = 8192] = "InheritTooltip";
    ConnectorConstraints[ConnectorConstraints["Interaction"] = 4218] = "Interaction";
    ConnectorConstraints[ConnectorConstraints["ReadOnly"] = 16384] = "ReadOnly";
    ConnectorConstraints[ConnectorConstraints["Default"] = 11838] = "Default";
})(ConnectorConstraints || (ConnectorConstraints = {}));
export var ThumbsConstraints;
(function (ThumbsConstraints) {
    ThumbsConstraints[ThumbsConstraints["Rotate"] = 2] = "Rotate";
    ThumbsConstraints[ThumbsConstraints["ConnectorSource"] = 4] = "ConnectorSource";
    ThumbsConstraints[ThumbsConstraints["ConnectorTarget"] = 8] = "ConnectorTarget";
    ThumbsConstraints[ThumbsConstraints["ResizeNorthEast"] = 16] = "ResizeNorthEast";
    ThumbsConstraints[ThumbsConstraints["ResizeEast"] = 32] = "ResizeEast";
    ThumbsConstraints[ThumbsConstraints["ResizeSouthEast"] = 64] = "ResizeSouthEast";
    ThumbsConstraints[ThumbsConstraints["ResizeSouth"] = 128] = "ResizeSouth";
    ThumbsConstraints[ThumbsConstraints["ResizeSouthWest"] = 256] = "ResizeSouthWest";
    ThumbsConstraints[ThumbsConstraints["ResizeWest"] = 512] = "ResizeWest";
    ThumbsConstraints[ThumbsConstraints["ResizeNorthWest"] = 1024] = "ResizeNorthWest";
    ThumbsConstraints[ThumbsConstraints["ResizeNorth"] = 2048] = "ResizeNorth";
    ThumbsConstraints[ThumbsConstraints["Default"] = 4094] = "Default";
})(ThumbsConstraints || (ThumbsConstraints = {}));
export var SelectorConstraints;
(function (SelectorConstraints) {
    SelectorConstraints[SelectorConstraints["None"] = 1] = "None";
    SelectorConstraints[SelectorConstraints["ConnectorSourceThumb"] = 2] = "ConnectorSourceThumb";
    SelectorConstraints[SelectorConstraints["ConnectorTargetThumb"] = 4] = "ConnectorTargetThumb";
    SelectorConstraints[SelectorConstraints["ResizeSouthEast"] = 8] = "ResizeSouthEast";
    SelectorConstraints[SelectorConstraints["ResizeSouthWest"] = 16] = "ResizeSouthWest";
    SelectorConstraints[SelectorConstraints["ResizeNorthEast"] = 32] = "ResizeNorthEast";
    SelectorConstraints[SelectorConstraints["ResizeNorthWest"] = 64] = "ResizeNorthWest";
    SelectorConstraints[SelectorConstraints["ResizeEast"] = 128] = "ResizeEast";
    SelectorConstraints[SelectorConstraints["ResizeWest"] = 256] = "ResizeWest";
    SelectorConstraints[SelectorConstraints["ResizeSouth"] = 512] = "ResizeSouth";
    SelectorConstraints[SelectorConstraints["ResizeNorth"] = 1024] = "ResizeNorth";
    SelectorConstraints[SelectorConstraints["Rotate"] = 2048] = "Rotate";
    SelectorConstraints[SelectorConstraints["UserHandle"] = 4096] = "UserHandle";
    SelectorConstraints[SelectorConstraints["ToolTip"] = 8192] = "ToolTip";
    SelectorConstraints[SelectorConstraints["ResizeAll"] = 2046] = "ResizeAll";
    SelectorConstraints[SelectorConstraints["All"] = 16382] = "All";
})(SelectorConstraints || (SelectorConstraints = {}));
export var NoOfSegments;
(function (NoOfSegments) {
    NoOfSegments[NoOfSegments["Zero"] = 0] = "Zero";
    NoOfSegments[NoOfSegments["One"] = 1] = "One";
    NoOfSegments[NoOfSegments["Two"] = 2] = "Two";
    NoOfSegments[NoOfSegments["Three"] = 3] = "Three";
    NoOfSegments[NoOfSegments["Four"] = 4] = "Four";
    NoOfSegments[NoOfSegments["Five"] = 5] = "Five";
})(NoOfSegments || (NoOfSegments = {}));
