// @syncfusion/ej2/ej2-resource/scripts/ej2-pivotview/blazor/pivotview/selection.js
import * as cls from '../common/constants';
import { EventHandler } from '@syncfusion/ej2-base';
import { removeClass, addClass } from '@syncfusion/ej2-base';
var Selection = (function () {
    function Selection(parent) {
        this.isPopupClicked = false;
        this.shiftLockedPos = [];
        this.savedSelectedCellsPos = [];
        this.cellSelectionPos = [];
        this.parent = parent;
        this.parent.selectionModule = this;
    }
    Selection.prototype.addInternalEvents = function () {
        this.wireEvents();
    };
    Selection.prototype.wireEvents = function () {
        this.unWireEvents();
        EventHandler.add(this.parent.element, this.parent.isAdaptive ? 'touchend' : 'click', this.mouseClickHandler, this);
    };
    Selection.prototype.unWireEvents = function () {
        EventHandler.remove(this.parent.element, this.parent.isAdaptive ? 'touchend' : 'click', this.mouseClickHandler);
    };
    Selection.prototype.mouseClickHandler = function (e) {
        if (e.which === 3) {
            this.lastCellClicked = e.target;
        }
        else if (e.which === 0) {
            this.lastCellClicked = e.target;
        }
        var target = e.target;
        if ((target.classList.contains(cls.HEADERCELL) ||
            target.classList.contains(cls.HEADER_CELL_DIV) ||
            target.classList.contains(cls.ROWSHEADER) ||
            target.classList.contains(cls.ROW_CELL_CLASS) ||
            target.classList.contains(cls.STACKED_HEADER_CELL_DIV) ||
            target.classList.contains(cls.HEADER_TEXT) ||
            target.classList.contains(cls.STACKED_HEADER_TEXT)) && this.parent.options.dataType === 'pivot') {
            var targetElement = null;
            if (target.classList.contains(cls.HEADERCELL) || target.classList.contains(cls.ROWSHEADER)
                || target.classList.contains(cls.ROW_CELL_CLASS)) {
                targetElement = target;
            }
            else if (target.classList.contains(cls.STACKED_HEADER_CELL_DIV) ||
                target.classList.contains(cls.CELLVALUE) || target.classList.contains(cls.HEADER_CELL_DIV) ||
                target.classList.contains(cls.ASCENDING_CLASS) || target.classList.contains(cls.DESCENDING_CLASS)) {
                targetElement = target.parentElement;
            }
            else if (target.classList.contains(cls.HEADER_TEXT)) {
                targetElement = target.parentElement.parentElement;
            }
            this.cellClicked(target, e);
        }
        else {
            this.cellClicked(target, e);
            return;
        }
    };
    Selection.prototype.cellClicked = function (target, e) {
        var targetElement = null;
        if (target.classList.contains(cls.HEADERCELL) ||
            target.classList.contains(cls.STACKED_HEADER_CELL) || target.classList.contains(cls.ROW_CELL_CLASS)) {
            targetElement = target;
        }
        else if (target.classList.contains(cls.STACKED_HEADER_CELL_DIV) ||
            target.classList.contains(cls.CELLVALUE) || target.classList.contains(cls.HEADER_CELL_DIV)) {
            targetElement = target.parentElement;
        }
        else if (target.classList.contains(cls.HEADER_TEXT)) {
            targetElement = target.parentElement.parentElement.parentElement;
        }
        else if (target.classList.contains(cls.STACKED_HEADER_TEXT)) {
            targetElement = target.parentElement.parentElement;
        }
        else if (target.classList.contains(cls.ROW_SELECT)) {
            if (target.classList.contains(cls.SPAN_CLICKED)) {
                this.isPopupClicked = false;
            }
            else {
                this.isPopupClicked = true;
            }
        }
        if (targetElement) {
            var colIndex = Number(targetElement.getAttribute('aria-colindex'));
            var rowIndex = Number(targetElement.getAttribute('index'));
            var colSpan = Number(targetElement.getAttribute('aria-colspan'));
            if (this.parent.gridSettings.allowSelection) {
                if (this.parent.gridSettings.selectionSettings.mode === 'Both' ? !targetElement.classList.contains(cls.ROW_CELL_CLASS) : this.parent.gridSettings.selectionSettings.mode !== 'Row') {
                    this.clearSelection(targetElement, e, colIndex, rowIndex);
                    this.applyColumnSelection(e, targetElement, colIndex, colIndex + (colSpan > 0 ? colSpan - 1 : 0), rowIndex);
                }
                else {
                    this.clearSelection(targetElement, e, colIndex, rowIndex);
                }
            }
            this.getSelectedCellsPos();
            var cellPos = JSON.stringify(this.savedSelectedCellsPos);
            if (this.parent.gridSettings.selectionSettings.mode === 'Both' ? !targetElement.classList.contains(cls.ROW_CELL_CLASS) : this.parent.gridSettings.selectionSettings.mode === 'Cell' ?
                targetElement.classList.contains(cls.COLUMNSHEADER) : this.parent.gridSettings.selectionSettings.mode !== 'Row') {
                this.parent.dotNetRef.invokeMethodAsync('CellClickedHandler', rowIndex, colIndex, e, JSON.stringify(window.sfBlazor.getDomObject('cellElement', targetElement)));
                this.parent.dotNetRef.invokeMethodAsync('SelectHandler', colIndex, rowIndex, cellPos);
                this.savedSelectedCellsPos = [];
            }
            if (!this.parent.gridSettings.allowSelection && !(target.classList.contains(cls.EXPAND_ICON) || target.classList.contains(cls.COLLAPSE_ICON))) {
                this.parent.dotNetRef.invokeMethodAsync('CellClickedHandler', rowIndex, colIndex, e, JSON.stringify(window.sfBlazor.getDomObject('cellElement', targetElement)));
            }
        }
    };
    Selection.prototype.clearSelection = function (targetElement, e, colIndex, rowIndex) {
        if ((!e.shiftKey && !e.ctrlKey) || this.parent.gridSettings.selectionSettings.type === 'Single') {
            if (this.parent.gridSettings.selectionSettings.mode === 'Cell') {
                if (targetElement.classList.contains(cls.COLUMNSHEADER)) {
                    removeClass(this.parent.element.querySelectorAll(('.' + cls.ROW_CELL_CLASS + '.') + cls.CELL_SELECTED_BGCOLOR), cls.CELL_SELECTED_BGCOLOR);
                }
                else {
                    removeClass(this.parent.element.querySelectorAll(('.' + cls.COLUMNSHEADER + '.') + cls.CELL_ACTIVE_BGCOLOR), [cls.CELL_ACTIVE_BGCOLOR, cls.SELECTED_BGCOLOR]);
                }
            }
            else if (this.parent.gridSettings.selectionSettings.mode === 'Both') {
                if (targetElement.classList.contains(cls.ROW_CELL_CLASS)) {
                    for (var _i = 0, _a = [].slice.call(this.parent.element.querySelectorAll('.' + cls.SELECTED_BGCOLOR + ', .' + cls.CELL_SELECTED_BGCOLOR)); _i < _a.length; _i++) {
                        var ele = _a[_i];
                        removeClass([ele], [cls.CELL_ACTIVE_BGCOLOR, cls.SELECTED_BGCOLOR, cls.CELL_SELECTED_BGCOLOR]);
                    }
                }
                else {
                    removeClass(this.parent.element.querySelectorAll('.' + cls.CELL_SELECTED_BGCOLOR), cls.CELL_SELECTED_BGCOLOR);
                }
            }
        }
    };
    Selection.prototype.applyColumnSelection = function (e, target, colStart, colEnd, rowStart) {
        var rowIndex = Number(target.getAttribute('index'));
        if (!target.classList.contains(cls.ROWSHEADER) || (this.parent.gridSettings.selectionSettings.mode === 'Cell' ? target.classList.contains(cls.COLUMNSHEADER) : true)) {
            var isCtrl = e.ctrlKey;
            if (this.parent.gridSettings.selectionSettings.type === 'Multiple' && this.parent.element.querySelector('.' + cls.ROW_SELECT) !== null) {
                if (this.isPopupClicked) {
                    this.parent.element.querySelector('.' + cls.ROW_SELECT).classList.add(cls.SPAN_CLICKED);
                    isCtrl = true;
                }
                else {
                    this.parent.element.querySelector('.' + cls.ROW_SELECT).classList.remove(cls.SPAN_CLICKED);
                    isCtrl = false;
                }
            }
            var queryStringArray = [];
            var type = this.parent.gridSettings.selectionSettings.type;
            var isToggle = target.classList.contains(cls.CELL_ACTIVE_BGCOLOR);
            var activeColumns = [];
            var actColPos = {};
            for (var cCnt = colStart; cCnt <= colEnd; cCnt++) {
                activeColumns.push(cCnt.toString());
            }
            if (!isCtrl || type === 'Single') {
                for (var _i = 0, _a = [].slice.call(this.parent.element.querySelectorAll('.' + cls.CELL_ACTIVE_BGCOLOR)); _i < _a.length; _i++) {
                    var targetElement = _a[_i];
                    removeClass([targetElement], [cls.CELL_ACTIVE_BGCOLOR, cls.SELECTED_BGCOLOR]);
                    if (activeColumns.indexOf(targetElement.getAttribute('aria-colindex')) === -1) {
                        isToggle = false;
                    }
                    var colIndex = Number(targetElement.getAttribute('aria-colindex'));
                    actColPos[colIndex] = colIndex;
                }
                activeColumns = Object.keys(actColPos).length > 0 ? Object.keys(actColPos).sort(function (a, b) {
                    return a - b;
                }) : activeColumns;
            }
            else {
                isToggle = false;
            }
            if (type === 'Multiple' && e.shiftKey) {
                this.shiftLockedPos = this.shiftLockedPos.length === 0 ? activeColumns : this.shiftLockedPos;
                if (Number(this.shiftLockedPos[0]) <= colStart) {
                    colStart = Number(this.shiftLockedPos[0]);
                }
                else {
                    colEnd = colEnd < Number(this.shiftLockedPos[this.shiftLockedPos.length - 1]) ?
                        Number(this.shiftLockedPos[this.shiftLockedPos.length - 1]) : colEnd;
                }
            }
            else {
                this.shiftLockedPos = [];
            }
            var rowSelectedList = [];
            if (e.ctrlKey && this.parent.gridSettings.selectionSettings.mode === 'Both' && type === 'Multiple' && !target.classList.contains(cls.ROWSHEADER)) {
                for (var _b = 0, _c = [].slice.call(this.parent.element.querySelectorAll('.' + cls.ROWSHEADER + '.' + cls.CELL_SELECTED_BGCOLOR)); _b < _c.length; _b++) {
                    var targetElement = _c[_b];
                    rowSelectedList.push(targetElement.getAttribute('index'));
                }
            }
            var count = colStart;
            while (count <= colEnd) {
                queryStringArray.push('[aria-colindex="' + count + '"]' + (this.parent.gridSettings.selectionSettings.mode === 'Cell' ? '[index="' + rowStart + '"]' : '') + '');
                count++;
            }
            if (!isToggle) {
                rowStart = target.classList.contains(cls.HEADERCELL) ? rowStart : (this.parent.scrollPageInfo.rowStartPos - 1);
                var isTargetSelected = target.classList.contains(cls.CELL_ACTIVE_BGCOLOR);
                for (var _d = 0, _e = [].slice.call(this.parent.element.querySelectorAll(queryStringArray.toString())); _d < _e.length; _d++) {
                    var targetElement = _e[_d];
                    if (Number(targetElement.getAttribute('index')) >= rowStart) {
                        if (isTargetSelected && isCtrl && rowSelectedList.indexOf(rowIndex.toString()) === -1) {
                            removeClass([targetElement], [cls.CELL_ACTIVE_BGCOLOR, cls.SELECTED_BGCOLOR]);
                        }
                        else {
                            addClass([targetElement], [cls.CELL_ACTIVE_BGCOLOR, cls.SELECTED_BGCOLOR]);
                        }
                    }
                }
            }
        }
    };
    Selection.prototype.getSelectedCellsPos = function () {
        for (var _i = 0, _a = [].slice.call(this.parent.element.querySelectorAll('.' + cls.SELECTED_BGCOLOR)); _i < _a.length; _i++) {
            var targetElement = _a[_i];
            this.savedSelectedCellsPos.push({ rowIndex: targetElement.getAttribute('index'), colIndex: targetElement.getAttribute('aria-colindex') });
        }
        for (var _b = 0, _c = [].slice.call(this.parent.element.querySelectorAll('.' + cls.CELL_SELECTED_BGCOLOR)); _b < _c.length; _b++) {
            var targetElement = _c[_b];
            this.cellSelectionPos.push({ rowIndex: targetElement.getAttribute('index'), colIndex: targetElement.getAttribute('aria-colindex') });
        }
    };
    Selection.prototype.destroy = function () {
        this.unWireEvents();
    };
    return Selection;
}());
export { Selection };
