// @syncfusion/ej2/ej2-resource/scripts/ej2-pivotview/blazor/pivotview/toolbar.js
import * as cls from '../common/constants';
import { removeClass, addClass } from '@syncfusion/ej2-base';
var Toolbar = (function () {
    function Toolbar(parent) {
        this.parent = parent;
        this.parent.toolbarModule = this;
    }
    Toolbar.prototype.renderToolbar = function () {
        if (this.parent.element.querySelector('.' + cls.GRID_TOOLBAR)) {
            this.updateItemElements(this.parent.element.querySelector('.' + cls.GRID_TOOLBAR));
        }
    };
    Toolbar.prototype.updateItemElements = function (toolbarElement) {
        for (var _i = 0, _a = [].slice.call(toolbarElement.querySelectorAll('.' + cls.TOOLBAR_ITEM)); _i < _a.length; _i++) {
            var element = _a[_i];
            if (element.querySelector('button')) {
                if (!element.classList.contains(cls.OVERLAY)) {
                    element.querySelector('button').setAttribute('tabindex', '0');
                }
            }
            else if (element.querySelector('.' + cls.MENU + '.' + cls.MENU_PARENT)) {
                element.querySelector('.' + cls.MENU + '.' + cls.MENU_PARENT).setAttribute('tabindex', '-1');
                if (!element.classList.contains(cls.OVERLAY) &&
                    element.querySelector('.' + cls.MENU_ITEM_CLASS + '.' + cls.MENU_CARET_ICON)) {
                    element.querySelector('.' + cls.MENU_ITEM_CLASS + '.' + cls.MENU_CARET_ICON).setAttribute('tabindex', '0');
                }
            }
        }
    };
    Toolbar.prototype.focusToolBar = function () {
        removeClass(document.querySelector('.' + cls.GRID_TOOLBAR).querySelectorAll('.' + cls.MENU_ITEM_CLASS + '.' + cls.FOCUSED_CLASS), cls.FOCUSED_CLASS);
        removeClass(document.querySelector('.' + cls.GRID_TOOLBAR).querySelectorAll('.' + cls.MENU_ITEM_CLASS + '.' + cls.SELECTED_CLASS), cls.SELECTED_CLASS);
        if (document.querySelector('.' + cls.TOOLBAR_ITEMS)) {
            addClass([document.querySelector('.' + cls.TOOLBAR_ITEMS)], cls.FOCUSED_CLASS);
        }
    };
    Toolbar.prototype.selectInputRange = function (element) {
        if (element.querySelector('.' + cls.GRID_REPORT_INPUT)) {
            var input = element.querySelector('.' + cls.GRID_REPORT_INPUT + ' input');
            input.setSelectionRange(input.value.length, input.value.length);
        }
    };
    Toolbar.prototype.copyMdxQuery = function (element) {
        if (element.querySelector('.' + cls.MDX_QUERY_CONTENT)) {
            var textArea = element.querySelector('.' + cls.MDX_QUERY_CONTENT);
            try {
                textArea.select();
                document.execCommand('copy');
            }
            catch (error) {
                window.alert('Oops, unable to copy');
            }
            return;
        }
    };
    return Toolbar;
}());
export { Toolbar };
