// @syncfusion/ej2/ej2-resource/scripts/ej2-pdf-export/implementation/pages/enum.js
export var PdfPageOrientation;
(function (PdfPageOrientation) {
    PdfPageOrientation[PdfPageOrientation["Portrait"] = 0] = "Portrait";
    PdfPageOrientation[PdfPageOrientation["Landscape"] = 1] = "Landscape";
})(PdfPageOrientation || (PdfPageOrientation = {}));
export var PdfPageRotateAngle;
(function (PdfPageRotateAngle) {
    PdfPageRotateAngle[PdfPageRotateAngle["RotateAngle0"] = 0] = "RotateAngle0";
    PdfPageRotateAngle[PdfPageRotateAngle["RotateAngle90"] = 1] = "RotateAngle90";
    PdfPageRotateAngle[PdfPageRotateAngle["RotateAngle180"] = 2] = "RotateAngle180";
    PdfPageRotateAngle[PdfPageRotateAngle["RotateAngle270"] = 3] = "RotateAngle270";
})(PdfPageRotateAngle || (PdfPageRotateAngle = {}));
export var PdfNumberStyle;
(function (PdfNumberStyle) {
    PdfNumberStyle[PdfNumberStyle["None"] = 0] = "None";
    PdfNumberStyle[PdfNumberStyle["Numeric"] = 1] = "Numeric";
    PdfNumberStyle[PdfNumberStyle["LowerLatin"] = 2] = "LowerLatin";
    PdfNumberStyle[PdfNumberStyle["LowerRoman"] = 3] = "LowerRoman";
    PdfNumberStyle[PdfNumberStyle["UpperLatin"] = 4] = "UpperLatin";
    PdfNumberStyle[PdfNumberStyle["UpperRoman"] = 5] = "UpperRoman";
})(PdfNumberStyle || (PdfNumberStyle = {}));
export var PdfDockStyle;
(function (PdfDockStyle) {
    PdfDockStyle[PdfDockStyle["None"] = 0] = "None";
    PdfDockStyle[PdfDockStyle["Bottom"] = 1] = "Bottom";
    PdfDockStyle[PdfDockStyle["Top"] = 2] = "Top";
    PdfDockStyle[PdfDockStyle["Left"] = 3] = "Left";
    PdfDockStyle[PdfDockStyle["Right"] = 4] = "Right";
    PdfDockStyle[PdfDockStyle["Fill"] = 5] = "Fill";
})(PdfDockStyle || (PdfDockStyle = {}));
export var PdfAlignmentStyle;
(function (PdfAlignmentStyle) {
    PdfAlignmentStyle[PdfAlignmentStyle["None"] = 0] = "None";
    PdfAlignmentStyle[PdfAlignmentStyle["TopLeft"] = 1] = "TopLeft";
    PdfAlignmentStyle[PdfAlignmentStyle["TopCenter"] = 2] = "TopCenter";
    PdfAlignmentStyle[PdfAlignmentStyle["TopRight"] = 3] = "TopRight";
    PdfAlignmentStyle[PdfAlignmentStyle["MiddleLeft"] = 4] = "MiddleLeft";
    PdfAlignmentStyle[PdfAlignmentStyle["MiddleCenter"] = 5] = "MiddleCenter";
    PdfAlignmentStyle[PdfAlignmentStyle["MiddleRight"] = 6] = "MiddleRight";
    PdfAlignmentStyle[PdfAlignmentStyle["BottomLeft"] = 7] = "BottomLeft";
    PdfAlignmentStyle[PdfAlignmentStyle["BottomCenter"] = 8] = "BottomCenter";
    PdfAlignmentStyle[PdfAlignmentStyle["BottomRight"] = 9] = "BottomRight";
})(PdfAlignmentStyle || (PdfAlignmentStyle = {}));
export var TemplateType;
(function (TemplateType) {
    TemplateType[TemplateType["None"] = 0] = "None";
    TemplateType[TemplateType["Top"] = 1] = "Top";
    TemplateType[TemplateType["Bottom"] = 2] = "Bottom";
    TemplateType[TemplateType["Left"] = 3] = "Left";
    TemplateType[TemplateType["Right"] = 4] = "Right";
})(TemplateType || (TemplateType = {}));
