// @syncfusion/ej2/ej2-resource/scripts/ej2-pdf-export/implementation/pages/pdf-section-collection.js
import { PdfNumber } from './../primitives/pdf-number';
import { PdfName } from './../primitives/pdf-name';
import { PdfArray } from './../primitives/pdf-array';
import { PdfDictionary, SaveSectionCollectionEventHandler } from './../primitives/pdf-dictionary';
import { PdfReferenceHolder } from './../primitives/pdf-reference';
import { RectangleF, PointF } from './../drawing/pdf-drawing';
import { PdfSection } from './pdf-section';
import { DictionaryProperties } from './../input-output/pdf-dictionary-properties';
var PdfSectionCollection = (function () {
    function PdfSectionCollection(document) {
        this.sections = [];
        this.dictionaryProperties = new DictionaryProperties();
        this.pdfDocument = document.clone();
        this.initialize();
    }
    Object.defineProperty(PdfSectionCollection.prototype, "section", {
        get: function () {
            return this.sections;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfSectionCollection.prototype, "document", {
        get: function () {
            return this.pdfDocument;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfSectionCollection.prototype, "count", {
        get: function () {
            return this.sections.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PdfSectionCollection.prototype, "element", {
        get: function () {
            return this.pages;
        },
        enumerable: true,
        configurable: true
    });
    PdfSectionCollection.prototype.initialize = function () {
        this.sectionCount = new PdfNumber(0);
        this.sectionCollection = new PdfArray();
        this.pages = new PdfDictionary();
        this.pages.beginSave = new SaveSectionCollectionEventHandler(this);
        this.pages.items.setValue(this.dictionaryProperties.type, new PdfName('Pages'));
        this.pages.items.setValue(this.dictionaryProperties.kids, this.sectionCollection);
        this.pages.items.setValue(this.dictionaryProperties.count, this.sectionCount);
        this.pages.items.setValue(this.dictionaryProperties.resources, new PdfDictionary());
        this.setPageSettings(this.pages, this.pdfDocument.pageSettings);
    };
    PdfSectionCollection.prototype.pdfSectionCollection = function (index) {
        if (index < 0 || index >= this.count) {
            throw new Error('IndexOutOfRangeException()');
        }
        return this.sections[index];
    };
    PdfSectionCollection.prototype.setPageSettings = function (container, pageSettings) {
        var bounds = new RectangleF(new PointF(), pageSettings.size);
        container.items.setValue(this.dictionaryProperties.mediaBox, PdfArray.fromRectangle(bounds));
    };
    PdfSectionCollection.prototype.add = function (section) {
        if (typeof section === 'undefined') {
            var section_1 = new PdfSection(this.pdfDocument);
            this.add(section_1);
            return section_1;
        }
        else {
            var r = this.checkSection(section);
            this.sections.push(section);
            section.parent = this;
            this.sectionCollection.add(r);
            return this.sections.indexOf(section);
        }
    };
    PdfSectionCollection.prototype.checkSection = function (section) {
        var r = new PdfReferenceHolder(section);
        var contains = this.sectionCollection.contains(r);
        return r;
    };
    PdfSectionCollection.prototype.countPages = function () {
        var count = 0;
        this.sections.forEach(function (n) { return (count += n.count); });
        return count;
    };
    PdfSectionCollection.prototype.beginSave = function () {
        this.sectionCount.intValue = this.countPages();
    };
    PdfSectionCollection.rotateFactor = 90;
    return PdfSectionCollection;
}());
export { PdfSectionCollection };
