// @syncfusion/ej2/ej2-resource/scripts/ej2-treegrid/blazor/sf-treegrid-fn.js
import { EventHandler } from '@syncfusion/ej2-base';
import { Clipboard } from './clipboard';
import { RowDD } from './rowdragdrop';
import { CheckboxColumn } from './checkboxcolumn';
var SfTreeGrid = (function () {
    function SfTreeGrid(element, options, dotnetRef) {
        this.columnModel = [];
        this.element = element;
        this.element.blazor_instance = this;
        this.grid = element.querySelector('#' + element.id + '_gridcontrol').blazor__instance;
        this.dotNetRef = dotnetRef;
        this.options = options;
        this.header = this.element.querySelector('.e-headercontent');
        this.content = this.element.querySelector('.e-content');
        this.footer = this.element.querySelector('.e-summarycontent');
        this.clipboardModule = new Clipboard(this);
        this.rowDragAndDropModule = new RowDD(this);
        this.checkboxcolumnModule = new CheckboxColumn(dotnetRef, this);
        this.wireEvents();
    }
    SfTreeGrid.prototype.wireEvents = function () {
        EventHandler.add(this.element, 'keydown', this.KeyDownHandler, this);
    };
    SfTreeGrid.prototype.getContent = function () { return this.content; };
    SfTreeGrid.prototype.getContentTable = function () { return this.content.querySelector('.e-table'); };
    SfTreeGrid.prototype.getHeaderContent = function () { return this.header; };
    SfTreeGrid.prototype.getHeaderTable = function () { return this.header.querySelector('.e-table'); };
    SfTreeGrid.prototype.getRows = function () { return [].slice.call(this.getContent().querySelectorAll('.e-row')); };
    SfTreeGrid.prototype.KeyDownHandler = function (e) {
        if (e.keyCode === 67 && e.ctrlKey) {
            this.clipboardModule.copy();
        }
        else if (e.keyCode === 72 && e.ctrlKey && e.shiftKey) {
            this.clipboardModule.copy(true);
        }
    };
    SfTreeGrid.prototype.unWireEvents = function () {
        EventHandler.remove(this.element, 'keydown', this.KeyDownHandler);
    };
    SfTreeGrid.prototype.destroy = function () {
        this.unWireEvents();
    };
    return SfTreeGrid;
}());
export { SfTreeGrid };
