// @syncfusion/ej2/ej2-resource/scripts/ej2-maps/maps/model/constants.js
export var load = 'load';
export var loaded = 'loaded';
export var click = 'click';
export var rightClick = 'rightClick';
export var doubleClick = 'doubleClick';
export var resize = 'resize';
export var tooltipRender = 'tooltipRender';
export var shapeSelected = 'shapeSelected';
export var shapeHighlight = 'shapeHighlight';
export var mousemove = 'mousemove';
export var mouseup = 'mouseup';
export var mousedown = 'mousedown';
export var layerRendering = 'layerRendering';
export var shapeRendering = 'shapeRendering';
export var markerRendering = 'markerRendering';
export var markerClick = 'markerClick';
export var markerMouseMove = 'markerMouseMove';
export var dataLabelRendering = 'dataLabelRendering';
export var bubbleRendering = 'bubbleRendering';
export var bubbleClick = 'bubbleClick';
export var bubbleMouseMove = 'bubbleMouseMove';
export var animationComplete = 'animationComplete';
export var legendRendering = 'legendRendering';
export var annotationRendering = 'annotationRendering';
export var itemSelection = 'itemSelection';
export var itemHighlight = 'itemHighlight';
export var beforePrint = 'beforePrint';
export var zoomIn = 'zoomIn';
export var zoomOut = 'zoomOut';
export var pan = 'pan';
