// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/rich-text-editor/renderer/content-renderer.js
import { getEditValue } from '../base/util';
var ContentRender = (function () {
    function ContentRender(parent, serviceLocator) {
        this.parent = parent;
        this.serviceLocator = serviceLocator;
    }
    ContentRender.prototype.renderPanel = function () {
        var rteObj = this.parent;
        var div = this.parent.createElement('div', { className: 'e-rte-content', id: this.parent.getID() + 'rte-view' });
        var rteContent = getEditValue(rteObj.value, rteObj);
        this.editableElement = this.parent.createElement('div', {
            className: 'e-content',
            id: this.parent.getID() + '_rte-edit-view',
            attrs: {
                'contenteditable': 'true'
            },
            innerHTML: rteContent
        });
        div.appendChild(this.editableElement);
        this.setPanel(div);
        rteObj.element.appendChild(div);
    };
    ContentRender.prototype.getPanel = function () {
        return this.contentPanel;
    };
    ContentRender.prototype.getEditPanel = function () {
        return this.editableElement;
    };
    ContentRender.prototype.getText = function () {
        return this.getEditPanel().innerText;
    };
    ContentRender.prototype.setPanel = function (panel) {
        this.contentPanel = panel;
    };
    ContentRender.prototype.getDocument = function () {
        return this.getEditPanel().ownerDocument;
    };
    return ContentRender;
}());
export { ContentRender };
