// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/rich-text-editor/renderer/markdown-renderer.js
var MarkdownRender = (function () {
    function MarkdownRender(parent) {
        this.parent = parent;
    }
    MarkdownRender.prototype.renderPanel = function () {
        var rteObj = this.parent;
        var div = this.parent.createElement('div', { id: this.parent.getID() + '_view', className: 'e-rte-content' });
        this.editableElement = this.parent.createElement('textarea', {
            className: 'e-content',
            id: this.parent.getID() + '_editable-content'
        });
        div.appendChild(this.editableElement);
        this.setPanel(div);
        rteObj.element.appendChild(div);
    };
    MarkdownRender.prototype.getPanel = function () {
        return this.contentPanel;
    };
    MarkdownRender.prototype.getEditPanel = function () {
        return this.editableElement;
    };
    MarkdownRender.prototype.getText = function () {
        return this.getEditPanel().value;
    };
    MarkdownRender.prototype.setPanel = function (panel) {
        this.contentPanel = panel;
    };
    MarkdownRender.prototype.getDocument = function () {
        return this.getEditPanel().ownerDocument;
    };
    return MarkdownRender;
}());
export { MarkdownRender };
