// @syncfusion/ej2/ej2-resource/scripts/ej2-barcode-generator/barcode/enum/enum.js
export var BarcodeEvent;
(function (BarcodeEvent) {
    BarcodeEvent[BarcodeEvent["invalid"] = 0] = "invalid";
})(BarcodeEvent || (BarcodeEvent = {}));
export var QuietZone;
(function (QuietZone) {
    QuietZone[QuietZone["All"] = 2] = "All";
})(QuietZone || (QuietZone = {}));
export var DataMatrixSize;
(function (DataMatrixSize) {
    DataMatrixSize[DataMatrixSize["Auto"] = 0] = "Auto";
    DataMatrixSize[DataMatrixSize["Size10x10"] = 1] = "Size10x10";
    DataMatrixSize[DataMatrixSize["Size12x12"] = 2] = "Size12x12";
    DataMatrixSize[DataMatrixSize["Size14x14"] = 3] = "Size14x14";
    DataMatrixSize[DataMatrixSize["Size16x16"] = 4] = "Size16x16";
    DataMatrixSize[DataMatrixSize["Size18x18"] = 5] = "Size18x18";
    DataMatrixSize[DataMatrixSize["Size20x20"] = 6] = "Size20x20";
    DataMatrixSize[DataMatrixSize["Size22x22"] = 7] = "Size22x22";
    DataMatrixSize[DataMatrixSize["Size24x24"] = 8] = "Size24x24";
    DataMatrixSize[DataMatrixSize["Size26x26"] = 9] = "Size26x26";
    DataMatrixSize[DataMatrixSize["Size32x32"] = 10] = "Size32x32";
    DataMatrixSize[DataMatrixSize["Size36x36"] = 11] = "Size36x36";
    DataMatrixSize[DataMatrixSize["Size40x40"] = 12] = "Size40x40";
    DataMatrixSize[DataMatrixSize["Size44x44"] = 13] = "Size44x44";
    DataMatrixSize[DataMatrixSize["Size48x48"] = 14] = "Size48x48";
    DataMatrixSize[DataMatrixSize["Size52x52"] = 15] = "Size52x52";
    DataMatrixSize[DataMatrixSize["Size64x64"] = 16] = "Size64x64";
    DataMatrixSize[DataMatrixSize["Size72x72"] = 17] = "Size72x72";
    DataMatrixSize[DataMatrixSize["Size80x80"] = 18] = "Size80x80";
    DataMatrixSize[DataMatrixSize["Size88x88"] = 19] = "Size88x88";
    DataMatrixSize[DataMatrixSize["Size96x96"] = 20] = "Size96x96";
    DataMatrixSize[DataMatrixSize["Size104x104"] = 21] = "Size104x104";
    DataMatrixSize[DataMatrixSize["Size120x120"] = 22] = "Size120x120";
    DataMatrixSize[DataMatrixSize["Size132x132"] = 23] = "Size132x132";
    DataMatrixSize[DataMatrixSize["Size144x144"] = 24] = "Size144x144";
    DataMatrixSize[DataMatrixSize["Size8x18"] = 25] = "Size8x18";
    DataMatrixSize[DataMatrixSize["Size8x32"] = 26] = "Size8x32";
    DataMatrixSize[DataMatrixSize["Size12x26"] = 27] = "Size12x26";
    DataMatrixSize[DataMatrixSize["Size12x36"] = 28] = "Size12x36";
    DataMatrixSize[DataMatrixSize["Size16x36"] = 29] = "Size16x36";
    DataMatrixSize[DataMatrixSize["Size16x48"] = 30] = "Size16x48";
})(DataMatrixSize || (DataMatrixSize = {}));
export var QRCodeVersion;
(function (QRCodeVersion) {
    QRCodeVersion[QRCodeVersion["Auto"] = 0] = "Auto";
    QRCodeVersion[QRCodeVersion["Version01"] = 1] = "Version01";
    QRCodeVersion[QRCodeVersion["Version02"] = 2] = "Version02";
    QRCodeVersion[QRCodeVersion["Version03"] = 3] = "Version03";
    QRCodeVersion[QRCodeVersion["Version04"] = 4] = "Version04";
    QRCodeVersion[QRCodeVersion["Version05"] = 5] = "Version05";
    QRCodeVersion[QRCodeVersion["Version06"] = 6] = "Version06";
    QRCodeVersion[QRCodeVersion["Version07"] = 7] = "Version07";
    QRCodeVersion[QRCodeVersion["Version08"] = 8] = "Version08";
    QRCodeVersion[QRCodeVersion["Version09"] = 9] = "Version09";
    QRCodeVersion[QRCodeVersion["Version10"] = 10] = "Version10";
    QRCodeVersion[QRCodeVersion["Version11"] = 11] = "Version11";
    QRCodeVersion[QRCodeVersion["Version12"] = 12] = "Version12";
    QRCodeVersion[QRCodeVersion["Version13"] = 13] = "Version13";
    QRCodeVersion[QRCodeVersion["Version14"] = 14] = "Version14";
    QRCodeVersion[QRCodeVersion["Version15"] = 15] = "Version15";
    QRCodeVersion[QRCodeVersion["Version16"] = 16] = "Version16";
    QRCodeVersion[QRCodeVersion["Version17"] = 17] = "Version17";
    QRCodeVersion[QRCodeVersion["Version18"] = 18] = "Version18";
    QRCodeVersion[QRCodeVersion["Version19"] = 19] = "Version19";
    QRCodeVersion[QRCodeVersion["Version20"] = 20] = "Version20";
    QRCodeVersion[QRCodeVersion["Version21"] = 21] = "Version21";
    QRCodeVersion[QRCodeVersion["Version22"] = 22] = "Version22";
    QRCodeVersion[QRCodeVersion["Version23"] = 23] = "Version23";
    QRCodeVersion[QRCodeVersion["Version24"] = 24] = "Version24";
    QRCodeVersion[QRCodeVersion["Version25"] = 25] = "Version25";
    QRCodeVersion[QRCodeVersion["Version26"] = 26] = "Version26";
    QRCodeVersion[QRCodeVersion["Version27"] = 27] = "Version27";
    QRCodeVersion[QRCodeVersion["Version28"] = 28] = "Version28";
    QRCodeVersion[QRCodeVersion["Version29"] = 29] = "Version29";
    QRCodeVersion[QRCodeVersion["Version30"] = 30] = "Version30";
    QRCodeVersion[QRCodeVersion["Version31"] = 31] = "Version31";
    QRCodeVersion[QRCodeVersion["Version32"] = 32] = "Version32";
    QRCodeVersion[QRCodeVersion["Version33"] = 33] = "Version33";
    QRCodeVersion[QRCodeVersion["Version34"] = 34] = "Version34";
    QRCodeVersion[QRCodeVersion["Version35"] = 35] = "Version35";
    QRCodeVersion[QRCodeVersion["Version36"] = 36] = "Version36";
    QRCodeVersion[QRCodeVersion["Version37"] = 37] = "Version37";
    QRCodeVersion[QRCodeVersion["Version38"] = 38] = "Version38";
    QRCodeVersion[QRCodeVersion["Version39"] = 39] = "Version39";
    QRCodeVersion[QRCodeVersion["Version40"] = 40] = "Version40";
})(QRCodeVersion || (QRCodeVersion = {}));
export var ErrorCorrectionLevel;
(function (ErrorCorrectionLevel) {
    ErrorCorrectionLevel[ErrorCorrectionLevel["Low"] = 7] = "Low";
    ErrorCorrectionLevel[ErrorCorrectionLevel["Medium"] = 15] = "Medium";
    ErrorCorrectionLevel[ErrorCorrectionLevel["Quartile"] = 25] = "Quartile";
    ErrorCorrectionLevel[ErrorCorrectionLevel["High"] = 30] = "High";
})(ErrorCorrectionLevel || (ErrorCorrectionLevel = {}));
