// @syncfusion/ej2/ej2-resource/scripts/ej2-spreadsheet/spreadsheet/common/event.js
export var ribbon = 'ribbon';
export var formulaBar = 'formulaBar';
export var sheetTabs = 'sheetTabs';
export var refreshSheetTabs = 'refreshSheetTabs';
export var isFormulaBarEdit = 'isFormulaBarEdit';
export var initialLoad = 'initialLoad';
export var contentLoaded = 'contentLoaded';
export var mouseDown = 'mouseDown';
export var spreadsheetDestroyed = 'spreadsheetDestroyed';
export var editOperation = 'editOperation';
export var formulaOperation = 'formulaOperation';
export var formulaBarOperation = 'formulaBarOperation';
export var click = 'click';
export var keyUp = 'keyUp';
export var keyDown = 'keyDown';
export var formulaKeyUp = 'formulaKeyUp';
export var formulaBarUpdate = 'formulaBarUpdate';
export var onVerticalScroll = 'verticalScroll';
export var onHorizontalScroll = 'horizontalScroll';
export var beforeContentLoaded = 'beforeContentLoaded';
export var beforeVirtualContentLoaded = 'beforeVirtualContentLoaded';
export var virtualContentLoaded = 'virtualContentLoaded';
export var contextMenuOpen = 'contextMenuOpen';
export var cellNavigate = 'cellNavigate';
export var mouseUpAfterSelection = 'mouseUpAfterSelection';
export var selectionComplete = 'selectionComplete';
export var cMenuBeforeOpen = 'contextmenuBeforeOpen';
export var insertSheetTab = 'insertSheetTab';
export var removeSheetTab = 'removeSheetTab';
export var renameSheetTab = 'renameSheetTab';
export var ribbonClick = 'ribboClick';
export var refreshRibbon = 'ribbonRefresh';
export var enableToolbarItems = 'enableToolbarItems';
export var tabSwitch = 'tabSwitch';
export var selectRange = 'selectRange';
export var cut = 'cut';
export var copy = 'copy';
export var paste = 'paste';
export var clearCopy = 'clearCopy';
export var dataBound = 'dataBound';
export var beforeDataBound = 'beforeDataBound';
export var addContextMenuItems = 'addContextMenuItems';
export var removeContextMenuItems = 'removeContextMenuItems';
export var enableContextMenuItems = 'enableContextMenuItems';
export var enableFileMenuItems = 'enableFileMenuItems';
export var hideFileMenuItems = 'hideFileMenuItems';
export var addFileMenuItems = 'addFileMenuItems';
export var hideRibbonTabs = 'hideRibbonTabs';
export var enableRibbonTabs = 'enableRibbonTabs';
export var addRibbonTabs = 'addRibbonTabs';
export var addToolbarItems = 'addToolbarItems';
export var hideToolbarItems = 'hideToolbarItems';
export var beforeRibbonCreate = 'beforeRibbonCreate';
export var rowHeightChanged = 'rowHeightChanged';
export var colWidthChanged = 'colWidthChanged';
export var onContentScroll = 'onContentScroll';
export var deInitProperties = 'deInitProperties';
export var activeSheetChanged = 'activeSheetChanged';
export var renameSheet = 'renameSheet';
export var initiateCustomSort = 'initiateCustomSort';
export var applySort = 'applySort';
export var collaborativeUpdate = 'collaborativeUpdate';
export var hideShow = 'hideShow';
export var autoFit = 'autoFitRowsColumns';
export var updateToggleItem = 'updateToggleItem';
export var initiateHyperlink = 'initiateHyperlink';
export var editHyperlink = 'editHyperlink';
export var openHyperlink = 'openHyperlink';
export var removeHyperlink = 'removeHyperlink';
export var createHyperlinkElement = 'createHyperlinkElement';
export var sheetNameUpdate = 'sheetNameUpdate';
export var hideSheet = 'hideSheet';
export var performUndoRedo = 'performUndoRedo';
export var updateUndoRedoCollection = 'updateUndoRedoCollection';
export var setActionData = 'setActionData';
export var getBeforeActionData = 'getBeforeActionData';
export var clearUndoRedoCollection = 'clearUndoRedoCollection';
export var initiateFilterUI = 'initiateFilterUI';
export var renderFilterCell = 'renderFilterCell';
export var reapplyFilter = 'reapplyFilter';
export var filterByCellValue = 'filterByCellValue';
export var clearFilter = 'clearFilter';
export var getFilteredColumn = 'getFilteredColumn';
export var completeAction = 'actionComplete';
export var beginAction = 'actionBegin';
export var filterCellKeyDown = 'filterCellKeyDown';
export var getFilterRange = 'getFilterRange';
export var setAutoFit = 'setAutoFit';
export var refreshFormulaDatasource = 'refreshFormulaDatasource';
export var setScrollEvent = 'setScrollEvent';
export var initiateDataValidation = 'initiatedatavalidation';
export var validationError = 'validationError';
export var startEdit = 'startEdit';
export var invalidData = 'invalidData';
export var clearInvalid = 'clearInvalid';
export var protectSheet = 'protectSheet';
export var applyProtect = 'applyProtect';
export var unprotectSheet = 'unprotectSheet';
export var protectCellFormat = 'protectCellFormat';
export var gotoDlg = 'renderGotoDlgt';
export var findDlg = 'renderFindDlg';
export var findHandler = 'findHandler';
export var replace = 'replaceHandler';
export var created = 'created';
export var editAlert = 'editAlert';
export var setUndoRedo = 'setUndoRedo';
export var enableFormulaInput = 'enableFormulaInput';
export var protectSelection = 'protectSelection';
export var hiddenMerge = 'hiddenMerge';
export var checkPrevMerge = 'checkPrevMerge';
export var checkMerge = 'checkMerge';
export var removeDataValidation = 'removeDataValidation';
export var showAggregate = 'showAggregate';
export var initiateConditionalFormat = 'initiateConditionalFormat';
export var checkConditionalFormat = 'checkConditionalFormat';
export var setCF = 'setCF';
export var clearViewer = 'clearViewer';
export var initiateFormulaReference = 'initiateFormulaReference';
export var initiateCur = 'initiateCur';
export var clearCellRef = 'clearCellRef';
export var editValue = 'editValue';
export var addressHandle = 'addressHandle';
export var initiateEdit = 'initiateEdit';
export var forRefSelRender = 'forRefSelRender';
export var blankWorkbook = 'blankWorkbook';
export var insertImage = 'insertImage';
export var refreshImgElem = 'refreshImgElem';
export var refreshImgCellObj = 'refreshImgCellObj';
export var getRowIdxFromClientY = 'getRowIdxFromClientY';
export var getColIdxFromClientX = 'getColIdxFromClientX';
export var createImageElement = 'createImageElement';
export var deleteImage = 'deleteImage';
export var deleteChart = 'deleteChart';
export var refreshChartCellObj = 'refreshChartCellObj';
export var refreshImagePosition = 'refreshImagePosition';
export var updateTableWidth = 'updateTableWidth';
export var focusBorder = 'focusBorder';
export var clearChartBorder = 'clearChartBorder';
export var insertChart = 'insertChart';
export var chartRangeSelection = 'chartRangeSelection';
export var insertDesignChart = 'insertDesignChart';
export var removeDesignChart = 'removeDesignChart';
export var chartDesignTab = 'chartDesignTab';
export var addChartEle = 'addChartEle';
export var isReact = 'isReact';
export var renderReactTemplates = 'renderReactTemplates';
export var clearTemplate = 'clearTemplate';
export var protectWorkbook = 'protectWorkbook';
export var unProtectWorkbook = 'unProtectWorkbook';
export var getPassWord = 'getPassWord';
export var importProtectWorkbook = 'importProtectWorkbook';
export var selectionStatus = 'selectionStatus';
export var freeze = 'freeze';
export var overlayEleSize = 'overlayEleSize';
