// @syncfusion/ej2/ej2-resource/scripts/ej2-barcode-generator/barcode/rendering/renderer.js
import { BarcodeCanvasRenderer } from './canvas-renderer';
import { BarcodeSVGRenderering } from './svg-renderer';
var BarcodeRenderer = (function () {
    function BarcodeRenderer(name, isSvgMode) {
        this.renderer = null;
        this.isSvgMode = null;
        this.isSvgMode = isSvgMode;
        this.renderer = isSvgMode ? new BarcodeSVGRenderering() : new BarcodeCanvasRenderer();
    }
    BarcodeRenderer.prototype.renderRootElement = function (attribute, backGroundColor, width, height) {
        var canvasObj = this.renderer.renderRootElement(attribute, backGroundColor, width, height);
        return canvasObj;
    };
    BarcodeRenderer.prototype.renderRectElement = function (canvas, attribute) {
        var canvasObj = this.renderer.renderRect(canvas, attribute);
        return canvasObj;
    };
    BarcodeRenderer.prototype.renderTextElement = function (canvas, attribute) {
        var canvasObj = this.renderer.renderText(canvas, attribute);
        return canvasObj;
    };
    return BarcodeRenderer;
}());
export { BarcodeRenderer };
