// @syncfusion/ej2/ej2-resource/scripts/ej2-diagrams/diagram/core/elements/html-element.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { DiagramElement } from './diagram-element';
import { getContent } from '../../utility/dom-util';
import { templateCompiler } from '../../utility/base-util';
var DiagramHtmlElement = (function (_super) {
    __extends(DiagramHtmlElement, _super);
    function DiagramHtmlElement(nodeId, diagramId, annotationId, nodeTemplate) {
        var _this = _super.call(this) || this;
        _this.data = '';
        _this.nodeId = '';
        _this.annotationId = '';
        _this.diagramId = '';
        _this.diagramId = diagramId;
        _this.nodeId = nodeId;
        _this.annotationId = annotationId;
        _this.templateFn = templateCompiler(nodeTemplate);
        return _this;
    }
    DiagramHtmlElement.prototype.getNodeTemplate = function () {
        return this.templateFn;
    };
    Object.defineProperty(DiagramHtmlElement.prototype, "content", {
        get: function () {
            return this.data;
        },
        set: function (value) {
            this.data = value;
            if (!this.canReset) {
                this.canReset = true;
                if (!this.isTemplate) {
                    this.template = getContent(this, true);
                }
                this.canReset = false;
                this.isDirt = true;
            }
        },
        enumerable: true,
        configurable: true
    });
    return DiagramHtmlElement;
}(DiagramElement));
export { DiagramHtmlElement };
