// @syncfusion/ej2/ej2-resource/scripts/ej2-spreadsheet/spreadsheet/integrations/number-format.js
import { refreshCellElement } from '../../workbook/common/event';
import { isNullOrUndefined } from '@syncfusion/ej2-base';
var NumberFormat = (function () {
    function NumberFormat(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    NumberFormat.prototype.refreshCellElement = function (args) {
        var cell = this.parent.getCell(args.rowIndex, args.colIndex);
        if (!isNullOrUndefined(cell)) {
            this.parent.refreshNode(cell, args);
        }
    };
    NumberFormat.prototype.addEventListener = function () {
        this.parent.on(refreshCellElement, this.refreshCellElement, this);
    };
    NumberFormat.prototype.removeEventListener = function () {
        if (!this.parent.isDestroyed) {
            this.parent.off(refreshCellElement, this.refreshCellElement);
        }
    };
    NumberFormat.prototype.destroy = function () {
        this.removeEventListener();
        this.parent = null;
    };
    NumberFormat.prototype.getModuleName = function () {
        return 'numberFormat';
    };
    return NumberFormat;
}());
export { NumberFormat };
