// @syncfusion/ej2/ej2-resource/scripts/ej2-spreadsheet/spreadsheet/integrations/save.js
import { dialog, focus } from '../index';
import { beginSave, saveCompleted, saveError } from '../../workbook/common/event';
var Save = (function () {
    function Save(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    Save.prototype.destroy = function () {
        this.removeEventListener();
        this.parent = null;
    };
    Save.prototype.addEventListener = function () {
        this.parent.on(beginSave, this.initiateSave, this);
        this.parent.on(saveCompleted, this.saveCompleted, this);
        this.parent.on(saveError, this.showErrorDialog, this);
    };
    Save.prototype.removeEventListener = function () {
        if (!this.parent.isDestroyed) {
            this.parent.off(beginSave, this.initiateSave);
            this.parent.off(saveCompleted, this.saveCompleted);
            this.parent.off(saveError, this.showErrorDialog);
        }
    };
    Save.prototype.getModuleName = function () {
        return 'save';
    };
    Save.prototype.initiateSave = function () {
        this.parent.showSpinner();
    };
    Save.prototype.saveCompleted = function () {
        this.parent.hideSpinner();
    };
    Save.prototype.showErrorDialog = function (args) {
        var _this = this;
        var dialogInst = this.parent.serviceLocator.getService(dialog);
        dialogInst.show({
            target: this.parent.element, isModal: true, showCloseIcon: true, height: 180, width: 400, content: args.content,
            beforeOpen: function () { return focus(_this.parent.element); }
        });
    };
    return Save;
}());
export { Save };
