// @syncfusion/ej2/ej2-resource/scripts/ej2-documenteditor/blazor/document-editor/implementation/track-changes/track-changes-pane.js
import { createElement, isNullOrUndefined, select } from '@syncfusion/ej2-base';
import { TextElementBox, ImageElementBox, FieldElementBox, TextFormField, DropDownFormField, ParagraphWidget } from '../viewer/page';
import { WRowFormat, WCharacterFormat } from '../index';
import { HelperMethods } from '../editor/editor-helper';
import { Dictionary } from '../base/index';
var TrackChangesPane = (function () {
    function TrackChangesPane(owner) {
        this.users = [];
        this.enableButtons = true;
        this.sortedRevisions = [];
        this.noChangesVisibleInternal = true;
        this.isTrackingPageBreak = false;
        this.owner = owner;
        this.selectedUser = 'All';
        this.selectedType = 'All';
    }
    Object.defineProperty(TrackChangesPane.prototype, "setNoChangesVisibility", {
        get: function () {
            return this.noChangesVisibleInternal;
        },
        set: function (visible) {
            this.noChangeDivElement.style.display = visible ? 'block' : 'none';
            this.noChangesVisibleInternal = visible;
            this.enableDisableToolbarItem(!visible);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TrackChangesPane.prototype, "currentSelectedRevision", {
        get: function () {
            return this.currentSelectedRevisionInternal;
        },
        set: function (value) {
            if (!isNullOrUndefined(this.changesInfoDiv)) {
                var selectedElement = select('.e-de-trckchanges-inner-select', this.changesInfoDiv);
                if (isNullOrUndefined(value)) {
                    if (!isNullOrUndefined(selectedElement)) {
                        selectedElement.classList.remove('e-de-trckchanges-inner-select');
                    }
                }
                else if (value !== this.currentSelectedRevisionInternal || isNullOrUndefined(selectedElement)) {
                    if (selectedElement) {
                        selectedElement.classList.remove('e-de-trckchanges-inner-select');
                    }
                    if (this.changes.length > 0 && this.changes.containsKey(value)) {
                        var currentChangeView = this.changes.get(value);
                        currentChangeView.singleInnerDiv.classList.add('e-de-trckchanges-inner-select');
                    }
                }
                this.currentSelectedRevisionInternal = value;
                selectedElement = select('.e-de-trckchanges-inner-select', this.changesInfoDiv);
                if (!isNullOrUndefined(selectedElement)) {
                    selectedElement.parentElement.scrollIntoView({ block: 'nearest' });
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    TrackChangesPane.prototype.initTrackChangePane = function () {
        this.changes = new Dictionary();
        this.changesInfoDiv = this.owner.element.getElementsByClassName('e-de-tc-pane-revision')[0];
        if (!this.changesInfoDiv) {
            this.changesInfoDiv = createElement('div', { id: 'e-de-tc-pane-revision', styles: 'overflow:auto' });
        }
        this.noChangeDivElement = createElement('div', { styles: 'display:none;', className: 'e-de-tc-no-chng' });
        this.noChangeDivElement.textContent = 'No changes';
        this.changesInfoDiv.appendChild(this.noChangeDivElement);
        this.updateTrackChanges();
    };
    TrackChangesPane.prototype.enableDisableToolbarItem = function (enable) {
    };
    TrackChangesPane.prototype.getSpanView = function (value, type) {
        return (type === 0 ? '<span class="e-de-track-span-user">' : '<span class="e-de-track-span-view">') + value + '</span>';
    };
    TrackChangesPane.prototype.onSelection = function (revision) {
        this.currentSelectedRevision = revision;
    };
    TrackChangesPane.prototype.onUserSelect = function (selectedText) {
        this.selectedUser = selectedText;
        this.sortCollectionToDisplay();
    };
    TrackChangesPane.prototype.onMenuSelect = function (selectedText) {
        var _this = this;
        if (selectedText === 'Accept all') {
            setTimeout(function () {
                _this.owner.revisionsInternal.handleRevisionCollection(true, _this.sortedRevisions);
            });
        }
        else if (selectedText === 'Reject all') {
            setTimeout(function () {
                _this.owner.revisionsInternal.handleRevisionCollection(false, _this.sortedRevisions);
            });
        }
        this.updateUsers();
    };
    TrackChangesPane.prototype.onTypeSelect = function (selectedText) {
        this.selectedType = selectedText;
        this.sortCollectionToDisplay();
    };
    TrackChangesPane.prototype.updateMenuOptions = function () {
        var revisionType;
        if (this.selectedType !== 'All') {
            revisionType = this.selectedType === 'Inserted' ? 'Insertion'
                : 'Deletion';
        }
    };
    TrackChangesPane.prototype.sortCollectionToDisplay = function () {
        var isRevisionVisible = false;
        this.sortedRevisions = [];
        this.updateMenuOptions();
        for (var i = 0; i < this.changes.length; i++) {
            var changes = this.changes.get(this.revisions[i]);
            var singleChangesDiv = changes.outerSingleDiv;
            if (this.selectedUser === 'All' && this.selectedType === 'All') {
                singleChangesDiv.style.display = 'block';
                isRevisionVisible = true;
            }
            else if (this.selectedUser === 'All' && this.selectedType !== 'All') {
                if (changes.revisionType === this.selectedType) {
                    singleChangesDiv.style.display = 'block';
                    isRevisionVisible = true;
                }
                else {
                    singleChangesDiv.style.display = 'none';
                }
            }
            else if (this.selectedUser !== 'All' && this.selectedType === 'All') {
                if (changes.user === this.selectedUser) {
                    singleChangesDiv.style.display = 'block';
                    isRevisionVisible = true;
                }
                else {
                    singleChangesDiv.style.display = 'none';
                }
            }
            else {
                if (changes.user === this.selectedUser && changes.revisionType === this.selectedType) {
                    singleChangesDiv.style.display = 'block';
                    isRevisionVisible = true;
                }
                else {
                    singleChangesDiv.style.display = 'none';
                }
            }
            if (singleChangesDiv.style.display === 'block') {
                this.sortedRevisions.push(this.revisions[i]);
            }
        }
        this.setNoChangesVisibility = !isRevisionVisible;
    };
    TrackChangesPane.prototype.enableDisableButton = function (enableButton) {
        this.enableButtons = enableButton;
        this.updateTrackChanges();
    };
    TrackChangesPane.prototype.updateTrackChanges = function (show) {
        if (show || isNullOrUndefined(show)) {
            if (isNullOrUndefined(this.changesInfoDiv)) {
                return;
            }
            this.removeAllChanges();
            for (var i = 0; i < this.owner.revisions.changes.length; i++) {
                var revision = this.owner.revisions.changes[i];
                this.addChanges(revision);
            }
            this.sortCollectionToDisplay();
            this.updateUsers();
            this.currentSelectedRevision = this.currentSelectedRevisionInternal;
            this.updateHeight();
            this.owner.resize();
        }
        else {
            this.currentSelectedRevision = undefined;
        }
    };
    TrackChangesPane.prototype.updateUsers = function () {
        if (this.users.length > 0) {
            this.users = [];
        }
        for (var i = 0; i < this.revisions.length; i++) {
            if (this.users.indexOf(this.revisions[i].author) === -1) {
                this.users.push(this.revisions[i].author);
            }
        }
        this.owner.dotNetRef.invokeMethodAsync("UpdateRevisionUsers", this.users);
    };
    TrackChangesPane.prototype.updateHeight = function () {
        var tabElement = this.owner.element.getElementsByClassName('e-de-review-pane')[0];
        if (tabElement) {
            var tabHeaderHeight = tabElement.getElementsByClassName('e-tab-header')[0].getBoundingClientRect().height;
            var toolbarElement = tabElement.getElementsByClassName('e-de-track-toolbar')[0];
            this.changesInfoDiv.style.height = tabElement.clientHeight - toolbarElement.clientHeight - tabHeaderHeight - 2 + 'px';
        }
    };
    TrackChangesPane.prototype.removeAllChanges = function () {
        if (isNullOrUndefined(this.changesInfoDiv)) {
            return;
        }
        while (this.changesInfoDiv.childNodes.length > 1) {
            this.changesInfoDiv.removeChild(this.changesInfoDiv.lastChild);
        }
        this.revisions = [];
        this.changes.clear();
    };
    TrackChangesPane.prototype.clear = function () {
        this.removeAllChanges();
        this.selectedUser = 'All';
        this.selectedType = 'All';
        this.currentSelectedRevision = undefined;
    };
    TrackChangesPane.prototype.destroy = function () {
        this.removeAllChanges();
        if (this.users.length > 0) {
            this.users = [];
        }
    };
    TrackChangesPane.prototype.addChanges = function (revision) {
        if (!isNullOrUndefined(this.changesInfoDiv)) {
            var currentChangeView = new ChangesSingleView(this.owner, this);
            this.changesInfoDiv.appendChild(currentChangeView.createSingleChangesDiv(revision));
            if (!this.enableButtons) {
                if (!(currentChangeView.acceptButtonElement.classList.contains('e-de-overlay'))) {
                    currentChangeView.acceptButtonElement.classList.add('e-de-overlay');
                    currentChangeView.rejectButtonElement.classList.add('e-de-overlay');
                }
            }
            else if (currentChangeView.acceptButtonElement.classList.contains('e-de-overlay')) {
                currentChangeView.acceptButtonElement.classList.remove('e-de-overlay');
                currentChangeView.rejectButtonElement.classList.remove('e-de-overlay');
            }
            this.revisions.push(revision);
            this.changes.add(revision, currentChangeView);
        }
    };
    TrackChangesPane.prototype.navigatePreviousChanges = function () {
        this.owner.selection.navigatePreviousRevision();
    };
    TrackChangesPane.prototype.navigateNextChanges = function () {
        this.owner.selection.navigateNextRevision();
    };
    return TrackChangesPane;
}());
export { TrackChangesPane };
var ChangesSingleView = (function () {
    function ChangesSingleView(owner, trackChangesPane) {
        this.owner = owner;
        this.trackChangesPane = trackChangesPane;
    }
    ChangesSingleView.prototype.createSingleChangesDiv = function (revision) {
        this.revision = revision;
        this.user = revision.author;
        this.outerSingleDiv = createElement('div', { className: 'e-de-tc-outer' });
        this.singleInnerDiv = createElement('div', { className: 'e-de-trckchanges-inner' });
        this.singleInnerDiv.addEventListener('click', this.selectRevision.bind(this));
        this.outerSingleDiv.appendChild(this.singleInnerDiv);
        var userNameTotalDiv = createElement('div', { className: 'e-de-track-usernme-div' });
        var userNameLabel = createElement('div', { innerHTML: revision.author, className: 'e-de-track-user-nme' });
        if (!isNullOrUndefined(revision.author)) {
            userNameTotalDiv.style.display = 'flex';
            this.owner.documentHelper.getAvatar(userNameTotalDiv, userNameLabel, undefined, revision);
        }
        var revisionTypeLabel = createElement('div');
        if (revision.revisionType === 'Insertion') {
            this.revisionType = 'Inserted';
            revisionTypeLabel.innerHTML = 'Inserted'.toUpperCase();
            revisionTypeLabel.classList.add('e-de-track-insert');
        }
        else if (revision.revisionType === 'Deletion') {
            this.revisionType = 'Deleted';
            revisionTypeLabel.innerHTML = 'Deleted'.toUpperCase();
            revisionTypeLabel.classList.add('e-de-track-delete');
        }
        else if (revision.revisionType === 'MoveFrom') {
            this.revisionType = 'MoveFrom';
            revisionTypeLabel.innerHTML = 'Move From'.toUpperCase();
            revisionTypeLabel.classList.add('e-de-track-delete');
            revisionTypeLabel.style.whiteSpace = 'nowrap';
        }
        else if (revision.revisionType === 'MoveTo') {
            this.revisionType = 'MoveTo';
            revisionTypeLabel.innerHTML = 'Move To'.toUpperCase();
            revisionTypeLabel.classList.add('e-de-track-insert');
            revisionTypeLabel.style.whiteSpace = 'nowrap';
        }
        userNameTotalDiv.appendChild(revisionTypeLabel);
        this.singleInnerDiv.appendChild(userNameTotalDiv);
        var dateView = createElement('div', {
            className: 'e-de-track-date',
            innerHTML: HelperMethods.getModifiedDate(revision.date)
        });
        this.singleInnerDiv.appendChild(dateView);
        var changesTextDiv = createElement('div', {
            className: 'e-de-track-chngs-text'
        });
        this.layoutElementText(revision.range, changesTextDiv);
        this.singleInnerDiv.appendChild(changesTextDiv);
        var buttonTotalDiv = createElement('div', {
            styles: 'display:inline-block;width:100%'
        });
        this.singleInnerDiv.appendChild(buttonTotalDiv);
        var buttonDiv = createElement('div', {
            styles: 'float:left;'
        });
        this.acceptButtonElement = createElement('button', { className: 'e-de-track-accept-button e-btn e-lib e-control e-outline' });
        this.acceptButtonElement.textContent = 'Accept';
        buttonDiv.appendChild(this.acceptButtonElement);
        buttonTotalDiv.appendChild(buttonDiv);
        this.acceptButtonElement.addEventListener('click', this.acceptButtonClick.bind(this));
        buttonDiv = createElement('div', {
            styles: 'float:left;'
        });
        this.rejectButtonElement = createElement('button', { className: 'e-de-track-reject-button e-btn e-lib e-control e-outline' });
        this.rejectButtonElement.textContent = 'Reject';
        buttonDiv.appendChild(this.rejectButtonElement);
        buttonTotalDiv.appendChild(buttonDiv);
        this.rejectButtonElement.addEventListener('click', this.rejectButtonClick.bind(this));
        var changesCount = createElement('div', { className: 'e-de-track-chngs-count', styles: 'float:right;' });
        var currentCount = this.owner.revisions.changes.indexOf(revision) + 1;
        var totalCount = this.owner.revisions.changes.length;
        changesCount.innerHTML = 'Changes' + ' ' + currentCount.toString() +
            ' ' + 'of' + ' ' + totalCount.toString();
        buttonTotalDiv.appendChild(changesCount);
        return this.outerSingleDiv;
    };
    ChangesSingleView.prototype.selectRevision = function () {
        this.owner.selection.selectRevision(this.revision);
        this.trackChangesPane.onSelection(this.revision);
    };
    ChangesSingleView.prototype.layoutElementText = function (range, changesText) {
        changesText.style.width = '100%';
        var text = '';
        var toSkip = false;
        for (var i = 0; i < range.length; i++) {
            var element = range[i];
            if (element instanceof FieldElementBox && element.fieldType === 1) {
                toSkip = false;
                continue;
            }
            if (toSkip) {
                continue;
            }
            if (element instanceof FieldElementBox && element.fieldType === 0) {
                toSkip = true;
            }
            if (element instanceof TextElementBox) {
                text += element.text;
            }
            else if (element instanceof FieldElementBox && element.fieldType === 0) {
                var fieldCode = this.owner.selection.getFieldCode(element);
                if (fieldCode.match('TOC ') || fieldCode.match('Toc')) {
                    text += '<Table of Content>';
                    changesText.appendChild(this.addSpan(text));
                    return;
                }
                else if (fieldCode.match('HYPERLINK ') || fieldCode.match('MERGEFIELD') || fieldCode.match('FORMTEXT') || fieldCode.match('PAGE ')) {
                    text += this.owner.editor.retrieveFieldResultantText(element.fieldEnd);
                }
                else if (element.formFieldData) {
                    var emptyChar = this.owner.documentHelper.textHelper.repeatChar(this.owner.documentHelper.textHelper.getEnSpaceCharacter(), 5);
                    if (text !== '') {
                        changesText.appendChild(this.addSpan(text));
                        text = '';
                    }
                    if (element.formFieldData instanceof TextFormField) {
                        changesText.appendChild(this.addSpan(element.formFieldData.defaultValue === '' ? emptyChar : element.formFieldData.defaultValue, 'e-de-tc-field'));
                    }
                    else if (element.formFieldData instanceof DropDownFormField) {
                        changesText.appendChild(this.addSpan(element.formFieldData.dropdownItems.length > 0 ? element.formFieldData.dropdownItems[0] : emptyChar, 'e-de-tc-field'));
                    }
                    else {
                        changesText.appendChild(this.addSpan(element.formFieldData.checked ? String.fromCharCode(9745) : String.fromCharCode(9744), 'e-de-tc-field'));
                    }
                }
            }
            else if (element instanceof ImageElementBox) {
                if (text !== '') {
                    changesText.appendChild(this.addSpan(text));
                    text = '';
                }
                var imageEle = createElement('img');
                imageEle.setAttribute('src', element.imageString);
                imageEle.classList.add('e-de-tc-shrink-img');
                changesText.appendChild(imageEle);
            }
            else if (element instanceof WRowFormat) {
                var tableEle = createElement('table');
                tableEle.classList.add('e-de-track-chng-table');
                tableEle.insertRow();
                for (var i_1 = 0; i_1 < element.ownerBase.childWidgets.length; i_1++) {
                    tableEle.rows[0].insertCell();
                    tableEle.rows[0].cells[i_1].classList.add('e-de-tc-tble-cell');
                }
                changesText.appendChild(tableEle);
                return;
            }
            else if (element instanceof WCharacterFormat) {
                if (text !== '') {
                    changesText.appendChild(this.addSpan(text));
                    text = '';
                }
                var paraMark = '¶';
                if (element.ownerBase instanceof ParagraphWidget && element.ownerBase.isEndsWithPageBreak) {
                    paraMark = '............Page Break............' + paraMark;
                }
                changesText.appendChild(this.addSpan(paraMark, 'e-de-tc-pmark'));
                changesText.appendChild(createElement('br'));
            }
        }
        changesText.appendChild(this.addSpan(text));
    };
    ChangesSingleView.prototype.addSpan = function (text, cssClass) {
        var span = createElement('span');
        span.textContent = text;
        if (cssClass) {
            span.classList.add(cssClass);
        }
        return span;
    };
    ChangesSingleView.prototype.acceptButtonClick = function () {
        this.trackChangesPane.changesInfoDiv.removeChild(this.outerSingleDiv);
        this.removeFromParentCollec();
        this.revision.accept();
    };
    ChangesSingleView.prototype.rejectButtonClick = function () {
        this.trackChangesPane.changesInfoDiv.removeChild(this.outerSingleDiv);
        this.removeFromParentCollec();
        this.revision.reject();
    };
    ChangesSingleView.prototype.removeFromParentCollec = function () {
        this.trackChangesPane.changes.remove(this.revision);
        this.trackChangesPane.revisions.splice(this.trackChangesPane.revisions.indexOf(this.revision), 1);
        if (this.trackChangesPane.changes.length === 0) {
            this.trackChangesPane.setNoChangesVisibility = true;
        }
        this.trackChangesPane.updateUsers();
    };
    return ChangesSingleView;
}());
export { ChangesSingleView };
