// @syncfusion/ej2/ej2-resource/scripts/ej2-diagrams/diagram/enum/enum.js
export var BlazorAction;
(function (BlazorAction) {
    BlazorAction[BlazorAction["Default"] = 0] = "Default";
    BlazorAction[BlazorAction["expandNode"] = 2] = "expandNode";
    BlazorAction[BlazorAction["interaction"] = 4] = "interaction";
    BlazorAction[BlazorAction["GroupingInProgress"] = 8] = "GroupingInProgress";
    BlazorAction[BlazorAction["GroupClipboardInProcess"] = 16] = "GroupClipboardInProcess";
    BlazorAction[BlazorAction["ClearObject"] = 32] = "ClearObject";
})(BlazorAction || (BlazorAction = {}));
export var PortVisibility;
(function (PortVisibility) {
    PortVisibility[PortVisibility["Visible"] = 1] = "Visible";
    PortVisibility[PortVisibility["Hidden"] = 2] = "Hidden";
    PortVisibility[PortVisibility["Hover"] = 4] = "Hover";
    PortVisibility[PortVisibility["Connect"] = 8] = "Connect";
})(PortVisibility || (PortVisibility = {}));
export var SnapConstraints;
(function (SnapConstraints) {
    SnapConstraints[SnapConstraints["None"] = 0] = "None";
    SnapConstraints[SnapConstraints["ShowHorizontalLines"] = 1] = "ShowHorizontalLines";
    SnapConstraints[SnapConstraints["ShowVerticalLines"] = 2] = "ShowVerticalLines";
    SnapConstraints[SnapConstraints["ShowLines"] = 3] = "ShowLines";
    SnapConstraints[SnapConstraints["SnapToHorizontalLines"] = 4] = "SnapToHorizontalLines";
    SnapConstraints[SnapConstraints["SnapToVerticalLines"] = 8] = "SnapToVerticalLines";
    SnapConstraints[SnapConstraints["SnapToLines"] = 12] = "SnapToLines";
    SnapConstraints[SnapConstraints["SnapToObject"] = 16] = "SnapToObject";
    SnapConstraints[SnapConstraints["All"] = 31] = "All";
})(SnapConstraints || (SnapConstraints = {}));
export var SelectorConstraints;
(function (SelectorConstraints) {
    SelectorConstraints[SelectorConstraints["None"] = 1] = "None";
    SelectorConstraints[SelectorConstraints["ConnectorSourceThumb"] = 2] = "ConnectorSourceThumb";
    SelectorConstraints[SelectorConstraints["ConnectorTargetThumb"] = 4] = "ConnectorTargetThumb";
    SelectorConstraints[SelectorConstraints["ResizeSouthEast"] = 8] = "ResizeSouthEast";
    SelectorConstraints[SelectorConstraints["ResizeSouthWest"] = 16] = "ResizeSouthWest";
    SelectorConstraints[SelectorConstraints["ResizeNorthEast"] = 32] = "ResizeNorthEast";
    SelectorConstraints[SelectorConstraints["ResizeNorthWest"] = 64] = "ResizeNorthWest";
    SelectorConstraints[SelectorConstraints["ResizeEast"] = 128] = "ResizeEast";
    SelectorConstraints[SelectorConstraints["ResizeWest"] = 256] = "ResizeWest";
    SelectorConstraints[SelectorConstraints["ResizeSouth"] = 512] = "ResizeSouth";
    SelectorConstraints[SelectorConstraints["ResizeNorth"] = 1024] = "ResizeNorth";
    SelectorConstraints[SelectorConstraints["Rotate"] = 2048] = "Rotate";
    SelectorConstraints[SelectorConstraints["UserHandle"] = 4096] = "UserHandle";
    SelectorConstraints[SelectorConstraints["ToolTip"] = 8192] = "ToolTip";
    SelectorConstraints[SelectorConstraints["ResizeAll"] = 2046] = "ResizeAll";
    SelectorConstraints[SelectorConstraints["All"] = 16382] = "All";
})(SelectorConstraints || (SelectorConstraints = {}));
export var ConnectionPointOrigin;
(function (ConnectionPointOrigin) {
    ConnectionPointOrigin["SamePoint"] = "SamePoint";
    ConnectionPointOrigin["DifferentPoint"] = "DifferentPoint";
})(ConnectionPointOrigin || (ConnectionPointOrigin = {}));
export var ChildArrangement;
(function (ChildArrangement) {
    ChildArrangement["Linear"] = "Linear";
    ChildArrangement["Nonlinear"] = "Nonlinear";
})(ChildArrangement || (ChildArrangement = {}));
export var ConnectorConstraints;
(function (ConnectorConstraints) {
    ConnectorConstraints[ConnectorConstraints["None"] = 1] = "None";
    ConnectorConstraints[ConnectorConstraints["Select"] = 2] = "Select";
    ConnectorConstraints[ConnectorConstraints["Delete"] = 4] = "Delete";
    ConnectorConstraints[ConnectorConstraints["Drag"] = 8] = "Drag";
    ConnectorConstraints[ConnectorConstraints["DragSourceEnd"] = 16] = "DragSourceEnd";
    ConnectorConstraints[ConnectorConstraints["DragTargetEnd"] = 32] = "DragTargetEnd";
    ConnectorConstraints[ConnectorConstraints["DragSegmentThumb"] = 64] = "DragSegmentThumb";
    ConnectorConstraints[ConnectorConstraints["AllowDrop"] = 128] = "AllowDrop";
    ConnectorConstraints[ConnectorConstraints["Bridging"] = 256] = "Bridging";
    ConnectorConstraints[ConnectorConstraints["BridgeObstacle"] = 512] = "BridgeObstacle";
    ConnectorConstraints[ConnectorConstraints["InheritBridging"] = 1024] = "InheritBridging";
    ConnectorConstraints[ConnectorConstraints["PointerEvents"] = 2048] = "PointerEvents";
    ConnectorConstraints[ConnectorConstraints["Tooltip"] = 4096] = "Tooltip";
    ConnectorConstraints[ConnectorConstraints["InheritTooltip"] = 8192] = "InheritTooltip";
    ConnectorConstraints[ConnectorConstraints["Interaction"] = 4218] = "Interaction";
    ConnectorConstraints[ConnectorConstraints["ReadOnly"] = 16384] = "ReadOnly";
    ConnectorConstraints[ConnectorConstraints["LineRouting"] = 32768] = "LineRouting";
    ConnectorConstraints[ConnectorConstraints["InheritLineRouting"] = 65536] = "InheritLineRouting";
    ConnectorConstraints[ConnectorConstraints["ConnectToNearByNode"] = 131072] = "ConnectToNearByNode";
    ConnectorConstraints[ConnectorConstraints["ConnectToNearByPort"] = 262144] = "ConnectToNearByPort";
    ConnectorConstraints[ConnectorConstraints["ConnectToNearByElement"] = 393216] = "ConnectToNearByElement";
    ConnectorConstraints[ConnectorConstraints["Default"] = 470590] = "Default";
})(ConnectorConstraints || (ConnectorConstraints = {}));
export var AnnotationConstraints;
(function (AnnotationConstraints) {
    AnnotationConstraints[AnnotationConstraints["ReadOnly"] = 2] = "ReadOnly";
    AnnotationConstraints[AnnotationConstraints["InheritReadOnly"] = 4] = "InheritReadOnly";
    AnnotationConstraints[AnnotationConstraints["Select"] = 8] = "Select";
    AnnotationConstraints[AnnotationConstraints["Drag"] = 16] = "Drag";
    AnnotationConstraints[AnnotationConstraints["Resize"] = 32] = "Resize";
    AnnotationConstraints[AnnotationConstraints["Rotate"] = 64] = "Rotate";
    AnnotationConstraints[AnnotationConstraints["Interaction"] = 120] = "Interaction";
    AnnotationConstraints[AnnotationConstraints["None"] = 0] = "None";
})(AnnotationConstraints || (AnnotationConstraints = {}));
export var NodeConstraints;
(function (NodeConstraints) {
    NodeConstraints[NodeConstraints["None"] = 0] = "None";
    NodeConstraints[NodeConstraints["Select"] = 2] = "Select";
    NodeConstraints[NodeConstraints["Drag"] = 4] = "Drag";
    NodeConstraints[NodeConstraints["Rotate"] = 8] = "Rotate";
    NodeConstraints[NodeConstraints["Shadow"] = 16] = "Shadow";
    NodeConstraints[NodeConstraints["PointerEvents"] = 32] = "PointerEvents";
    NodeConstraints[NodeConstraints["Delete"] = 64] = "Delete";
    NodeConstraints[NodeConstraints["InConnect"] = 128] = "InConnect";
    NodeConstraints[NodeConstraints["OutConnect"] = 256] = "OutConnect";
    NodeConstraints[NodeConstraints["Individual"] = 512] = "Individual";
    NodeConstraints[NodeConstraints["Expandable"] = 1024] = "Expandable";
    NodeConstraints[NodeConstraints["AllowDrop"] = 2048] = "AllowDrop";
    NodeConstraints[NodeConstraints["Inherit"] = 78] = "Inherit";
    NodeConstraints[NodeConstraints["ResizeNorthEast"] = 4096] = "ResizeNorthEast";
    NodeConstraints[NodeConstraints["ResizeEast"] = 8192] = "ResizeEast";
    NodeConstraints[NodeConstraints["ResizeSouthEast"] = 16384] = "ResizeSouthEast";
    NodeConstraints[NodeConstraints["ResizeSouth"] = 32768] = "ResizeSouth";
    NodeConstraints[NodeConstraints["ResizeSouthWest"] = 65536] = "ResizeSouthWest";
    NodeConstraints[NodeConstraints["ResizeWest"] = 131072] = "ResizeWest";
    NodeConstraints[NodeConstraints["ResizeNorthWest"] = 262144] = "ResizeNorthWest";
    NodeConstraints[NodeConstraints["ResizeNorth"] = 524288] = "ResizeNorth";
    NodeConstraints[NodeConstraints["Resize"] = 1044480] = "Resize";
    NodeConstraints[NodeConstraints["AspectRatio"] = 1048576] = "AspectRatio";
    NodeConstraints[NodeConstraints["Tooltip"] = 2097152] = "Tooltip";
    NodeConstraints[NodeConstraints["InheritTooltip"] = 4194304] = "InheritTooltip";
    NodeConstraints[NodeConstraints["ReadOnly"] = 8388608] = "ReadOnly";
    NodeConstraints[NodeConstraints["HideThumbs"] = 16777216] = "HideThumbs";
    NodeConstraints[NodeConstraints["AllowMovingOutsideLane"] = 33554432] = "AllowMovingOutsideLane";
    NodeConstraints[NodeConstraints["Default"] = 5240814] = "Default";
})(NodeConstraints || (NodeConstraints = {}));
export var ElementAction;
(function (ElementAction) {
    ElementAction[ElementAction["None"] = 0] = "None";
    ElementAction[ElementAction["ElementIsPort"] = 2] = "ElementIsPort";
    ElementAction[ElementAction["ElementIsGroup"] = 4] = "ElementIsGroup";
    ElementAction[ElementAction["HorizontalLaneHeader"] = 8] = "HorizontalLaneHeader";
})(ElementAction || (ElementAction = {}));
export var ThumbsConstraints;
(function (ThumbsConstraints) {
    ThumbsConstraints[ThumbsConstraints["Rotate"] = 2] = "Rotate";
    ThumbsConstraints[ThumbsConstraints["ConnectorSource"] = 4] = "ConnectorSource";
    ThumbsConstraints[ThumbsConstraints["ConnectorTarget"] = 8] = "ConnectorTarget";
    ThumbsConstraints[ThumbsConstraints["ResizeNorthEast"] = 16] = "ResizeNorthEast";
    ThumbsConstraints[ThumbsConstraints["ResizeEast"] = 32] = "ResizeEast";
    ThumbsConstraints[ThumbsConstraints["ResizeSouthEast"] = 64] = "ResizeSouthEast";
    ThumbsConstraints[ThumbsConstraints["ResizeSouth"] = 128] = "ResizeSouth";
    ThumbsConstraints[ThumbsConstraints["ResizeSouthWest"] = 256] = "ResizeSouthWest";
    ThumbsConstraints[ThumbsConstraints["ResizeWest"] = 512] = "ResizeWest";
    ThumbsConstraints[ThumbsConstraints["ResizeNorthWest"] = 1024] = "ResizeNorthWest";
    ThumbsConstraints[ThumbsConstraints["ResizeNorth"] = 2048] = "ResizeNorth";
    ThumbsConstraints[ThumbsConstraints["Default"] = 4094] = "Default";
})(ThumbsConstraints || (ThumbsConstraints = {}));
export var DiagramConstraints;
(function (DiagramConstraints) {
    DiagramConstraints[DiagramConstraints["None"] = 1] = "None";
    DiagramConstraints[DiagramConstraints["Bridging"] = 2] = "Bridging";
    DiagramConstraints[DiagramConstraints["UndoRedo"] = 4] = "UndoRedo";
    DiagramConstraints[DiagramConstraints["Tooltip"] = 8] = "Tooltip";
    DiagramConstraints[DiagramConstraints["UserInteraction"] = 16] = "UserInteraction";
    DiagramConstraints[DiagramConstraints["ApiUpdate"] = 32] = "ApiUpdate";
    DiagramConstraints[DiagramConstraints["PageEditable"] = 48] = "PageEditable";
    DiagramConstraints[DiagramConstraints["Zoom"] = 64] = "Zoom";
    DiagramConstraints[DiagramConstraints["PanX"] = 128] = "PanX";
    DiagramConstraints[DiagramConstraints["PanY"] = 256] = "PanY";
    DiagramConstraints[DiagramConstraints["Pan"] = 384] = "Pan";
    DiagramConstraints[DiagramConstraints["ZoomTextEdit"] = 512] = "ZoomTextEdit";
    DiagramConstraints[DiagramConstraints["Virtualization"] = 1024] = "Virtualization";
    DiagramConstraints[DiagramConstraints["LineRouting"] = 2048] = "LineRouting";
    DiagramConstraints[DiagramConstraints["Default"] = 500] = "Default";
})(DiagramConstraints || (DiagramConstraints = {}));
export var DiagramTools;
(function (DiagramTools) {
    DiagramTools[DiagramTools["None"] = 0] = "None";
    DiagramTools[DiagramTools["SingleSelect"] = 1] = "SingleSelect";
    DiagramTools[DiagramTools["MultipleSelect"] = 2] = "MultipleSelect";
    DiagramTools[DiagramTools["ZoomPan"] = 4] = "ZoomPan";
    DiagramTools[DiagramTools["DrawOnce"] = 8] = "DrawOnce";
    DiagramTools[DiagramTools["ContinuousDraw"] = 16] = "ContinuousDraw";
    DiagramTools[DiagramTools["Default"] = 3] = "Default";
})(DiagramTools || (DiagramTools = {}));
export var Transform;
(function (Transform) {
    Transform[Transform["Self"] = 1] = "Self";
    Transform[Transform["Parent"] = 2] = "Parent";
})(Transform || (Transform = {}));
export var RenderMode;
(function (RenderMode) {
    RenderMode[RenderMode["Canvas"] = 0] = "Canvas";
    RenderMode[RenderMode["Svg"] = 1] = "Svg";
})(RenderMode || (RenderMode = {}));
export var KeyModifiers;
(function (KeyModifiers) {
    KeyModifiers[KeyModifiers["None"] = 0] = "None";
    KeyModifiers[KeyModifiers["Control"] = 1] = "Control";
    KeyModifiers[KeyModifiers["Meta"] = 1] = "Meta";
    KeyModifiers[KeyModifiers["Alt"] = 2] = "Alt";
    KeyModifiers[KeyModifiers["Shift"] = 4] = "Shift";
})(KeyModifiers || (KeyModifiers = {}));
export var Keys;
(function (Keys) {
    Keys[Keys["None"] = null] = "None";
    Keys[Keys["Number0"] = 0] = "Number0";
    Keys[Keys["Number1"] = 1] = "Number1";
    Keys[Keys["Number2"] = 2] = "Number2";
    Keys[Keys["Number3"] = 3] = "Number3";
    Keys[Keys["Number4"] = 4] = "Number4";
    Keys[Keys["Number5"] = 5] = "Number5";
    Keys[Keys["Number6"] = 6] = "Number6";
    Keys[Keys["Number7"] = 7] = "Number7";
    Keys[Keys["Number8"] = 8] = "Number8";
    Keys[Keys["Number9"] = 9] = "Number9";
    Keys[Keys["A"] = 65] = "A";
    Keys[Keys["B"] = 66] = "B";
    Keys[Keys["C"] = 67] = "C";
    Keys[Keys["D"] = 68] = "D";
    Keys[Keys["E"] = 69] = "E";
    Keys[Keys["F"] = 70] = "F";
    Keys[Keys["G"] = 71] = "G";
    Keys[Keys["H"] = 72] = "H";
    Keys[Keys["I"] = 73] = "I";
    Keys[Keys["J"] = 74] = "J";
    Keys[Keys["K"] = 75] = "K";
    Keys[Keys["L"] = 76] = "L";
    Keys[Keys["M"] = 77] = "M";
    Keys[Keys["N"] = 78] = "N";
    Keys[Keys["O"] = 79] = "O";
    Keys[Keys["P"] = 80] = "P";
    Keys[Keys["Q"] = 81] = "Q";
    Keys[Keys["R"] = 82] = "R";
    Keys[Keys["S"] = 83] = "S";
    Keys[Keys["T"] = 84] = "T";
    Keys[Keys["U"] = 85] = "U";
    Keys[Keys["V"] = 86] = "V";
    Keys[Keys["W"] = 87] = "W";
    Keys[Keys["X"] = 88] = "X";
    Keys[Keys["Y"] = 89] = "Y";
    Keys[Keys["Z"] = 90] = "Z";
    Keys[Keys["Left"] = 37] = "Left";
    Keys[Keys["Up"] = 38] = "Up";
    Keys[Keys["Right"] = 39] = "Right";
    Keys[Keys["Down"] = 40] = "Down";
    Keys[Keys["Escape"] = 27] = "Escape";
    Keys[Keys["Space"] = 32] = "Space";
    Keys[Keys["PageUp"] = 33] = "PageUp";
    Keys[Keys["PageDown"] = 34] = "PageDown";
    Keys[Keys["End"] = 35] = "End";
    Keys[Keys["Home"] = 36] = "Home";
    Keys[Keys["Delete"] = 46] = "Delete";
    Keys[Keys["Tab"] = 9] = "Tab";
    Keys[Keys["Enter"] = 13] = "Enter";
    Keys[Keys["BackSpace"] = 8] = "BackSpace";
    Keys[Keys["F1"] = 112] = "F1";
    Keys[Keys["F2"] = 113] = "F2";
    Keys[Keys["F3"] = 114] = "F3";
    Keys[Keys["F4"] = 115] = "F4";
    Keys[Keys["F5"] = 116] = "F5";
    Keys[Keys["F6"] = 117] = "F6";
    Keys[Keys["F7"] = 118] = "F7";
    Keys[Keys["F8"] = 119] = "F8";
    Keys[Keys["F9"] = 120] = "F9";
    Keys[Keys["F10"] = 121] = "F10";
    Keys[Keys["F11"] = 122] = "F11";
    Keys[Keys["F12"] = 123] = "F12";
    Keys[Keys["Star"] = 56] = "Star";
    Keys[Keys["Plus"] = 187] = "Plus";
    Keys[Keys["Minus"] = 189] = "Minus";
})(Keys || (Keys = {}));
export var DiagramAction;
(function (DiagramAction) {
    DiagramAction[DiagramAction["Render"] = 2] = "Render";
    DiagramAction[DiagramAction["PublicMethod"] = 4] = "PublicMethod";
    DiagramAction[DiagramAction["ToolAction"] = 8] = "ToolAction";
    DiagramAction[DiagramAction["UndoRedo"] = 16] = "UndoRedo";
    DiagramAction[DiagramAction["TextEdit"] = 32] = "TextEdit";
    DiagramAction[DiagramAction["Group"] = 64] = "Group";
    DiagramAction[DiagramAction["Clear"] = 128] = "Clear";
    DiagramAction[DiagramAction["PreventClearSelection"] = 256] = "PreventClearSelection";
    DiagramAction[DiagramAction["Interactions"] = 512] = "Interactions";
    DiagramAction[DiagramAction["PreventHistory"] = 1024] = "PreventHistory";
    DiagramAction[DiagramAction["PreventIconsUpdate"] = 2048] = "PreventIconsUpdate";
    DiagramAction[DiagramAction["PreventCollectionChangeOnDragOver"] = 4096] = "PreventCollectionChangeOnDragOver";
    DiagramAction[DiagramAction["PreventZIndexOnDragging"] = 8192] = "PreventZIndexOnDragging";
    DiagramAction[DiagramAction["isGroupDragging"] = 16384] = "isGroupDragging";
    DiagramAction[DiagramAction["DragUsingMouse"] = 32768] = "DragUsingMouse";
    DiagramAction[DiagramAction["DecoratorPropertyChange"] = 65536] = "DecoratorPropertyChange";
    DiagramAction[DiagramAction["PreventLaneContainerUpdate"] = 131072] = "PreventLaneContainerUpdate";
})(DiagramAction || (DiagramAction = {}));
export var RendererAction;
(function (RendererAction) {
    RendererAction[RendererAction["None"] = 2] = "None";
    RendererAction[RendererAction["DrawSelectorBorder"] = 4] = "DrawSelectorBorder";
    RendererAction[RendererAction["PreventRenderSelector"] = 8] = "PreventRenderSelector";
})(RendererAction || (RendererAction = {}));
export var RealAction;
(function (RealAction) {
    RealAction[RealAction["None"] = 0] = "None";
    RealAction[RealAction["PreventDrag"] = 2] = "PreventDrag";
    RealAction[RealAction["PreventScale"] = 4] = "PreventScale";
    RealAction[RealAction["PreventDataInit"] = 8] = "PreventDataInit";
    RealAction[RealAction["hScrollbarMoved"] = 16] = "hScrollbarMoved";
    RealAction[RealAction["vScrollbarMoved"] = 32] = "vScrollbarMoved";
    RealAction[RealAction["AnimationClick"] = 64] = "AnimationClick";
    RealAction[RealAction["EnableGroupAction"] = 128] = "EnableGroupAction";
    RealAction[RealAction["PanInProgress"] = 256] = "PanInProgress";
    RealAction[RealAction["OverViewAction"] = 512] = "OverViewAction";
})(RealAction || (RealAction = {}));
export var ScrollActions;
(function (ScrollActions) {
    ScrollActions[ScrollActions["None"] = 0] = "None";
    ScrollActions[ScrollActions["PropertyChange"] = 1024] = "PropertyChange";
    ScrollActions[ScrollActions["Interaction"] = 2048] = "Interaction";
})(ScrollActions || (ScrollActions = {}));
export var NoOfSegments;
(function (NoOfSegments) {
    NoOfSegments[NoOfSegments["Zero"] = 0] = "Zero";
    NoOfSegments[NoOfSegments["One"] = 1] = "One";
    NoOfSegments[NoOfSegments["Two"] = 2] = "Two";
    NoOfSegments[NoOfSegments["Three"] = 3] = "Three";
    NoOfSegments[NoOfSegments["Four"] = 4] = "Four";
    NoOfSegments[NoOfSegments["Five"] = 5] = "Five";
})(NoOfSegments || (NoOfSegments = {}));
export var DiagramEvent;
(function (DiagramEvent) {
    DiagramEvent[DiagramEvent["collectionChange"] = 0] = "collectionChange";
    DiagramEvent[DiagramEvent["rotateChange"] = 1] = "rotateChange";
    DiagramEvent[DiagramEvent["positionChange"] = 2] = "positionChange";
    DiagramEvent[DiagramEvent["propertyChange"] = 3] = "propertyChange";
    DiagramEvent[DiagramEvent["selectionChange"] = 4] = "selectionChange";
    DiagramEvent[DiagramEvent["sizeChange"] = 5] = "sizeChange";
    DiagramEvent[DiagramEvent["drop"] = 6] = "drop";
    DiagramEvent[DiagramEvent["sourcePointChange"] = 7] = "sourcePointChange";
    DiagramEvent[DiagramEvent["targetPointChange"] = 8] = "targetPointChange";
    DiagramEvent[DiagramEvent["connectionChange"] = 9] = "connectionChange";
    DiagramEvent[DiagramEvent["animationComplete"] = 10] = "animationComplete";
    DiagramEvent[DiagramEvent["click"] = 11] = "click";
    DiagramEvent[DiagramEvent["doubleClick"] = 12] = "doubleClick";
    DiagramEvent[DiagramEvent["scrollChange"] = 13] = "scrollChange";
    DiagramEvent[DiagramEvent["dragEnter"] = 14] = "dragEnter";
    DiagramEvent[DiagramEvent["dragLeave"] = 15] = "dragLeave";
    DiagramEvent[DiagramEvent["dragOver"] = 16] = "dragOver";
    DiagramEvent[DiagramEvent["textEdit"] = 17] = "textEdit";
    DiagramEvent[DiagramEvent["paletteSelectionChange"] = 18] = "paletteSelectionChange";
    DiagramEvent[DiagramEvent["historyChange"] = 19] = "historyChange";
    DiagramEvent[DiagramEvent["mouseEnter"] = 20] = "mouseEnter";
    DiagramEvent[DiagramEvent["mouseLeave"] = 21] = "mouseLeave";
    DiagramEvent[DiagramEvent["mouseOver"] = 22] = "mouseOver";
    DiagramEvent[DiagramEvent["expandStateChange"] = 23] = "expandStateChange";
    DiagramEvent[DiagramEvent["segmentCollectionChange"] = 24] = "segmentCollectionChange";
    DiagramEvent[DiagramEvent["commandExecute"] = 25] = "commandExecute";
    DiagramEvent[DiagramEvent["historyStateChange"] = 26] = "historyStateChange";
    DiagramEvent[DiagramEvent["onUserHandleMouseDown"] = 27] = "onUserHandleMouseDown";
    DiagramEvent[DiagramEvent["onUserHandleMouseUp"] = 28] = "onUserHandleMouseUp";
    DiagramEvent[DiagramEvent["onUserHandleMouseEnter"] = 29] = "onUserHandleMouseEnter";
    DiagramEvent[DiagramEvent["onUserHandleMouseLeave"] = 30] = "onUserHandleMouseLeave";
    DiagramEvent[DiagramEvent["onImageLoad"] = 31] = "onImageLoad";
    DiagramEvent[DiagramEvent["onDoBindingInit"] = 32] = "onDoBindingInit";
    DiagramEvent[DiagramEvent["keyUp"] = 33] = "keyUp";
    DiagramEvent[DiagramEvent["keyDown"] = 34] = "keyDown";
    DiagramEvent[DiagramEvent["fixedUserHandleClick"] = 35] = "fixedUserHandleClick";
})(DiagramEvent || (DiagramEvent = {}));
export var PortConstraints;
(function (PortConstraints) {
    PortConstraints[PortConstraints["None"] = 1] = "None";
    PortConstraints[PortConstraints["Drag"] = 2] = "Drag";
    PortConstraints[PortConstraints["Draw"] = 4] = "Draw";
    PortConstraints[PortConstraints["InConnect"] = 8] = "InConnect";
    PortConstraints[PortConstraints["OutConnect"] = 16] = "OutConnect";
    PortConstraints[PortConstraints["Default"] = 24] = "Default";
})(PortConstraints || (PortConstraints = {}));
export var contextMenuClick = 'contextMenuClick';
export var contextMenuOpen = 'contextMenuOpen';
export var contextMenuBeforeItemRender = 'contextMenuBeforeItemRender';
