// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/blazor/actions/base-quick-toolbar.js
import { append, selectAll, closest, isNullOrUndefined as isNOU } from '@syncfusion/ej2-base';
import { addClass, removeClass, Browser, setStyleAttribute } from '@syncfusion/ej2-base';
import { isCollide } from '@syncfusion/ej2-popups';
import * as constant from '../constant';
import * as classes from '../classes';
import { isIDevice, setToolbarStatus } from '../util';
import { PopupRenderer } from '../renderer/popup-renderer';
var BaseQuickToolbar = (function () {
    function BaseQuickToolbar(parent) {
        this.parent = parent;
        this.isDOMElement = false;
        this.popupRenderer = new PopupRenderer(parent);
    }
    BaseQuickToolbar.prototype.render = function (element, type) {
        this.element = element;
        this.popupRenderer.renderPopup(this, type);
        this.addEventListener();
    };
    BaseQuickToolbar.prototype.setPosition = function (e) {
        var x;
        var y;
        var imgContainer = closest(e.target, '.' + classes.CLS_CAPTION);
        var target = !isNOU(imgContainer) ? imgContainer : e.target;
        var targetOffsetTop = target.offsetTop;
        var parentOffsetTop = window.pageYOffset + e.parentData.top;
        if ((targetOffsetTop - e.editTop) > e.popHeight) {
            y = parentOffsetTop + e.tBarElementHeight + (targetOffsetTop - e.editTop) - e.popHeight - 5;
        }
        else if (((e.editTop + e.editHeight) - (targetOffsetTop + target.offsetHeight)) > e.popHeight) {
            y = parentOffsetTop + e.tBarElementHeight + (targetOffsetTop - e.editTop) + target.offsetHeight + 5;
        }
        else {
            y = e.y;
        }
        if (target.offsetWidth > e.popWidth) {
            x = (target.offsetWidth / 2) - (e.popWidth / 2) + e.parentData.left + target.offsetLeft;
        }
        else {
            x = e.parentData.left + target.offsetLeft;
        }
        this.popupObj.position.X = ((x + e.popWidth) > e.parentData.right) ? e.parentData.right - e.popWidth : x;
        this.popupObj.position.Y = (y >= 0) ? y : e.y + 5;
        this.popupObj.dataBind();
    };
    BaseQuickToolbar.prototype.checkCollision = function (e, viewPort, type) {
        var x;
        var y;
        var parentTop = e.parentData.top;
        var contentTop = e.windowY + parentTop + e.tBarElementHeight;
        var collision = [];
        if (viewPort === 'document') {
            collision = isCollide(e.popup);
        }
        else {
            collision = isCollide(e.popup, e.parentElement);
        }
        for (var i = 0; i < collision.length; i++) {
            switch (collision[i]) {
                case 'top':
                    if (viewPort === 'document') {
                        y = e.windowY;
                    }
                    else {
                        y = (window.pageYOffset + parentTop) + e.tBarElementHeight;
                    }
                    break;
                case 'bottom':
                    var posY = void 0;
                    if (viewPort === 'document') {
                        if (type === 'inline') {
                            posY = (e.y - e.popHeight - 10);
                        }
                        else {
                            if ((e.windowHeight - (parentTop + e.tBarElementHeight)) > e.popHeight) {
                                if ((contentTop - e.windowHeight) > e.popHeight) {
                                    posY = (contentTop + (e.windowHeight - parentTop)) - e.popHeight;
                                }
                                else {
                                    posY = contentTop;
                                }
                            }
                            else {
                                posY = e.windowY + (parentTop + e.tBarElementHeight);
                            }
                        }
                    }
                    else {
                        if (e.target.tagName !== 'IMG') {
                            posY = (e.parentData.bottom + window.pageYOffset) - e.popHeight - 10;
                        }
                        else {
                            posY = (e.parentData.bottom + window.pageYOffset) - e.popHeight - 5;
                        }
                    }
                    y = posY;
                    break;
                case 'right':
                    if (type === 'inline') {
                        x = e.windowWidth - (e.popWidth + e.bodyRightSpace + 10);
                    }
                    else {
                        x = e.x - e.popWidth;
                    }
                    break;
                case 'left':
                    if (type === 'inline') {
                        x = 0;
                    }
                    else {
                        x = e.parentData.left;
                    }
                    break;
            }
        }
        this.popupObj.position.X = (x) ? x : this.popupObj.position.X;
        this.popupObj.position.Y = (y) ? y : this.popupObj.position.Y;
        this.popupObj.dataBind();
    };
    BaseQuickToolbar.prototype.showPopup = function (x, y, target, type) {
        var _this = this;
        if (this.parent.onQuickTbOpenEnabled) {
            this.parent.dotNetRef.invokeMethodAsync(constant.beforeQuickToolbarOpenEvent).then(function (args) {
                if (!args.cancel) {
                    _this.onQuickTbOpenCallback(x, y, target, type);
                }
            });
        }
        else {
            this.onQuickTbOpenCallback(x, y, target, type);
        }
    };
    BaseQuickToolbar.prototype.onQuickTbOpenCallback = function (x, y, target, type) {
        var editPanelTop;
        var editPanelHeight;
        var bodyStyle = window.getComputedStyle(document.body);
        var bodyRight = parseFloat(bodyStyle.marginRight.split('px')[0]) + parseFloat(bodyStyle.paddingRight.split('px')[0]);
        var windowHeight = window.innerHeight;
        var windowWidth = window.innerWidth;
        var parent = this.parent.element;
        var toolbarAvail = !isNOU(this.parent.getToolbar());
        var tbHeight = toolbarAvail && this.parent.toolbarModule.getToolbarHeight();
        var expTBHeight = toolbarAvail && this.parent.toolbarModule.getExpandTBarPopHeight();
        var tBarHeight = (toolbarAvail) ? (tbHeight + expTBHeight) : 0;
        addClass([this.element], [classes.CLS_HIDE]);
        if (Browser.isDevice && !isIDevice()) {
            addClass([this.parent.getToolbar()], [classes.CLS_HIDE]);
        }
        if (this.parent.iframeSettings.enable) {
            var cntEle = this.parent.getPanel().contentWindow;
            editPanelTop = cntEle.pageYOffset;
            editPanelHeight = cntEle.innerHeight;
        }
        else {
            var cntEle = closest(target, '.' + classes.CLS_RTE_CONTENT);
            editPanelTop = (cntEle) ? cntEle.scrollTop : 0;
            editPanelHeight = (cntEle) ? cntEle.offsetHeight : 0;
        }
        if (!this.parent.inlineMode.enable && !closest(target, 'table')) {
        }
        append([this.element], document.body);
        this.popupObj.position.X = x + 20;
        this.popupObj.position.Y = y + ((this.parent.iframeSettings.enable) ? 35 : 20);
        this.popupObj.dataBind();
        this.popupObj.element.classList.add(classes.CLS_POPUP_OPEN);
        var showPopupData = {
            x: x, y: y,
            target: target,
            editTop: editPanelTop,
            editHeight: editPanelHeight,
            popup: this.popupObj.element,
            popHeight: this.popupObj.element.offsetHeight,
            popWidth: this.popupObj.element.offsetWidth,
            parentElement: parent,
            bodyRightSpace: bodyRight,
            windowY: window.pageYOffset,
            windowHeight: windowHeight,
            windowWidth: windowWidth,
            parentData: parent.getBoundingClientRect(),
            tBarElementHeight: tBarHeight
        };
        if (target.tagName === 'IMG') {
            this.setPosition(showPopupData);
        }
        if (!this.parent.inlineMode.enable) {
            this.checkCollision(showPopupData, 'parent', '');
        }
        this.checkCollision(showPopupData, 'document', ((this.parent.inlineMode.enable) ? 'inline' : ''));
        this.popupObj.element.classList.remove(classes.CLS_POPUP_CLOSE);
        removeClass([this.element], [classes.CLS_HIDE]);
        this.popupObj.show({ name: 'ZoomIn', duration: (Browser.isIE ? 250 : 400) });
        setStyleAttribute(this.element, {
            maxWidth: this.parent.element.offsetWidth + 'px'
        });
        addClass([this.element], [classes.CLS_POP]);
        this.isDOMElement = true;
        this.parent.dotNetRef.invokeMethodAsync(constant.updateClass, this.popupObj.element.classList.toString(), type);
    };
    BaseQuickToolbar.prototype.hidePopup = function () {
        var viewSourcePanel = this.parent.viewSourceModule.getViewPanel();
        if (Browser.isDevice && !isIDevice()) {
            removeClass([this.parent.getToolbar()], [classes.CLS_HIDE]);
        }
        if (!isNOU(this.parent.getToolbar()) && !this.parent.inlineMode.enable) {
            if (isNOU(viewSourcePanel) || viewSourcePanel.style.display === 'none') {
            }
        }
        this.removeEleFromDOM();
        this.isDOMElement = false;
    };
    BaseQuickToolbar.prototype.removeEleFromDOM = function () {
        this.popupObj.hide();
        this.element.classList.add(classes.CLS_HIDE);
        this.element.classList.add(classes.CLS_RTE_QUICK_POPUP_HIDE);
        if (this.isDOMElement) {
            removeClass([this.element], [classes.CLS_POP]);
            this.popupObj.element.removeAttribute('style');
            this.popupObj.destroy();
            if (this.parent.quickTbClosedEnabled) {
                this.parent.dotNetRef.invokeMethodAsync(constant.quickToolbarCloseEvent);
            }
        }
    };
    BaseQuickToolbar.prototype.updateStatus = function (args) {
        var tbElements = selectAll('.' + classes.CLS_TB_ITEM, this.element);
        if (tbElements.length <= 0) {
            return;
        }
        var options = {
            args: args,
            dropDownModule: null,
            parent: this.parent,
            tbElements: tbElements,
            tbItems: this.parent.toolbarSettings.items
        };
        setToolbarStatus(options, true);
    };
    BaseQuickToolbar.prototype.addEventListener = function () {
        this.parent.observer.on(constant.destroy, this.destroy, this);
        if (this.parent.inlineMode.enable) {
            this.parent.observer.on(constant.toolbarUpdated, this.updateStatus, this);
        }
    };
    BaseQuickToolbar.prototype.removeEventListener = function () {
        this.parent.observer.off(constant.destroy, this.destroy);
        if (this.parent.inlineMode.enable) {
            this.parent.observer.off(constant.toolbarUpdated, this.updateStatus);
        }
    };
    BaseQuickToolbar.prototype.destroy = function () {
        if (this.popupObj && !this.popupObj.isDestroyed) {
            this.popupObj.destroy();
            this.removeEleFromDOM();
        }
        this.removeEventListener();
    };
    return BaseQuickToolbar;
}());
export { BaseQuickToolbar };
