// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/blazor/actions/markdown-editor.js
import { extend, isNullOrUndefined, addClass, removeClass } from '@syncfusion/ej2-base';
import * as events from '../constant';
import * as classes from '../classes';
import { MarkdownToolbarStatus } from './markdown-toolbar-status';
import { MarkdownFormatter } from '../formatter/markdown-formatter';
import { MarkdownSelection } from '../../src/markdown-parser/plugin/markdown-selection';
var MarkdownEditor = (function () {
    function MarkdownEditor(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    MarkdownEditor.prototype.addEventListener = function () {
        this.saveSelection = new MarkdownSelection();
        this.parent.observer.on(events.initialEnd, this.render, this);
        this.parent.observer.on(events.markdownToolbarClick, this.onToolbarClick, this);
        this.parent.observer.on(events.destroy, this.destroy, this);
        this.parent.observer.on(events.selectAll, this.selectAll, this);
        this.parent.observer.on(events.getSelectedHtml, this.getSelectedHtml, this);
        this.parent.observer.on(events.selectionSave, this.onSelectionSave, this);
        this.parent.observer.on(events.selectionRestore, this.onSelectionRestore, this);
        this.parent.observer.on(events.readOnlyMode, this.updateReadOnly, this);
    };
    MarkdownEditor.prototype.updateReadOnly = function () {
        if (this.parent.readonly) {
            this.parent.getEditPanel().setAttribute('readonly', 'readonly');
            addClass([this.parent.element], classes.CLS_RTE_READONLY);
        }
        else {
            this.parent.getEditPanel().removeAttribute('readonly');
            removeClass([this.parent.element], classes.CLS_RTE_READONLY);
        }
    };
    MarkdownEditor.prototype.onSelectionSave = function () {
        var textArea = this.parent.getEditPanel();
        this.saveSelection.save(textArea.selectionStart, textArea.selectionEnd);
    };
    MarkdownEditor.prototype.onSelectionRestore = function (e) {
        this.parent.getEditPanel().focus();
        var textArea = this.parent.getEditPanel();
        this.saveSelection.restore(textArea);
    };
    MarkdownEditor.prototype.onToolbarClick = function (args) {
        var text;
        var startOffset;
        var endOffset;
        var item = args.item;
        var textArea = this.parent.getEditPanel();
        textArea.focus();
        startOffset = textArea.selectionStart;
        endOffset = textArea.selectionEnd;
        text = textArea.value.substring(startOffset, endOffset);
        switch (item.subCommand) {
            case 'Maximize':
                this.parent.observer.notify(events.enableFullScreen, { args: args });
                break;
            case 'Minimize':
                this.parent.observer.notify(events.disableFullScreen, { args: args });
                break;
            case 'CreateLink':
                this.parent.observer.notify(events.insertLink, { member: 'link', args: args, text: text, module: 'Markdown' });
                break;
            case 'Image':
                this.parent.observer.notify(events.insertImage, { member: 'image', args: args, text: text, module: 'Markdown' });
                break;
            case 'CreateTable':
                var tableConstant = {
                    'headingText': this.parent.localeData.headingText,
                    'colText': this.parent.localeData.colText
                };
                this.parent.formatter.process(this.parent, args, args.originalEvent, tableConstant);
                break;
            default:
                this.parent.formatter.process(this.parent, args, args.originalEvent, null);
                break;
        }
    };
    MarkdownEditor.prototype.removeEventListener = function () {
        this.parent.observer.off(events.initialEnd, this.render);
        this.parent.observer.off(events.destroy, this.destroy);
        this.parent.observer.off(events.markdownToolbarClick, this.onToolbarClick);
        this.parent.observer.off(events.selectAll, this.selectAll);
        this.parent.observer.off(events.getSelectedHtml, this.getSelectedHtml);
        this.parent.observer.off(events.selectionSave, this.onSelectionSave);
        this.parent.observer.off(events.selectionRestore, this.onSelectionRestore);
        this.parent.observer.off(events.readOnlyMode, this.updateReadOnly);
    };
    MarkdownEditor.prototype.render = function () {
        var editElement = this.parent.getEditPanel();
        var option = { undoRedoSteps: this.parent.undoRedoSteps, undoRedoTimer: this.parent.undoRedoTimer };
        if (isNullOrUndefined(this.parent.adapter)) {
            this.parent.formatter = new MarkdownFormatter({
                element: editElement,
                options: option
            });
        }
        else {
            this.parent.formatter = new MarkdownFormatter(extend({}, this.parent.adapter, {
                element: editElement,
                options: option
            }));
        }
        if (this.parent.toolbarSettings.enable) {
            this.toolbarUpdate = new MarkdownToolbarStatus(this.parent);
        }
        this.parent.observer.notify(events.bindOnEnd, {});
    };
    MarkdownEditor.prototype.selectAll = function () {
        this.parent.formatter.editorManager.markdownSelection.setSelection(this.parent.getEditPanel(), 0, this.parent.getEditPanel().value.length);
    };
    MarkdownEditor.prototype.getSelectedHtml = function (e) {
        e.callBack(this.parent.formatter.editorManager.markdownSelection.getSelectedText(this.parent.getEditPanel()));
    };
    MarkdownEditor.prototype.destroy = function () {
        this.removeEventListener();
    };
    return MarkdownEditor;
}());
export { MarkdownEditor };
