// @syncfusion/ej2/ej2-resource/scripts/ej2-richtexteditor/blazor/actions/quick-toolbar.js
import { select, isNullOrUndefined as isNOU, Browser, EventHandler, closest } from '@syncfusion/ej2-base';
import * as events from '../constant';
import { BaseQuickToolbar } from './base-quick-toolbar';
import { pageYOffset, hasClass, isIDevice } from '../util';
import { CLS_INLINE_POP, CLS_POPUP, CLS_HIDE, CLS_POPUP_CLOSE } from '../classes';
var QuickToolbar = (function () {
    function QuickToolbar(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    QuickToolbar.prototype.showImageQTBar = function (x, y, target, type) {
        if (this.parent.readonly || !this.parent.quickToolbarSettings.enable) {
            return;
        }
        var popupTarget = document.querySelector('#' + this.parent.element.id + events.imageQuickPopup);
        this.imageQTBar = new BaseQuickToolbar(this.parent);
        this.imageQTBar.render(popupTarget, type);
        this.parent.dotNetRef.invokeMethodAsync(events.showImagePopup, this.imageQTBar.popupObj.element.classList.toString(), type);
        this.imageQTBar.showPopup(x, y, target, type);
    };
    QuickToolbar.prototype.hideImageQTBar = function () {
        if (this.imageQTBar) {
            this.imageQTBar.hidePopup();
            this.imageQTBar = undefined;
            this.parent.dotNetRef.invokeMethodAsync(events.hideImagePopup);
        }
    };
    QuickToolbar.prototype.showLinkQTBar = function (x, y, target, type) {
        if (this.parent.readonly || !this.parent.quickToolbarSettings.enable) {
            return;
        }
        var popupTarget = document.querySelector('#' + this.parent.element.id + events.linkQuickPopup);
        this.linkQTBar = new BaseQuickToolbar(this.parent);
        this.linkQTBar.render(popupTarget, type);
        this.parent.dotNetRef.invokeMethodAsync(events.showLinkPopup, this.linkQTBar.popupObj.element.classList.toString());
        this.linkQTBar.showPopup(x, y, target, type);
    };
    QuickToolbar.prototype.hideLinkQTBar = function () {
        if (this.linkQTBar) {
            this.linkQTBar.hidePopup();
            this.linkQTBar = undefined;
            this.parent.dotNetRef.invokeMethodAsync(events.hideLinkPopup);
        }
    };
    QuickToolbar.prototype.showTableQTBar = function (x, y, target, type) {
        if (this.parent.readonly || !this.parent.quickToolbarSettings.enable) {
            return;
        }
        var popupTarget = document.querySelector('#' + this.parent.element.id + events.tableQuickPopup);
        this.tableQTBar = new BaseQuickToolbar(this.parent);
        this.tableQTBar.render(popupTarget, type);
        this.parent.dotNetRef.invokeMethodAsync(events.showTablePopup, this.tableQTBar.popupObj.element.classList.toString());
        this.tableQTBar.showPopup(x, y, target, type);
    };
    QuickToolbar.prototype.hideTableQTBar = function () {
        if (this.tableQTBar) {
            this.tableQTBar.hidePopup();
            this.tableQTBar = undefined;
            this.parent.dotNetRef.invokeMethodAsync(events.hideTablePopup);
        }
    };
    QuickToolbar.prototype.showInlineQTBar = function (x, y, target) {
        if (this.parent.readonly || !this.parent.inlineMode.enable) {
            return;
        }
        if (this.parent.inlineMode.enable && (!Browser.isDevice || isIDevice())) {
            var popupTarget = document.querySelector('#' + this.parent.element.id + events.inlineQuickPopup);
            this.inlineQTBar = new BaseQuickToolbar(this.parent);
            this.inlineQTBar.render(popupTarget, 'Inline');
            EventHandler.add(this.inlineQTBar.element, 'mousedown', this.onMouseDown, this);
            this.parent.dotNetRef.invokeMethodAsync(events.showInlinePopup, this.inlineQTBar.popupObj.element.classList.toString());
            this.inlineQTBar.showPopup(x, y, target, 'Inline');
        }
    };
    QuickToolbar.prototype.hideInlineQTBar = function () {
        if (this.inlineQTBar) {
            this.inlineQTBar.hidePopup();
            this.inlineQTBar = undefined;
            this.parent.dotNetRef.invokeMethodAsync(events.hideInlinePopup);
        }
    };
    QuickToolbar.prototype.hideQuickToolbars = function () {
        if (!isNOU(this.linkQTBar)) {
            this.hideLinkQTBar();
        }
        if (!isNOU(this.imageQTBar)) {
            this.hideImageQTBar();
        }
        if (!isNOU(this.tableQTBar)) {
            this.hideTableQTBar();
        }
        if (this.parent.inlineMode.enable && (!Browser.isDevice || isIDevice())) {
            this.hideInlineQTBar();
        }
    };
    QuickToolbar.prototype.scrollHandler = function () {
        if (this.parent.quickToolbarSettings.actionOnScroll.toLocaleLowerCase() === 'hide') {
            this.hideQuickToolbars();
        }
    };
    QuickToolbar.prototype.selectionChangeHandler = function (e) {
        var _this = this;
        if (!this.parent.inlineMode.onSelection) {
            return;
        }
        clearTimeout(this.deBouncer);
        this.deBouncer = window.setTimeout(function () { _this.onSelectionChange(e); }, 1000);
    };
    QuickToolbar.prototype.onSelectionChange = function (e) {
        if (!isNOU(select('.' + CLS_INLINE_POP + '.' + CLS_POPUP, document.body))) {
            return;
        }
        var selection = this.parent.getDocument().getSelection();
        if (!selection.isCollapsed) {
            this.mouseUpHandler({ args: e });
        }
    };
    QuickToolbar.prototype.toolbarUpdated = function (args) {
        if (!isNOU(this.linkQTBar)) {
            this.hideLinkQTBar();
        }
        if (!isNOU(this.imageQTBar)) {
            this.hideImageQTBar();
        }
        if (!isNOU(this.tableQTBar)) {
            this.hideTableQTBar();
        }
    };
    QuickToolbar.prototype.deBounce = function (x, y, target) {
        var _this = this;
        clearTimeout(this.deBouncer);
        this.deBouncer = window.setTimeout(function () { _this.showInlineQTBar(x, y, target); }, 1000);
    };
    QuickToolbar.prototype.onMouseDown = function (e) {
        this.parent.isBlur = false;
        this.parent.isRTE = true;
    };
    QuickToolbar.prototype.mouseUpHandler = function (e) {
        if (this.parent.inlineMode.enable && (!Browser.isDevice || isIDevice())) {
            var coordinates = void 0;
            coordinates = e.args.touches ? e.args.changedTouches[0] : e.args;
            var range = this.parent.getRange();
            var target = e.args.target;
            var inlinePopEle = select('.' + CLS_INLINE_POP, document.body);
            if (isNOU(inlinePopEle) || inlinePopEle.classList.contains(CLS_HIDE)) {
                if (isIDevice() && e.touchData && e.touchData.prevClientX !== e.touchData.clientX
                    && e.touchData.prevClientY !== e.touchData.clientY) {
                    return;
                }
                this.hideInlineQTBar();
                this.offsetX = coordinates.pageX;
                this.offsetY = pageYOffset(coordinates, this.parent.element, this.parent.iframeSettings.enable);
                if (target.nodeName === 'TEXTAREA') {
                    this.showInlineQTBar(this.offsetX, this.offsetY, target);
                }
                else {
                    var closestAnchor = closest(target, 'a');
                    target = closestAnchor ? closestAnchor : target;
                    if (target.tagName !== 'IMG' && target.tagName !== 'A' && (!closest(target, 'td,th') || !range.collapsed)) {
                        if (this.parent.inlineMode.onSelection && range.collapsed) {
                            return;
                        }
                        this.target = target;
                        this.showInlineQTBar(this.offsetX, this.offsetY, target);
                    }
                }
            }
        }
    };
    QuickToolbar.prototype.keyDownHandler = function () {
        if ((this.parent.inlineMode.enable && (!Browser.isDevice || isIDevice()))
            && !isNOU(select('.' + CLS_INLINE_POP + '.' + CLS_POPUP, document))) {
            this.hideInlineQTBar();
        }
    };
    QuickToolbar.prototype.inlineQTBarMouseDownHandler = function () {
        if ((this.parent.inlineMode.enable && (!Browser.isDevice || isIDevice()))
            && !isNOU(select('.' + CLS_INLINE_POP + '.' + CLS_POPUP, document))) {
            this.hideInlineQTBar();
        }
    };
    QuickToolbar.prototype.keyUpHandler = function (e) {
        if (this.parent.inlineMode.enable && !Browser.isDevice) {
            if (this.parent.inlineMode.onSelection) {
                return;
            }
            var args = e.args;
            this.deBounce(this.offsetX, this.offsetY, args.target);
        }
    };
    QuickToolbar.prototype.onKeyDown = function (e) {
        var args = e.args;
        if (args.which === 8 || args.which === 46) {
            if (this.imageQTBar && !hasClass(this.imageQTBar.element, CLS_POPUP_CLOSE)) {
                this.imageQTBar.hidePopup();
            }
        }
    };
    QuickToolbar.prototype.onIframeMouseDown = function () {
        this.hideQuickToolbars();
        this.hideInlineQTBar();
    };
    QuickToolbar.prototype.addEventListener = function () {
        if (this.parent.inlineMode.enable && this.parent.inlineMode.onSelection && isIDevice()) {
            EventHandler.add(this.parent.getDocument(), 'selectionchange', this.selectionChangeHandler, this);
        }
        this.parent.observer.on(events.toolbarUpdated, this.toolbarUpdated, this);
        this.wireInlineQTBarEvents();
        this.parent.observer.on(events.scroll, this.scrollHandler, this);
        this.parent.observer.on(events.contentscroll, this.scrollHandler, this);
        this.parent.observer.on(events.focusChange, this.hideQuickToolbars, this);
        this.parent.observer.on(events.iframeMouseDown, this.onIframeMouseDown, this);
        this.parent.observer.on(events.destroy, this.destroy, this);
        this.parent.observer.on(events.keyDown, this.onKeyDown, this);
    };
    QuickToolbar.prototype.removeEventListener = function () {
        EventHandler.remove(this.parent.getDocument(), 'selectionchange', this.selectionChangeHandler);
        this.parent.observer.off(events.toolbarUpdated, this.toolbarUpdated);
        this.unWireInlineQTBarEvents();
        this.parent.observer.off(events.scroll, this.scrollHandler);
        this.parent.observer.off(events.contentscroll, this.scrollHandler);
        this.parent.observer.off(events.focusChange, this.hideQuickToolbars);
        this.parent.observer.off(events.destroy, this.destroy);
        this.parent.observer.off(events.iframeMouseDown, this.onIframeMouseDown);
        this.parent.observer.off(events.keyDown, this.onKeyDown);
    };
    QuickToolbar.prototype.wireInlineQTBarEvents = function () {
        this.parent.observer.on(events.mouseUp, this.mouseUpHandler, this);
        this.parent.observer.on(events.mouseDown, this.inlineQTBarMouseDownHandler, this);
        this.parent.observer.on(events.keyDown, this.keyDownHandler, this);
        this.parent.observer.on(events.keyUp, this.keyUpHandler, this);
        this.parent.observer.on(events.sourceCodeMouseDown, this.mouseUpHandler, this);
    };
    QuickToolbar.prototype.unWireInlineQTBarEvents = function () {
        this.parent.observer.off(events.mouseUp, this.mouseUpHandler);
        this.parent.observer.off(events.mouseDown, this.inlineQTBarMouseDownHandler);
        this.parent.observer.off(events.keyDown, this.keyDownHandler);
        this.parent.observer.off(events.keyUp, this.keyUpHandler);
        this.parent.observer.off(events.sourceCodeMouseDown, this.mouseUpHandler);
    };
    QuickToolbar.prototype.destroy = function () {
        if (this.linkQTBar) {
            EventHandler.remove(this.linkQTBar.element, 'mousedown', this.onMouseDown);
            this.linkQTBar.destroy();
        }
        if (this.imageQTBar) {
            EventHandler.remove(this.imageQTBar.element, 'mousedown', this.onMouseDown);
            this.imageQTBar.destroy();
        }
        if (this.tableQTBar) {
            EventHandler.remove(this.tableQTBar.element, 'mousedown', this.onMouseDown);
            this.tableQTBar.destroy();
        }
        if (this.inlineQTBar) {
            EventHandler.remove(this.inlineQTBar.element, 'mousedown', this.onMouseDown);
            if (isIDevice()) {
                EventHandler.remove(document, 'selectionchange', this.selectionChangeHandler);
            }
            this.inlineQTBar.destroy();
        }
        this.removeEventListener();
    };
    return QuickToolbar;
}());
export { QuickToolbar };
