// @syncfusion/ej2/ej2-resource/scripts/ej2-gantt/gantt/renderer/virtual-content-render.js
import { formatUnit, createElement } from '@syncfusion/ej2-base';
var VirtualContentRenderer = (function () {
    function VirtualContentRenderer(parent) {
        this.parent = parent;
    }
    VirtualContentRenderer.prototype.renderWrapper = function (height) {
        this.wrapper = createElement('div', { className: 'e-virtualtable', styles: "min-height:" + formatUnit(height) });
        this.parent.ganttChartModule.chartBodyContent.appendChild(this.wrapper);
        this.appendChildElements(this.parent.chartRowsModule.taskTable);
    };
    VirtualContentRenderer.prototype.appendChildElements = function (element) {
        this.wrapper.appendChild(element);
    };
    VirtualContentRenderer.prototype.adjustTable = function () {
        var content = this.parent.treeGrid.getContent().querySelector('.e-content').querySelector('.e-virtualtable');
        this.parent.ganttChartModule.virtualRender.wrapper.style.transform = content.style.transform;
    };
    return VirtualContentRenderer;
}());
export { VirtualContentRenderer };
