// @syncfusion/ej2/ej2-resource/scripts/ej2-gantt/blazor/sf-gantt-fn.js
import { Splitter } from './splitter';
var SfGantt = (function () {
    function SfGantt(element, options, dotnetRef) {
        this.toolbarHeight = 0;
        this.element = element;
        this.element.blazor_instance = this;
        this.dotNetRef = dotnetRef;
        this.options = options;
        this.initModules();
    }
    SfGantt.prototype.initModules = function () {
        this.splitterModule = new Splitter(this);
    };
    SfGantt.prototype.getOffsetRect = function (element) {
        var box = element.getBoundingClientRect();
        var scrollTop = window.pageYOffset || document.documentElement.scrollTop
            || document.body.scrollTop;
        var scrollLeft = window.pageXOffset || document.documentElement.scrollLeft ||
            document.body.scrollLeft;
        var clientTop = document.documentElement.clientTop || document.body.clientTop || 0;
        var clientLeft = document.documentElement.clientLeft || document.body.clientLeft || 0;
        var top = box.top + scrollTop - clientTop;
        var left = box.left + scrollLeft - clientLeft;
        return { top: Math.round(top), left: Math.round(left), width: box.width, height: box.height };
    };
    SfGantt.prototype.getParentElement = function (elem, selector, isID) {
        var parent = elem;
        while (parent) {
            if (isID ? parent.id === selector : parent.classList.contains(selector)) {
                break;
            }
            parent = parent.parentElement;
        }
        return parent;
    };
    return SfGantt;
}());
export { SfGantt };
