// @syncfusion/ej2/ej2-resource/scripts/ej2-gantt/blazor/treegrid.js
import { EventHandler } from '@syncfusion/ej2-base';
var TreeGrid = (function () {
    function TreeGrid(ganttParent, content) {
        this.previousScroll = { top: 0, left: 0 };
        this.parent = ganttParent;
        this.content = content;
        this.addEventListeners();
    }
    TreeGrid.prototype.addEventListeners = function () {
        EventHandler.add(this.content, 'scroll', this.scrollHandler, this);
    };
    TreeGrid.prototype.removeEventListeners = function () {
        EventHandler.remove(this.content, 'scroll', this.scrollHandler);
    };
    TreeGrid.prototype.scrollHandler = function (e) {
        if (this.content.scrollTop !== this.previousScroll.top) {
            this.parent.chartScrollModule.gridScrollHandler(this.content.scrollTop);
        }
        this.previousScroll.top = this.content.scrollTop;
    };
    TreeGrid.prototype.updateScrollTop = function (top) {
        this.content.scrollTop = top;
        this.previousScroll.top = this.content.scrollTop;
    };
    TreeGrid.prototype.destroy = function () {
        this.removeEventListeners();
    };
    return TreeGrid;
}());
export { TreeGrid };
